---
title: "Analysis of Poplar microbiome seasonal patterns"
output:
  html_notebook: default
---

# Set up and load libraries  

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_packages}
library(tidyverse)

library(qiime2R)

library(phyloseq)

library(vegan)

library(SRS)

library(hillR)

library(viridis)

library(cowplot)

library(igraph)

library(codyn)

library(psych)

library(lubridate)

library(mgcv)

library(pals)
```

# Generate tables  

## 16S  

### Compile 16S data as `phyloseq` object  

```{r 16s_qza_to_phyloseq}
# Create phyloseq object with ASVs and taxonomy
physeq_16s <- qza_to_phyloseq(
  
  features = "data/final_unfiltered_qzas/merged_table_16S.qza", 
  
  taxonomy = "data/final_unfiltered_qzas/tax_16S.qza"
  
)

# Get species for genotypes
host_species <- read_tsv("data/metadata/metadata_for_host_spp.txt") %>% 
  
  select(genotype, HostSpecies) %>% 
  
  distinct() %>% 
  
  filter(!is.na(genotype)) %>% 
  
  rename(host_species = "HostSpecies")

# Read in and format working metadata
metadata_16s <- read_csv("data/metadata/working_metadata_16S.csv") %>% 
  
  mutate(
    
    genotype = ifelse(sample_id == "Besccorv-R1-Oct2019-16S-LE", "BESC103", genotype), 
    
    genotype = ifelse(sample_id == "S6A-16S-LE-REDO", "BESC148", genotype), 
    
    genotype = ifelse(sample_id == "S11A-P4-Root-16S-RE", "BESC22", genotype), 
    
    replicate = ifelse(sample_id == "besc835-R-Jun2019-16S-LE", "R08", replicate), 
    
    sample_type = ifelse(sample_id == "S1F-16S-LE-REDO", "LE", sample_type)
    
  ) %>% 
  
  select(-note, -`...10`) %>% 
  
  inner_join(., host_species, by = "genotype") %>% 
  
  mutate(sample_id_for_physeq = sample_id) %>% 
  
  as.data.frame() %>% 
  
  column_to_rownames(var = "sample_id_for_physeq")

# Convert working metadata to sampledata object for phyloseq
metadata_16s_physeq <- sample_data(metadata_16s)

physeq_16s_full <- merge_phyloseq(physeq_16s, metadata_16s_physeq)

# Get sequence totals
physeq_16s_full_pre_filter_seqs <- sum(sample_sums(physeq_16s_full))

# Remove sequences that are not bacteria or archaea and drop ASVs with no seqs
physeq_16s_full <- subset_taxa(physeq_16s_full, Kingdom == "d__Bacteria" | Kingdom == "d__Archaea")

physeq_16s_full <- subset_taxa(physeq_16s_full, Order != "Chloroplast")

physeq_16s_full <- subset_taxa(physeq_16s_full, Family != "Mitochondria")

physeq_16s_full <- prune_taxa(taxa_sums(physeq_16s_full) > 0, physeq_16s_full)

# Get sequence totals
physeq_16s_full_post_filter_seqs <- sum(sample_sums(physeq_16s_full))

# Create phyloseq object with ASVs and taxonomy for RH redos
redo_16s <- qza_to_phyloseq(

  features = "data/final_unfiltered_qzas/16S_RH_redo_table.qza",

  taxonomy = "data/final_unfiltered_qzas/16S_RH_redo_taxonomy.qza"

)

# Format working metadata
metadata_redo_16s <- read_csv("data/metadata/rh_redo_metadata.csv") %>% 
  
  select(ID, sample_id, sample_code, sample_type) %>% 
  
  distinct() %>% 
  
  separate(
    
    col = sample_code, 
    
    into = c("genotype", "cutting_location", "replicate", "month", "year", "sample_type2", "host_species2")) %>% 
  
  inner_join(., host_species, by = "genotype") %>% 
  
  mutate(sample_id = str_replace(sample_id, "\\_", "\\-")) %>% 
  
  mutate(sample_id = paste(sample_id, "-redo-16S", sep = "")) %>% 
  
  select(-sample_type2, -host_species2) %>% 
  
  mutate(sample_id_for_physeq = sample_id) %>% 
  
  as.data.frame() %>% 
  
  column_to_rownames(var = "sample_id_for_physeq")

# Convert working metadata to sampledata object for phyloseq
metadata_16s_redo_physeq <- sample_data(metadata_redo_16s)

physeq_16s_redo <- merge_phyloseq(redo_16s, metadata_16s_redo_physeq)

# Get sequence totals
physeq_16s_redo_pre_filter_seqs <- sum(sample_sums(physeq_16s_redo))

# Remove sequences that are not bacteria or archaea and drop ASVs with no seqs
physeq_16s_redo <- subset_taxa(physeq_16s_redo, Kingdom == "d__Bacteria" | Kingdom == "d__Archaea")

physeq_16s_redo <- subset_taxa(physeq_16s_redo, Order != "Chloroplast")

physeq_16s_redo <- subset_taxa(physeq_16s_redo, Family != "Mitochondria")

physeq_16s_redo <- subset_samples(physeq_16s_redo, sample_id != "NTC-Amp1-redo-16S" & sample_id != "NTC-Amp2-redo-16S")

physeq_16s_redo <- prune_taxa(taxa_sums(physeq_16s_redo) > 0, physeq_16s_redo)

# Get sequence totals
physeq_16s_redo_post_filter_seqs <- sum(sample_sums(physeq_16s_redo))

# Merge complete dataset
physeq_16s_merged <- merge_phyloseq(physeq_16s_full, physeq_16s_redo)

physeq_16s_metadata <- as.data.frame(physeq_16s_merged@sam_data) %>% 
  
  as_tibble(rownames = "physeq_sample_id")

physeq_16s_asv <- as.data.frame(physeq_16s_merged@otu_table) %>% 
  
  as_tibble(rownames = "asv_id")

physeq_16s_tax <- as.data.frame(physeq_16s_merged@tax_table) %>% 
  
  as_tibble(rownames = "asv_id") %>% 
  
  rename(Domain = "Kingdom") %>% 
  
  mutate(Domain = str_remove(Domain, "d__"))

# Clean up environment
rm(
  
  physeq_16s, 
  
  redo_16s, 
  
  physeq_16s_full, 
  
  physeq_16s_redo, 
  
  physeq_16s_merged, 
  
  metadata_16s, 
  
  metadata_16s_physeq, 
  
  metadata_16s_redo_physeq, 
  
  metadata_redo_16s
  
)

# Free up memory
gc()
```

```{r 16s_sequence_totals}
# Get file names
bc_16s_rawseq_count_files <- list.files(
  
  path = "data/sample_counts/16S",
  
  full.names = TRUE
  
)

# Read in and combine
bc_16s_rawseq_counts <- map(
  
  bc_16s_rawseq_count_files,
  
  .f = read_tsv
  
) %>%
  
  bind_rows(.) %>%
  
  summarise(total_seqs = sum(`forward sequence count`)) %>%
  
  pull(total_seqs)

# Pre-filtered
bc_16s_pre_filter_seqs <- sum(
  
  physeq_16s_full_pre_filter_seqs,
  
  physeq_16s_redo_pre_filter_seqs
  
)

# Post-filtered
bc_16s_post_filter_seqs <- sum(
  
  physeq_16s_full_post_filter_seqs,
  
  physeq_16s_redo_post_filter_seqs
  
)

# Percentages
bc_16s_hq_perc <- 100 * (bc_16s_pre_filter_seqs / bc_16s_rawseq_counts)

bc_16s_filter_perc <- 100 * (bc_16s_post_filter_seqs / bc_16s_pre_filter_seqs)
```

## ITS  

### Compile ITS data as `phyloseq` object  

```{r its_qza_to_phyloseq}
# Create phyloseq object with ASVs and taxonomy
physeq_its <- qza_to_phyloseq(
  
  features = "data/final_unfiltered_qzas/merged_table_ITS.qza", 
  
  taxonomy = "data/final_unfiltered_qzas/taxonomy-ITS.qza"
  
)

# Read in and format working metadata
metadata_its <- read_csv("data/metadata/working_metadata_ITS.csv") %>% 
  
  rename(ID = "...1") %>% 
  
  mutate(
    
    genotype = ifelse(genotype == "BESC104" | genotype == "BESC105" | genotype == "BESC106" | genotype == "BESC107", "BESC103", genotype), 
    
    genotype = ifelse(sample_id == "Besccorv-R1-Oct2019-ITS-LE", "BESC103", genotype), 
    
    genotype = ifelse(sample_id == "S6A-ITS-LE-REDO_redo", "BESC148", genotype), 
    
    genotype = ifelse(sample_id == "S11A-P4-Root-ITS", "BESC22", genotype), 
    
    replicate = ifelse(sample_id == "besc835-R-Jun2019-ITS-LE", "R08", replicate), 
    
    sample_type = ifelse(sample_id == "Besc103-R24-Oct2019-ITS-LE", "LE", sample_type), 
    
    month = ifelse(sample_id == "Besc103-R23-Jun2018-ITS_redo", "June", month), 
    
    year = ifelse(sample_id == "Besc103-R23-Jun2018-ITS_redo", 2018, year)
    
  ) %>% 
  
  select(-note, -X) %>% 
  
  inner_join(., host_species, by = "genotype") %>% 
  
  mutate(sample_id_for_physeq = sample_id) %>% 
  
  as.data.frame() %>% 
  
  column_to_rownames(var = "sample_id_for_physeq")

# Convert working metadata to sampledata object for phyloseq
metadata_its_physeq <- sample_data(metadata_its)

physeq_its_full <- merge_phyloseq(physeq_its, metadata_its_physeq)

# Get sequence totals
physeq_its_full_pre_filter_seqs <- sum(sample_sums(physeq_its_full))

# Remove non-fungal sequences
physeq_its_full <- subset_taxa(physeq_its_full, Kingdom == "Fungi")

physeq_its_full <- prune_taxa(taxa_sums(physeq_its_full) > 0, physeq_its_full)

physeq_its_metadata <- as.data.frame(physeq_its_full@sam_data) %>% 
  
  as_tibble(rownames = "physeq_sample_id")

# Get sequence totals
physeq_its_full_post_filter_seqs <- sum(sample_sums(physeq_its_full))

# Create phyloseq object with ASVs and taxonomy for RH redos
redo_its <- qza_to_phyloseq(

  features = "data/final_unfiltered_qzas/ITS_RH_redo2_table.qza",

  taxonomy = "data/final_unfiltered_qzas/final_ITS_RH_redo2_taxonomy.qza"

)

# Format working metadata
metadata_redo_its <- read_csv("data/metadata/rh_redo_metadata2.csv") %>% 
  
  select(ID, sample_id, sample_code, sample_type) %>% 
  
  distinct() %>% 
  
  separate(
    
    col = sample_code, 
    
    into = c("genotype", "cutting_location", "replicate", "month", "year", "sample_type2", "host_species2")) %>% 
  
  inner_join(., host_species, by = "genotype") %>% 
  
  mutate(sample_id = str_replace(sample_id, "\\_", "\\-")) %>% 
  
  select(-sample_type2, -host_species2) %>% 
  
  mutate(sample_id_for_physeq = sample_id) %>% 
  
  as.data.frame() %>% 
  
  column_to_rownames(var = "sample_id_for_physeq")

# Convert working metadata to sampledata object for phyloseq
metadata_its_redo_physeq <- sample_data(metadata_redo_its)

physeq_its_redo <- merge_phyloseq(redo_its, metadata_its_redo_physeq)

# Get sequence totals
physeq_its_redo_pre_filter_seqs <- sum(sample_sums(physeq_its_redo))

# Remove sequences that are not fungi and drop ASVs with no seqs
physeq_its_redo <- subset_taxa(physeq_its_redo, Kingdom == "Fungi")

physeq_its_redo <- subset_samples(physeq_its_redo, sample_id != "NTC-Amp1-redo-ITS" & sample_id != "NTC-Amp2-redo-ITS")

physeq_its_redo <- prune_taxa(taxa_sums(physeq_its_redo) > 0, physeq_its_redo)

# Get sequence totals
physeq_its_redo_post_filter_seqs <- sum(sample_sums(physeq_its_redo))

# Merge complete dataset
physeq_its_merged <- merge_phyloseq(physeq_its_full, physeq_its_redo)

physeq_its_metadata <- as.data.frame(physeq_its_merged@sam_data) %>% 
  
  as_tibble(rownames = "physeq_sample_id")

physeq_its_asv <- as.data.frame(physeq_its_merged@otu_table) %>% 
  
  as_tibble(rownames = "asv_id")

physeq_its_tax <- as.data.frame(physeq_its_merged@tax_table) %>% 
  
  as_tibble(rownames = "asv_id")

# Clean up environment
rm(
  
  physeq_its, 
  
  redo_its, 
  
  physeq_its_full, 
  
  physeq_its_redo, 
  
  physeq_its_merged, 
  
  metadata_its, 
  
  metadata_its_physeq, 
  
  metadata_its_redo_physeq, 
  
  metadata_redo_its
  
)

# Free up memory
gc()
```

```{r its_sequence_totals}
# Get file names
bc_its_rawseq_count_files <- list.files(
  
  path = "data/sample_counts/ITS",
  
  full.names = TRUE
  
)

# Read in and combine
bc_its_rawseq_counts <- map(
  
  bc_its_rawseq_count_files,
  
  .f = read_tsv
  
) %>%
  
  bind_rows(.) %>%
  
  filter(`forward sequence count` >= 100) %>%
  
  summarise(total_seqs = sum(`forward sequence count`)) %>%
  
  pull(total_seqs)

# Pre-filtered
bc_its_pre_filter_seqs <- sum(
  
  physeq_its_full_pre_filter_seqs,
  
  physeq_its_redo_pre_filter_seqs
  
)

# Post-filtered
bc_its_post_filter_seqs <- sum(
  
  physeq_its_full_post_filter_seqs,
  
  physeq_its_redo_post_filter_seqs
  
)

# Percentage
bc_its_hq_perc <- 100 * (bc_its_pre_filter_seqs / bc_its_rawseq_counts)

bc_its_filter_perc <- 100 * (bc_its_post_filter_seqs / bc_its_pre_filter_seqs)
```

# Trim tables  

## Format ASV tables  

```{r functions_format_asv_tables}
#### Function for getting plant habitats from metadata abbreviations ####

plant_hab <- function(x) {
  
  tmp1 <- x
  
  if(tmp1 == "LE") {
    
    tmp2 <- "Leaf endosphere"
    
  } else if(tmp1 == "RE") {
    
    tmp2 <- "Root endosphere"
    
  } else if(tmp1 == "RH") {
    
    tmp2 <- "Rhizosphere"
    
  } else {
    
    tmp2 <- NA
    
  }
  
  return(tmp2)
  
}

#### Make unique sample id ####

make_unique_sample_id <- function(input.metadata) {
  
  # Save to temp object
  tmp1 <- input.metadata
  
  # Create unique sample ID
  tmp2 <- tmp1 %>% 
    
    unite(
      
      col = sample_unique_id, 
      
      genotype, cutting_location, replicate, month, year, sample_type, host_species, 
      
      remove = FALSE
      
    )
  
  # Return tibble
  return(tmp2)
  
}

#### Convert ASV table to long format ####

asv_pivot_longer <- function(x) {
  
  tmp1 <- x %>% 
    
    # Convert to long format
    pivot_longer(
      
      -asv_id, 
      
      names_to = "sample_id", 
      
      values_to = "n_seqs"
      
    )
  
  return(tmp1)
  
}
```

```{r format_asv_tables}
# Format metadata
bc_metadata <- list(physeq_16s_metadata, physeq_its_metadata) %>% 
  
  lapply(., make_unique_sample_id) %>% 
  
  set_names(nm = c("bc_16s_metadata", "bc_its_metadata"))

# Filter 2017 out of ASV tables
bc_asv <- list(physeq_16s_asv, physeq_its_asv) %>% 
  
  # Convert to long format for addition of metadata and downstream trimming
  lapply(asv_pivot_longer) %>% 
  
  # Add metadata and drop 2017 samples from ASV tables
  map2(bc_metadata, ., inner_join, by = "sample_id") %>% 
  
  set_names(nm = c("bc_16s_asv", "bc_its_asv")) %>% 
  
  lapply(., filter, year != 2017)

# Clean up environment
rm(
  
  physeq_16s_asv, 
  
  physeq_its_asv, 
  
  physeq_16s_metadata, 
  
  physeq_its_metadata
  
)

# Free up memory
gc()
```

## Trim duplicate samples  

```{r functions_trim_dups}
#### Get duplicate samples ####

get_dups <- function(x) {
  
  # x = metadata
  tmp1 <- x %>% 
    
    select(sample_id, sample_unique_id) %>% 
    
    group_by(sample_unique_id) %>% 
    
    filter(n() > 1) %>% 
    
    ungroup() %>% 
    
    pull(sample_unique_id) %>% 
    
    unique()
  
  # Return
  return(tmp1)
  
}

#### Get sequence counts of duplicated samples ####

dup_check <- function(asv_tab, dups) {
  
  tmp1 <- asv_tab %>% 
    
    filter(sample_unique_id %in% dups) %>% 
    
    group_by(sample_unique_id, sample_id) %>% 
    
    summarise(sample_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    group_by(sample_unique_id) %>% 
    
    mutate(max_n_seqs = max(sample_n_seqs)) %>% 
    
    ungroup()
  
}

#### Filter out duplicate samples with lowest seq counts ####

remove_dups <- function(asv_tab, dup_filter) {
  
  tmp1 <- asv_tab %>% 
    
    filter(!sample_id %in% dup_filter)
  
}

#### Drop samples with 0 seqs ####

drop_0seq_samples <- function(x) {
  
  tmp1 <- x %>% 
    
    group_by(sample_unique_id) %>% 
    
    mutate(sample_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    # Filter out samples with no seqs
    filter(sample_n_seqs > 0) %>% 
    
    select(-sample_n_seqs)
  
  return(tmp1)
  
}

#### Drop ASVs with 0 seqs ####

drop_0seq_asv <- function(x) {
  
  tmp1 <- x %>% 
    
    group_by(asv_id) %>% 
    
    mutate(asv_total = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    # Drop ASVs with no sequences after removal of duplicates
    filter(asv_total > 0) %>% 
    
    select(-asv_total)
  
  return(tmp1)
  
}

#### Get temporary metadata ####

temp_metadata <- function(x) {
  
  tmp1 <- x %>% 
    
    select(sample_unique_id, sample_id, genotype, cutting_location, replicate, month, year, sample_type, host_species) %>% 
    
    distinct()
  
  return(tmp1)
  
}
```

```{r trim_dups}
# Determine read counts of duplicated samples; create filter for sample removal
bc_dup_filter <- bc_metadata %>% 
  
  # Get duplicated samples
  map(., .f = get_dups) %>% 
  
  # Get sequence counts of duplicated samples
  map2(bc_asv, ., .f = dup_check) %>% 
  
  set_names(nm = c("bc_16s_asv_dups", "bc_its_asv_dups")) %>% 
  
  # Get duplicated samples with lowest sequence counts
  lapply(., filter, sample_n_seqs != max_n_seqs) %>% 
  
  lapply(., pull, sample_id)

# Filter out duplicate samples from ASV tables
bc_asv_no_dups <- map2(bc_asv, bc_dup_filter, .f = remove_dups) %>% 
  
  # Remove samples and ASVs with no sequences
  map(., .f = drop_0seq_samples) %>% 
  
  map(., .f = drop_0seq_asv) %>% 
  
  set_names(nm = c("bc_16s_asv_no_dups", "bc_its_asv_no_dups"))

# Get temporary metadata
bc_metadata_temp <- map(bc_asv_no_dups, .f = temp_metadata) %>% 
  
  set_names(nm = c("bc_16s_metadata_temp", "bc_its_metadata_temp"))

# Clean up environment
rm(
  
  bc_asv, 
  
  bc_dup_filter
  
)

# Free up memory
gc()
```

## Get tables for downstream analyses  

```{r functions_final_tables}
#### Get filter for extra samples (Aug 2019) ####

get_extra_filter <- function(asv.tab.nodups, metadata.temp) {
  
  tmp1 <- asv.tab.nodups %>% 
    
    group_by(sample_unique_id) %>% 
    
    summarise(sample_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    # Add metadata
    inner_join(metadata.temp, ., by = "sample_unique_id") %>% 
    
    arrange(sample_n_seqs) %>% 
    
    filter(year == 2019 & month == "Aug" & sample_type != "RH") %>%
    
    # filter(year == 2019 & month == "Aug") %>% 
    
    group_by(sample_type, year, month, genotype, cutting_location, host_species) %>% 
    
    mutate(total_samples = n()) %>% 
    
    ungroup() %>% 
    
    unite(sample_id2, sample_type, year, month, genotype, cutting_location, host_species, remove = FALSE) %>% 
    
    arrange(sample_id2, sample_n_seqs) %>% 
    
    filter(total_samples > 3) %>% 
    
    group_by(sample_id2) %>% 
    
    mutate(
      
      first_min = min(sample_n_seqs), 
      
      second_min = nth(sample_n_seqs, 2, order_by = sample_n_seqs), 
      
      third_min = nth(sample_n_seqs, 3, order_by = sample_n_seqs)
      
    ) %>% 
    
    ungroup() %>% 
    
    mutate(
      
      filter_key = NA, 
      
      filter_key = ifelse(total_samples >= 4 & sample_n_seqs == first_min, "remove", filter_key), 
      
      filter_key = ifelse(total_samples >= 5 & sample_n_seqs == second_min, "remove", filter_key), 
      
      filter_key = ifelse(total_samples == 6 & sample_n_seqs == third_min, "remove", filter_key)
      
    ) %>% 
    
    filter(filter_key == "remove") %>% 
    
    pull(sample_unique_id)
  
  # Return
  return(tmp1)
  
}

#### Remove extra samples ####

remove_extra <- function(asv.tab.nodups, extra.filter) {
  
  tmp1 <- asv.tab.nodups  %>% 
    
    filter(!sample_unique_id %in% extra.filter) %>% 
    
    select(sample_unique_id, asv_id, n_seqs)
  
  return(tmp1)
  
}

#### Format final filtered metadata ####

filter_metadata <- function(x) {
  
  tmp1 <- x %>% 
    
    select(sample_unique_id) %>% 
    
    distinct() %>% 
    
    separate(
      
      sample_unique_id, 
      
      into = c("genotype", "cutting_location", "replicate", "month", "year", "sample_type", "host_species"), 
      
      remove = FALSE
      
    ) %>% 
    
    mutate(
      
      plant_habitat = sapply(sample_type, plant_hab, simplify = TRUE), 
      
      plant_habitat = factor(
        
        plant_habitat, 
        
        levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
        
      ), 
      
      seasonality = paste(month, year, sep = " "), 
      
      seasonality = factor(
        
        seasonality, 
        
        levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018", 
                   "Feb 2019", "June 2019", "Aug 2019", "Oct 2019")
        
      )
      
    )
  
  # Return
  return(tmp1)
  
}

#### Get asv IDs ####

get_asvid <- function(x) {
  
  tmp1 <- x %>% 
    
    select(asv_id) %>% 
    
    distinct() %>% 
    
    pull(asv_id)
  
  return(tmp1)
  
}

#### Filter taxonomy ####

filter_taxonomy <- function(tax.tab, asv.id) {
  
  tmp1 <- tax.tab %>% 
    
    filter(asv_id %in% asv.id)
  
  return(tmp1)
  
}
```

```{r final_tables}
# Get filter for removal of extra samples
bc_extra_filter <- map2(bc_asv_no_dups, bc_metadata_temp, .f = get_extra_filter) %>% 
  
  set_names(nm = c("bc_16s_extra", "bc_its_extra"))

# Get filter to match 16S and ITS samples
bc_match_filter <- bc_metadata_temp$bc_16s_metadata_temp %>%

  filter(sample_type == "RH" & year == 2019 & month == "Aug") %>%

  select(sample_unique_id)

bc_match_filter_full <- bc_metadata_temp$bc_its_metadata_temp %>%

  filter(!(sample_type == "RH" & year == 2019 & month == "Aug")) %>%

  select(sample_unique_id) %>%

  bind_rows(., bc_match_filter) %>%

  pull(sample_unique_id)

# Final filtered ASV table
bc_asv_filtered <- map2(bc_asv_no_dups, bc_extra_filter, .f = remove_extra)

bc_asv_filtered$bc_its_asv_no_dups <- bc_asv_filtered$bc_its_asv_no_dups %>%

  filter(sample_unique_id %in% bc_match_filter_full)

bc_asv_filtered <- bc_asv_filtered %>% 
  
  # Remove samples and ASVs with no sequences
  map(., .f = drop_0seq_samples) %>% 
  
  map(., .f = drop_0seq_asv) %>% 
  
  set_names(nm = c("bc_16s_asv_filtered", "bc_its_asv_filtered"))

# Final metadata
bc_metadata_filtered <- map(bc_asv_filtered, .f = filter_metadata) %>% 
  
  set_names(nm = c("bc_16s_metadata_filtered", "bc_its_metadata_filtered"))

# Get unique ASV IDs
bc_asv_id_filter <- map(bc_asv_filtered, .f = get_asvid)

# Final taxonomy tables
bc_tax_filtered <- list(physeq_16s_tax, physeq_its_tax) %>% 
  
  map2(., bc_asv_id_filter, .f = filter_taxonomy) %>% 
  
  set_names(nm = c("bc_16s_tax_filtered", "bc_its_tax_filtered"))

# Clean up environment
rm(
  
  physeq_16s_tax, 
  
  physeq_its_tax, 
  
  bc_metadata, 
  
  bc_metadata_temp, 
  
  bc_asv_no_dups, 
  
  bc_asv_id_filter, 
  
  bc_extra_filter, 
  
  bc_match_filter, 
  
  bc_match_filter_full
  
)

# Free up memory
gc()
```

# Alpha diversity  

## Get sample sequence counts  

```{r functions_sample_seq_counts}
#### Calculate sample seq counts #####

sample_seq_counts <- function(x) {
  
  tmp1 <- x %>% 
    
    group_by(sample_unique_id) %>% 
    
    summarise(sample_n_seqs = sum(n_seqs)) %>% 
    
    ungroup()
  
  return(tmp1)
  
}
```

```{r sample_seq_counts}
# Sample counts
bc_sample_counts <- bc_asv_filtered %>% 
  
  map(., .f = sample_seq_counts) %>% 
  
  map2(bc_metadata_filtered, ., .f = inner_join, by = "sample_unique_id") %>% 
  
  set_names(nm = c("16S", "ITS")) %>% 
  
  map(., .f = arrange, sample_n_seqs) %>% 
  
  map(., .f = group_by, sample_type) %>% 
  
  map(., .f = group_split)  %>% 
  
  flatten(.) %>% 
  
  map(., .f = ungroup) %>% 
  
  set_names(nm = c("bc_16s_le_sample_counts", "bc_16s_re_sample_counts", "bc_16s_rh_sample_counts", "bc_its_le_sample_counts", "bc_its_re_sample_counts", "bc_its_rh_sample_counts"))
```

## Trim and subsample ASV tables 

```{r functions_trim_and_subsample_asv}
#### Trim to min sample depth ####

trim_to_seq_min <- function(asv.tab, sample.seq.min) {
  
  tmp1 <- asv.tab %>% 
    
    group_by(sample_unique_id) %>% 
    
    mutate(sample_n_seqs = sum(n_seqs)) %>%
    
    ungroup() %>% 
    
    # Remove samples with a total sample count of <7min seqs
    filter(sample_n_seqs > sample.seq.min)
  
  return(tmp1)
  
}

#### Calculate min sampling depth ####

min_sample_n_seqs <- function(sample.counts, sample.cutoff) {
  
  tmp1 <- sample.counts %>% 
    
    filter(sample_n_seqs > sample.cutoff)
  
  tmp2 <- tmp1 %>% 
    
    summarise(min_n_seqs = min(sample_n_seqs)) %>% 
    
    pull(min_n_seqs)
  
  return(tmp2)
  
}

#### ASV long to wide df ####

asv_long_to_wide_df_SRS <- function(x) {
  
  tmp1 <- x %>% 
    
    select(sample_unique_id, asv_id, n_seqs) %>% 
    
    pivot_wider(
      
      names_from = "sample_unique_id", 
      
      values_from = "n_seqs", 
      
      values_fill = 0
      
    ) %>% 
    
    as.data.frame() %>% 
    
    column_to_rownames(var = "asv_id")
  
  return(tmp1)
  
}

#### Scaling with ranked subsampling ####

srs2 <- function(asv.tab, seq.min) {
  
  tmp1 <- asv.tab
  
  tmp2 <- seq.min
  
  tmp3 <- SRS(asv.tab, Cmin = seq.min, set_seed = TRUE, seed = 12345)
  
  rownames(tmp3) <- rownames(tmp1)
  
  tmp4 <- as_tibble(tmp3, rownames = NA) %>% 
    
    rownames_to_column(var = "asv_id") %>% 
    
    pivot_longer(
      
      -asv_id,
      
      names_to = "sample_unique_id",
      
      values_to = "n_seqs"
      
    )
  
  # Remove ASVs with no seqs
  tmp5 <- drop_0seq_asv(tmp4) %>%
    
    pivot_wider(
      
      names_from = "asv_id",
      
      values_from = "n_seqs",
      
      values_fill = 0
      
    ) %>%
    
    as.data.frame() %>%
    
    column_to_rownames(var = "sample_unique_id")
  
  return(tmp5)
  
}
```

```{r trim_and_subsample_asv}
# Create list of minimum sample counts
bc_sample_seq_cutoffs <- list(500, 2400, 7000, 900, 500, 5000) %>% 
  
  set_names(nm = c("bc_16s_le_sample_seq_min", "bc_16s_re_sample_seq_min", "bc_16s_rh_sample_seq_min", "bc_its_le_sample_seq_min", "bc_its_re_sample_seq_min", "bc_its_rh_sample_seq_min"))

bc_sample_seq_min <- map2(bc_sample_counts, bc_sample_seq_cutoffs, .f = min_sample_n_seqs)

# Split asv tables by habitat
bc_asv_habsplit_df_SRS_input <- bc_asv_filtered %>% 
  
  map2(bc_metadata_filtered, ., .f = inner_join, by = "sample_unique_id") %>% 
  
  # Split by habitat
  map(., .f = group_by, sample_type) %>% 
  
  map(., .f = group_split)  %>% 
  
  flatten(.) %>% 
  
  map(., .f = ungroup) %>% 
  
  set_names(nm = c("bc_16s_le_asv_df", "bc_16s_re_asv_df", "bc_16s_rh_asv_df", "bc_its_le_asv_df", "bc_its_re_asv_df", "bc_its_rh_asv_df")) %>% 
  
  # Drop samples below minimum sequence depth
  map2(., bc_sample_seq_cutoffs, .f = trim_to_seq_min) %>% 
  
  # Drop ASVs with no seqs
  map(., .f = drop_0seq_asv) %>% 
  
  # Convert to wide format data frame
  map(., .f = asv_long_to_wide_df_SRS)

# Subsample with ranked scaling
bc_asv_habsplit_df_SRS_output <- map2(bc_asv_habsplit_df_SRS_input, bc_sample_seq_min, .f = srs2)
```

## Calculate Hill diversity indices

```{r functions_calculate_hill_div}
#### Hill diversity ####

hill_div <- function(x) {
  
  tmp1 <- hill_taxa(x, q = 0)
  
  tmp2 <- hill_taxa(x, q = 1)
  
  tmp3 <- hill_taxa(x, q = 2)
  
  tmp4 <- tibble(
    
    sample_unique_id = names(tmp1), 
    
    hill_index = "hill0", 
    
    hill = tmp1
    
  )
  
  tmp5 <- tibble(
    
    sample_unique_id = names(tmp2), 
    
    hill_index = "hill1", 
    
    hill = tmp2
    
  )
  
  tmp6 <- tibble(
    
    sample_unique_id = names(tmp3), 
    
    hill_index = "hill2", 
    
    hill = tmp3
    
  )
  
  tmp7 <- bind_rows(tmp4, tmp5, tmp6)
  
  return(tmp7)
  
}
```

```{r calculate_hill_div}
bc_metadata_habsplit_filtered <- bc_metadata_filtered %>% 
  
  # Split by habitat
  map(., .f = group_by, sample_type) %>% 
  
  map(., .f = group_split)  %>% 
  
  flatten(.) %>% 
  
  map(., .f = ungroup) %>% 
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh"))

bc_hill_div <- map(bc_asv_habsplit_df_SRS_output, .f = hill_div) %>% 
  
  map2(bc_metadata_habsplit_filtered, ., .f = inner_join, by = "sample_unique_id")
```

## Remove outliers  

```{r functions_remove_outliers_based_on_alpha_div}
#### Get hill div outlier ####

get_hill_outlier <- function(x) {
  
  tmp1 <- x %>% 
    
    group_by(hill_index) %>%
    
    filter(hill == max(hill)) %>% 
    
    ungroup() %>% 
    
    select(sample_unique_id) %>% 
    
    distinct() %>% 
    
    pull(sample_unique_id)
  
  return(tmp1)
  
}
```

```{r remove_outliers_based_on_alpha_div}
# Get outlier
bc_16s_hill_le_outlier <- get_hill_outlier(bc_hill_div$bc_16s_le)

# Drop outlier
bc_hill_div$bc_16s_le <- bc_hill_div$bc_16s_le %>% 
  
  filter(!sample_unique_id %in% bc_16s_hill_le_outlier)

# Get outlier
bc_its_hill_le_outlier <- get_hill_outlier(bc_hill_div$bc_its_le)

# Drop outlier
bc_hill_div$bc_its_le <- bc_hill_div$bc_its_le %>% 
  
  filter(!sample_unique_id %in% bc_its_hill_le_outlier)
```

## Hill diversity ANOVAs

```{r functions_hill_div_aov}
#### Hill anova ####

hill_aov <- function(x) {
  
  tmp1 <- aov(hill ~ host_species * year * month, data = x)
  
  return(tmp1)
  
}

#### Hill list anova ####

get_hill_div_aov_list <- function(x) {
  
  tmp1 <- x %>% 
    
    group_by(hill_index) %>% 
    
    group_split() %>% 
    
    set_names(nm = c("hill0", "hill1", "hill2"))
  
  tmp2 <- map(tmp1, .f = hill_aov)
  
  return(tmp2)
  
}

#### Hill list anova summary ####

get_hill_div_aov_summary_list <- function(x) {
  
  tmp1 <- x %>% 
    
    map(., .f = summary.aov)
  
  return(tmp1)
  
}

#### Tukey HSD ####

get_tukeyHSD <- function(x) {
  
  tmp1 <- map(x, .f = TukeyHSD)
  
  return(tmp1)
  
}

#### Get seasonality P values from ANOVA ####

get_seasonality_aov_Pval <- function(x) {
  
  tmp1 <- as.data.frame(x[[1]]) %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "Effect") %>% 
    
    mutate(Effect = str_replace_all(Effect, "\\s", "")) %>% 
    
    filter(Effect == "year:month") %>% 
    
    select(`Pr(>F)`)
  
  return(tmp1)
  
}

#### Get seasonality P values for Hill div ####

get_hill_div_seasonality_aov_Pval <- function(x) {
  
  tmp1 <- map(x, .f = get_seasonality_aov_Pval)
  
  tmp2 <- map_dfr(tmp1, .f = bind_rows) %>% 
    
    mutate(hill_index = names(tmp1)) %>% 
    
    mutate(pval_label = NA) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` < 0.001, "season%*%year*','~italic(P) < 0.001", pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` <= 0.05 & `Pr(>F)` >= 0.001, paste("season%*%year*','~italic(P) == ", round(`Pr(>F)`, 3), sep = ""), pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` > 0.05, "season%*%year*','~italic(n.s.)", pval_label))
  
  return(tmp2)
  
}

#### Hill Div CLD inner function ####

hill_div_cld_inner <- function(x, y) {
  
  tmp1 <- x$`year:month`$Letters
  
  tmp2 <- tibble(
    
    sample_type = names(tmp1), 
    
    cld = str_remove_all(tmp1, " "), 
    
    hill_index = y
    
  ) %>% 
    
    select(hill_index, everything())
  
  return(tmp2)
  
}

#### Hill div seasonality CLD updated function ####

hill_div_cld <- function(anova.model, tukeyhsd.model) {
  
  tmp1 <- map2(
    
    anova.model, tukeyhsd.model, 
    
    .f = multcompView::multcompLetters4
    
  )
  
  tmp2 <- names(anova.model)
  
  tmp3 <- map2(tmp1, tmp2, .f = hill_div_cld_inner) %>% 
    
    bind_rows() %>% 
    
    rename(timepoint = "sample_type") %>% 
    
    mutate(
      
      hill_label = NA, 
      
      hill_label = ifelse(hill_index == "hill0", "phantom()^0*italic(D)", hill_label), 
      
      hill_label = ifelse(hill_index == "hill1", "phantom()^1*italic(D)", hill_label), 
      
      hill_label = ifelse(hill_index == "hill2", "phantom()^2*italic(D)", hill_label)
      
    ) %>%
    
    separate(timepoint, into = c("year", "month"), sep = ":") %>%
    
    unite(seasonality, month, year, sep = " ") %>%
    
    mutate(seasonality = factor(seasonality, levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018",
                                                        "Feb 2019", "June 2019", "Aug 2019", "Oct 2019")))
  
  return(tmp3)
  
}
```

```{r hill_div_aov}
# Run ANOVA
bc_hill_div_aov <- map(bc_hill_div, .f = get_hill_div_aov_list)

# Get ANOVA summaries
bc_hill_div_aov_summary <- map(bc_hill_div_aov, .f = get_hill_div_aov_summary_list)

# Run Tukey's HSD
bc_hill_div_tukeyHSD <- map(bc_hill_div_aov, .f = get_tukeyHSD)

# Get month x year interaction P values
bc_hill_div_seasonality_Pval <- map(bc_hill_div_aov_summary, .f = get_hill_div_seasonality_aov_Pval)

# Get hill diversity CLD for plots
bc_hill_div_seasonality_cld <- map2(
  
  bc_hill_div_aov, 
  
  bc_hill_div_tukeyHSD, 
  
  .f = hill_div_cld)
```

### Hill diversity ANOVA, table S7  

```{r functions_hill_div_pvalues_and_table_S7}
#### Get hill div effect Pvalues ####

get_hill_div_aov_Pvalue <- function(aov.summary, hill.index, variable) {
  
  tmp1 <- aov.summary %>% 
    
    map(., .f = `[[`, 1) %>% 
    
    map(., .f = as_tibble, rownames = NA) %>% 
    
    map(., .f = rownames_to_column, var = "Variable") %>% 
    
    bind_rows(.id = "hill_index") %>% 
    
    mutate(Variable = str_squish(Variable))
  
  tmp2 <- tmp1 %>% 
    
    filter(hill_index == "hill2") %>% 
    
    filter(Variable == "year:month")
  
  return(tmp2)
  
}

#### Get hill div anova tables ####

get_hill_div_aov_tables_inner <- function(aov.summary) {
  
  tmp1 <- aov.summary %>% 
    
    map(., .f = `[[`, 1) %>% 
    
    map(., .f = as_tibble, rownames = NA) %>% 
    
    map(., .f = rownames_to_column, var = "Variable") %>% 
    
    bind_rows(.id = "hill_index") %>% 
    
    mutate(Variable = str_squish(Variable))
  
  return(tmp1)
  
}

#### Get supplemental hill div anova tables ####

get_hill_div_aov_tables <- function(anova.summary) {
  
  tmp1 <- map(anova.summary, .f = get_hill_div_aov_tables_inner) %>% 
    
    bind_rows(.id = "Plant habitat") %>% 
    
    mutate(
      
      `Plant habitat` = str_remove(`Plant habitat`, "bc_16s_"), 
      
      `Plant habitat` = str_remove(`Plant habitat`, "bc_its_"), 
      
      `Plant habitat` = toupper(`Plant habitat`)
      
    )
  
  tmp2 <- unique(tmp1$`Plant habitat`)
  
  tmp3 <- tmp1 %>% 
    
    rename(
      
      P = "Pr(>F)", 
      
      `Sum of sqs.` = "Sum Sq"
      
    ) %>% 
    
    mutate(
      
      `Sum of sqs.` = format(round(`Sum of sqs.`, 1), nsmall = 1), 
      
      `F value` = format(round(`F value`, 2), nsmall = 2), 
      
      P = ifelse(P >= 0.001, format(round(P, 3), nsmall = 3, scientific = FALSE), P), 
      
      P = ifelse(as.numeric(P) < 0.001, "< 0.001", P)
      
    ) %>% 
    
    mutate(across(.cols = everything(), .fns = str_trim)) %>% 
    
    mutate(
      
      `F value` = ifelse(`F value` == "NA", "", `F value`), 
      
      `F value` = ifelse(is.na(`F value`), "", `F value`), 
      
      P = ifelse(P == "NA", "", P), 
      
      P = ifelse(is.na(P), "", P)
      
    ) %>% 
    
    mutate(
      
      Variable = str_to_title(Variable), 
      
      Variable = str_replace_all(Variable, "\\:", " x "), 
      
      hill_index = ifelse(hill_index == "hill0", "q = 0", hill_index), 
      
      hill_index = ifelse(hill_index == "hill1", "q = 1", hill_index), 
      
      hill_index = ifelse(hill_index == "hill2", "q = 2", hill_index), 
      
    ) %>% 
    
    rename(`Hill index` = "hill_index") %>% 
    
    select(-`Mean Sq`) %>% 
    
    group_by(`Plant habitat`) %>% 
    
    group_split() %>% 
    
    set_names(nm = tmp2) %>% 
    
    map(., .f = ungroup) %>% 
    
    map(., select, -`Plant habitat`)
  
  return(tmp3)
  
}

#### P value cutoff ####

get_pval_cutoff <- function(x) {
  
  tmp1 <- max(x)
  
  if(tmp1 == 0.05) {
    
    tmp2 <- "= 0.05"
    
  } else if(tmp1 > 0.05) {
    
    tmp2 <- "> 0.05"
    
  } else if(tmp1 < 0.05 & tmp1 >= 0.01) {
    
    tmp2 <- "< 0.05"
    
  } else if(tmp1 < 0.01 & tmp1 > 0.001) {
    
    tmp2 <- "< 0.01"
    
  } else if(tmp1 <= 0.001) {
    
    tmp2 <- "< 0.001"
    
  } else {
    
    tmp2 <- NULL
    
  }
  
  return(tmp2)
  
}
```

```{r hill_div_pvalues_and_table_S7}
# Table S7 input, 16S
tableS7_input_16S <- get_hill_div_aov_tables(bc_hill_div_aov_summary[1:3]) %>% 
  
  bind_rows(.id = "Plant habitat") %>% 
  
  mutate(Community = "Bacteria/Archaea")

# Table S7 input
tableS7_input <- get_hill_div_aov_tables(bc_hill_div_aov_summary[4:6]) %>% 
  
  bind_rows(.id = "Plant habitat") %>% 
  
  mutate(Community = "Fungi") %>% 
  
  bind_rows(tableS7_input_16S, .) %>% 
  
  select(Community, everything()) %>% 
  
  mutate(
    
    `Plant habitat` = map_chr(`Plant habitat`, .f = plant_hab), 
    
    Variable = str_replace(Variable, "Host_species", "Host species"), 
    
    Variable = str_replace(Variable, "Month", "Season"), 
    
    Variable = str_replace(Variable, "month", "season")
    
  )

# Save
write_tsv(
  
  tableS7_input,
  
  file = "data/results/tableS7.txt"
  
)
```

## Plot Hill diversity  

```{r functions_plot_hill_div}
#### Prep hill div data ####

prep_hill_div_data <- function(hill.div, hill.index) {
  
  tmp1 <- hill.div %>% 
    
    bind_rows() %>% 
    
    filter(hill_index == hill.index) %>% 
    
    mutate(
      
      host_species = ifelse(host_species == "deltoides", "italic(P*'.'~deltoides)", host_species),
      
      host_species = ifelse(host_species == "trichocarpa", "italic(P*'.'~trichocarpa)", host_species)
      
    )
  
  return(tmp1)
  
}

#### Prep hill div CLD and P values ####

prep_hill_div_cld <- function(cld.data, hill.index) {
  
  tmp1 <- cld.data %>% 
    
    bind_rows(.id = "sample_type") %>% 
    
    mutate(
      
      sample_type = str_remove(sample_type, "bc_16s_"), 
      
      sample_type = str_remove(sample_type, "bc_its_"), 
      
      sample_type = toupper(sample_type)
      
    ) %>% 
    
    mutate(plant_habitat = map(sample_type, .f = plant_hab)) %>% 
    
    mutate(
      
      plant_habitat = factor(
        
        plant_habitat, 
        
        levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
        
      )
      
    ) %>% 
    
    filter(hill_index == hill.index)
  
  return(tmp1)
  
}

#### Seasonality labels ####

seasonality_labels <- function(x) {
  
  tmp0 <- as.character(x)
  
  if(tmp0 == "Feb 2018") {
    
    tmp1 <- "Winter 2018"
    
  } else if(tmp0 == "June 2018") {
    
    tmp1 <- "Spring 2018"
    
  } else if(tmp0 == "Aug 2018") {
    
    tmp1 <- "Summer 2018"
    
  } else if(tmp0 == "Oct 2018") {
    
    tmp1 <- "Fall 2018"
    
  } else if(tmp0 == "Feb 2019") {
    
    tmp1 <- "Winter 2019"
    
  } else if(tmp0 == "June 2019") {
    
    tmp1 <- "Spring 2019"
    
  } else if(tmp0 == "Aug 2019") {
    
    tmp1 <- "Summer 2019"
    
  } else if(tmp0 == "Oct 2019") {
    
    tmp1 <- "Fall 2019"
    
  } else {
    
    tmp1 <- x
    
  }
  
  return(tmp1)
  
}

#### Seasonality labels 2 ####

seasonality_labels2 <- function(x) {
  
  tmp1 <- map(x, .f = seasonality_labels)
  
  return(tmp1)
  
}

#### Plot hill diversity 2 ####

plot_hill_div2 <- function(hill.div, cdl.stats, stat.labels.P, plot.title, hill.index, community) {
  
  tmp0 <- hill.div %>% 
    
    group_by(plant_habitat, year, seasonality) %>% 
    
    summarise(
      
      hill_mean = mean(hill), 
      
      hill_se = se(hill)
      
    ) %>% 
    
    ungroup()
  
  tmp1 <- ggplot() + 
    
    scale_y_continuous(expand = expansion(mult = c(0.15, 0.15))) +
    
    scale_x_discrete(
      
      labels = function(x) seasonality_labels2(x)
      
    ) +
    
    geom_point(
      
      data = hill.div, 
      
      aes(x = seasonality, y = hill, colour = plant_habitat, fill = plant_habitat), 
      
      position = position_jitter(width = 0.2, seed = 12345), 
      
      pch = 21, 
      
      alpha = 0.5, 
      
      size = 1
      
    ) + 
    
    geom_boxplot(
      
      data = hill.div, 
      
      aes(x = seasonality, y = hill, colour = plant_habitat), 
      
      outlier.shape = NA,
      
      position = position_dodge(width = 0.9),
      
      fill = "white",
      
      alpha = 0.4,
      
      show.legend = FALSE
      
    ) +
    
    geom_line(
      
      data = tmp0, 
      
      aes(x = seasonality, y = hill_mean, group = year, colour = plant_habitat), 
      
      show.legend = FALSE
      
    ) + 
    
    # stat_summary(
    #   
    #   data = hill.div, 
    #   
    #   aes(x = seasonality, y = hill), 
    #   
    #   colour = "black",
    #   
    #   fun.data = mean_cl_normal,
    #   
    #   geom = "errorbar", 
    #   
    #   linewidth = 0.25, 
    #   
    #   width = 0
    #   
    # ) +
    
    stat_summary(
      
      data = hill.div, 
      
      aes(x = seasonality, y = hill, fill = plant_habitat), 
      
      colour = "black",
      
      fun = mean, 
      
      geom = "point", 
      
      pch = 21,
      
      size = 2
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    geom_text(
      
      data = cdl.stats,
      
      aes(x = seasonality, y = Inf, vjust = 1.55, label = cld),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    geom_text(
      
      data = stat.labels.P,
      
      aes(x = 3, y = -Inf, vjust = -0.25, label = pval_label),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    facet_wrap(
      
      ~ plant_habitat, 
      
      scales = "free_y", 
      
      nrow = 3, 
      
      ncol = 1
      
    ) + 
    
    labs(
      
      title = plot.title, 
      
      x = NULL, 
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )
  
  if(hill.index == "hill0" & community == "16s") {
    
    tmp2 <- tmp1 + 
      
      labs(y = bquote('Bacterial/archaeal alpha diversity ('*phantom()^0*italic(D)*')'))
    
  } else if(hill.index == "hill1" & community == "16s") {
    
    tmp2 <- tmp1 + 
      
      labs(y = bquote('Bacterial/archaeal alpha diversity ('*phantom()^1*italic(D)*')'))
    
  } else if(hill.index == "hill2" & community == "16s") {
    
    tmp2 <- tmp1 + 
      
      labs(y = bquote('Bacterial/archaeal alpha diversity ('*phantom()^2*italic(D)*')'))
    
  } else if(hill.index == "hill0" & community == "its") {
    
    tmp2 <- tmp1 + 
      
      labs(y = bquote('Fungal alpha diversity ('*phantom()^0*italic(D)*')'))
    
  } else if(hill.index == "hill1" & community == "its") {
    
    tmp2 <- tmp1 + 
      
      labs(y = bquote('Fungal alpha diversity ('*phantom()^1*italic(D)*')'))
    
  } else if(hill.index == "hill2" & community == "its") {
    
    tmp2 <- tmp1 + 
      
      labs(y = bquote('Fungal alpha diversity ('*phantom()^2*italic(D)*')'))
    
  } else {
    
    tmp2 <- tmp1
    
  }
  
  return(tmp2)
  
}

#### SE function ####

se <- function(x) {
  
  tmp1 <- sd(x) / sqrt(length(x))
  
  return(tmp1)
  
}
```

### Fig. S10  

```{r figS10}
# Fig. S12a data (bacteria/archaea)
bc_hill_div_data_figS10a <- prep_hill_div_data(bc_hill_div[1:3], "hill0")

# Fig. S12b data (fungi)
bc_hill_div_data_figS10b <- prep_hill_div_data(bc_hill_div[4:6], "hill0")

# Fig. S12a CLD (bacteria/archaea)
bc_hill_div_cld_figS10a <- prep_hill_div_cld(bc_hill_div_seasonality_cld[1:3], "hill0")

# Fig. S12b CLD (fungi)
bc_hill_div_cld_figS10b <- prep_hill_div_cld(bc_hill_div_seasonality_cld[4:6], "hill0")

# Fig. S12a P (bacteria/archaea)
bc_hill_div_pval_figS10a <- prep_hill_div_cld(bc_hill_div_seasonality_Pval[1:3], "hill0")

# Fig. S12b P (fungi)
bc_hill_div_pval_figS10b <- prep_hill_div_cld(bc_hill_div_seasonality_Pval[4:6], "hill0")

# Fig S12a
figS10a <- plot_hill_div2(
  
  bc_hill_div_data_figS10a, 
  
  bc_hill_div_cld_figS10a, 
  
  bc_hill_div_pval_figS10a, 
  
  "(a)", 
  
  "hill0", 
  
  "16s"
  
)

# Fig S12b
figS10b <- plot_hill_div2(
  
  bc_hill_div_data_figS10b, 
  
  bc_hill_div_cld_figS10b, 
  
  bc_hill_div_pval_figS10b, 
  
  "(b)", 
  
  "hill0", 
  
  "its"
  
)

# Get shared legend
figS10_legend <- get_legend(
  
  figS10a + 
    
    theme(legend.position = "bottom")
  
)

# Arrange figures in grid
figS10_grid <- plot_grid(
  
  figS10a + theme(legend.position = "none"), 
  
  figS10b + theme(legend.position = "none"), 
  
  align = 'vh', 
  
  nrow = 1, 
  
  ncol = 2
  
)

# Construct figure
figS10 <- plot_grid(figS10_grid, figS10_legend, ncol = 1, rel_heights = c(1, .033))

# Save figure
ggsave2(
  
  filename = "data/results/figS10.png", 
  
  plot = figS10, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS10.pdf", 
  
  plot = figS10, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)
```

### Fig. S9  

```{r figS9}
# Fig. S11a data (bacteria/archaea)
bc_hill_div_data_figS9a <- prep_hill_div_data(bc_hill_div[1:3], "hill1")

# Fig. S11b data (fungi)
bc_hill_div_data_figS9b <- prep_hill_div_data(bc_hill_div[4:6], "hill1")

# Fig. S11a CLD (bacteria/archaea)
bc_hill_div_cld_figS9a <- prep_hill_div_cld(bc_hill_div_seasonality_cld[1:3], "hill1")

# Fig. S11b CLD (fungi)
bc_hill_div_cld_figS9b <- prep_hill_div_cld(bc_hill_div_seasonality_cld[4:6], "hill1")

# Fig. S11a P (bacteria/archaea)
bc_hill_div_pval_figS9a <- prep_hill_div_cld(bc_hill_div_seasonality_Pval[1:3], "hill1")

# Fig. S11b P (fungi)
bc_hill_div_pval_figS9b <- prep_hill_div_cld(bc_hill_div_seasonality_Pval[4:6], "hill1")

# Fig S11a
figS9a <- plot_hill_div2(
  
  bc_hill_div_data_figS9a, 
  
  bc_hill_div_cld_figS9a, 
  
  bc_hill_div_pval_figS9a, 
  
  "(a)", 
  
  "hill1", 
  
  "16s"
  
)

# Fig S11b
figS9b <- plot_hill_div2(
  
  bc_hill_div_data_figS9b, 
  
  bc_hill_div_cld_figS9b, 
  
  bc_hill_div_pval_figS9b, 
  
  "(b)", 
  
  "hill1", 
  
  "its"
  
)

# Get shared legend
figS9_legend <- get_legend(
  
  figS9a + 
    
    theme(legend.position = "bottom")
  
)

# Arrange figures in grid
figS9_grid <- plot_grid(
  
  figS9a + theme(legend.position = "none"), 
  
  figS9b + theme(legend.position = "none"), 
  
  align = 'vh', 
  
  nrow = 1, 
  
  ncol = 2
  
)

# Construct figure
figS9 <- plot_grid(figS9_grid, figS9_legend, ncol = 1, rel_heights = c(1, .033))

# Save figure
ggsave2(
  
  filename = "data/results/figS9.png", 
  
  plot = figS9, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS9.pdf", 
  
  plot = figS9, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)
```

### Fig. S11  

```{r figS11}
# Fig. S13a data (bacteria/archaea)
bc_hill_div_data_figS11a <- prep_hill_div_data(bc_hill_div[1:3], "hill2")

# Fig. S13b data (fungi)
bc_hill_div_data_figS11b <- prep_hill_div_data(bc_hill_div[4:6], "hill2")

# Fig. S13a CLD (bacteria/archaea)
bc_hill_div_cld_figS11a <- prep_hill_div_cld(bc_hill_div_seasonality_cld[1:3], "hill2")

# Fig. S13b CLD (fungi)
bc_hill_div_cld_figS11b <- prep_hill_div_cld(bc_hill_div_seasonality_cld[4:6], "hill2")

# Fig. S13a P (bacteria/archaea)
bc_hill_div_pval_figS11a <- prep_hill_div_cld(bc_hill_div_seasonality_Pval[1:3], "hill2")

# Fig. S13b P (fungi)
bc_hill_div_pval_figS11b <- prep_hill_div_cld(bc_hill_div_seasonality_Pval[4:6], "hill2")

# Fig S13a
figS11a <- plot_hill_div2(
  
  bc_hill_div_data_figS11a, 
  
  bc_hill_div_cld_figS11a, 
  
  bc_hill_div_pval_figS11a, 
  
  "(a)", 
  
  "hill2", 
  
  "16s"
  
)

# Fig S13b
figS11b <- plot_hill_div2(
  
  bc_hill_div_data_figS11b, 
  
  bc_hill_div_cld_figS11b, 
  
  bc_hill_div_pval_figS11b, 
  
  "(b)", 
  
  "hill2", 
  
  "its"
  
)

# Get shared legend
figS11_legend <- get_legend(
  
  figS11a + 
    
    theme(legend.position = "bottom")
  
)

# Arrange figures in grid
figS11_grid <- plot_grid(
  
  figS11a + theme(legend.position = "none"), 
  
  figS11b + theme(legend.position = "none"), 
  
  align = 'vh', 
  
  nrow = 1, 
  
  ncol = 2
  
)

# Construct figure
figS11 <- plot_grid(figS11_grid, figS11_legend, ncol = 1, rel_heights = c(1, .033))

# Save figure
ggsave2(
  
  filename = "data/results/figS11.png", 
  
  plot = figS11, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS11.pdf", 
  
  plot = figS11, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)
```

## Alpha diversity by host species, Fig. S12  

```{r alpha_div_figS12_functions}
#### Plot hill diversity SI ####

hill_div_SI_fig <- function(hill.div) {
  
  tmp0 <- hill.div %>%
    
    group_by(plant_habitat, hill_label, host_species, year, seasonality) %>%
    
    summarise(
      
      hill_mean = mean(hill)
      
    ) %>%
    
    ungroup()
  
  tmp1 <- ggplot() +
    
    scale_x_discrete(
      
      labels = function(x) seasonality_labels2(x)
      
    ) +
    
    geom_point(
      
      data = hill.div, 
      
      aes(
        
        x = seasonality, 
        
        y = hill, 
        
        colour = plant_habitat,
        
        fill = plant_habitat, 
        
        group = host_species, 
        
        shape = host_species
        
      ), 
      
      position = position_jitterdodge(
        
        jitter.width = 0.2, 
        
        dodge.width = 0.5, 
        
        seed = 12345
        
      ),
      
      alpha = 0.2, 
      
      size = 1
      
    ) +
    
    geom_line(
      
      data = tmp0, 
      
      aes(
        
        x = seasonality, 
        
        y = hill_mean, 
        
        group = interaction(hill_label, host_species, year), 
        
        colour = plant_habitat
        
      ), 
      
      linetype = 1,
      
      position = position_dodge(width = 0.5), 
      
      show.legend = FALSE
      
    ) + 
    
    stat_summary(
      
      data = hill.div, 
      
      aes(
        
        x = seasonality, 
        
        y = hill, 
        
        # colour = plant_habitat, 
        
        group = host_species
        
      ), 
      
      position = position_dodge(width = 0.5), 
      
      fun.data = mean_cl_normal,
      
      geom = "errorbar", 
      
      colour = "black",
      
      linewidth = 0.5, 
      
      width = 0.2
      
    ) +
    
    stat_summary(
      
      data = hill.div,
      
      aes(
        
        x = seasonality,
        
        y = hill,
        
        fill = plant_habitat,
        
        #colour = plant_habitat,
        
        group = host_species, 
        
        shape = host_species
        
      ),
      
      colour = "black",
      
      position = position_dodge(width = 0.5), 
      
      fun = mean,
      
      geom = "point",
      
      size = 1.5
      
    ) +
    
    scale_shape_manual(
      
      name = "Host species",
      
      values = c(21, 24), 
      
      labels = function(x) parse(text = x)
      
    ) +
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.4, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      begin = 0.4, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    facet_grid(
      
      rows = vars(plant_habitat), 
      
      cols = vars(hill_label),
      
      scales = "free_y",
      
      labeller = label_parsed
      
    ) + 
    
    labs(
      
      title = NULL, 
      
      x = NULL, 
      
      y = bquote("Bacterial/archaeal"~alpha*"-diversity")
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    ) +
    
    guides(
      
      shape = guide_legend(override.aes = list(
        
        alpha = 0.75, 
        
        pch = c(21, 24),  
        
        size = 2.5, 
        
        colour = "black", 
        
        fill = "black"
        
      ))
      
    )
  
  return(tmp1)
  
}
```

```{r figS12}
# Get figS12 input data
figS12_input <- bc_hill_div %>%
  
  bind_rows(.id = "Community") %>% 
  
  mutate(
    
    Community = str_remove(Community, "bc_"),
    
    Community = str_remove(Community, "_le"),
    
    Community = str_remove(Community, "_re"),
    
    Community = str_remove(Community, "_rh")
    
  ) %>% 
  
  filter(Community == "16s" & sample_type != "LE" & hill_index != "hill0") %>% 
  
  mutate(
    
    hill_label = NA,
    
    hill_label = ifelse(hill_index == "hill1", "phantom()^1*italic(D)", hill_label),
    
    hill_label = ifelse(hill_index == "hill2", "phantom()^2*italic(D)", hill_label),
    
    plant_habitat = str_replace_all(plant_habitat, " ", "~"),
    
    plant_habitat = factor(plant_habitat, levels = c("Root~endosphere", "Rhizosphere"))
    
  )

# Get figure
figS12 <- hill_div_SI_fig(figS12_input)

# Save figure
ggsave2(
  
  filename = "data/results/figS12.png", 
  
  plot = figS12, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS12.pdf", 
  
  plot = figS12, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)
```

# Beta diversity  

## Remove outlier

```{r functions_remove_outlier_from_subsampled_asv_tables}
#### Convert ASV table to long format ####

asv_pivot_longer2 <- function(x) {
  
  tmp1 <- x %>% 
    
    # Convert to long format
    pivot_longer(
      
      -sample_unique_id, 
      
      names_to = "asv_id", 
      
      values_to = "n_seqs"
      
    )
  
}

#### ASV long to wide df 2 ####

asv_long_to_wide_df <- function(x) {
  
  tmp1 <- x %>% 
    
    select(sample_unique_id, asv_id, n_seqs) %>% 
    
    pivot_wider(
      
      names_from = "asv_id", 
      
      values_from = "n_seqs", 
      
      values_fill = 0
      
    ) %>% 
    
    as.data.frame() %>% 
    
    column_to_rownames(var = "sample_unique_id")
  
  return(tmp1)
  
}
```

```{r remove_outlier_from_subsampled_asv_tables}
# Remove outliers from subsampled 16S ASV tables
bc_asv_habsplit_df_SRS_output$bc_16s_le_asv_df <- bc_asv_habsplit_df_SRS_output$bc_16s_le_asv_df %>% 
  
  as_tibble(rownames = NA) %>% 
  
  rownames_to_column(var = "sample_unique_id") %>% 
  
  asv_pivot_longer2(.) %>% 
  
  filter(!sample_unique_id %in% bc_16s_hill_le_outlier) %>% 
  
  drop_0seq_asv(.) %>% 
  
  asv_long_to_wide_df(.)

# Remove outliers from subsampled ITS ASV tables
bc_asv_habsplit_df_SRS_output$bc_its_le_asv_df <- bc_asv_habsplit_df_SRS_output$bc_its_le_asv_df %>% 
  
  as_tibble(rownames = NA) %>% 
  
  rownames_to_column(var = "sample_unique_id") %>% 
  
  asv_pivot_longer2(.) %>% 
  
  filter(!sample_unique_id %in% bc_its_hill_le_outlier) %>% 
  
  drop_0seq_asv(.) %>% 
  
  asv_long_to_wide_df(.)
```

## Run PERMANOVA and PCoA  

### PERMANOVA  

```{r functions_bray_dist}
#### Bray Curtis distance matrices ####

get_bray_dist <- function(x) {
  
  tmp1 <- vegdist(x, method = "bray")
  
  return(tmp1)
  
}

#### PERMANOVA ####

permanova <- function(dist.mat, env.table) {
  
  # Create filter
  tmp1 <- rownames(as.matrix(dist.mat))
  
  # Order env data
  tmp2 <- env.table %>% 
    
    droplevels() %>% 
    
    filter(sample_unique_id %in% tmp1) %>% 
    
    arrange(match(sample_unique_id, tmp1))
  
  # Set seed
  set.seed(12345)
  
  # PERMANOVA
  tmp3 <- adonis2(
    
    dist.mat ~ host_species * year * month,
    
    data = tmp2,
    
    permutations = 9999, 
    
    parallel = 10
    
  )
  
  # Return
  return(tmp3)
  
}
```

```{r bray_dist}
# Bray curtis distance matrices
bc_dist <- map(bc_asv_habsplit_df_SRS_output, .f = get_bray_dist) %>% 
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh"))

# PERMANOVA
bc_permanova <- map2(bc_dist, bc_metadata_habsplit_filtered, .f = permanova)

write_rds(

  bc_permanova,

  file = "data/working/bc_permanova.rds"

)

bc_permanova <- read_rds(
  
  file = "data/working/bc_permanova.rds"
  
)
```

### PERMANOVA results, table S5  

```{r functions_tableS5}
#### Get PERMANOVA tables for Table S5 ####

get_permanova_table <- function(x) {
  
  tmp1 <- x %>% 
    
    bind_rows(.id = "permanova") %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "Variable") %>% 
    
    separate(
      
      col = Variable, 
      
      into = c("Variable", "remaining_number"), 
      
      sep = "\\..."
      
    ) %>% 
    
    select(-remaining_number) %>% 
    
    mutate(
      
      Variable = str_to_title(Variable), 
      
      Variable = str_replace_all(Variable, "\\:", " x "), 
      
      permanova = str_remove(permanova, "bc_")
      
    ) %>% 
    
    separate(
      
      col = permanova, 
      
      into = c("Community", "sample_type"), 
      
      sep = "_"
      
    ) %>% 
    
    mutate(
      
      sample_type = toupper(sample_type), 
      
      plant_habitat = sample_type, 
      
      Community = ifelse(Community == "16s", "Bacteria/Archaea", Community), 
      
      Community = ifelse(Community == "its", "Fungi", Community)
      
    ) %>% 
    
    select(
      
      Community, 
      
      plant_habitat, 
      
      Variable, 
      
      everything()
      
    ) %>% 
    
    select(-sample_type) %>% 
    
    rename(
      
      `Plant habitat` = "plant_habitat", 
      
      P = "Pr(>F)", 
      
      `Sum of sqs.` = "SumOfSqs", 
      
      `Pseudo-F` = "F"
      
    ) %>% 
    
    mutate(
      
      `Sum of sqs.` = format(round(`Sum of sqs.`, 1), nsmall = 1), 
      
      R2 = format(round(R2, 3), nsmall = 3), 
      
      `Pseudo-F` = format(round(`Pseudo-F`, digits = 2), nsmall = 2, scientific = FALSE), 
      
      P = ifelse(as.numeric(P) >= 0.001, format(round(P, digits = 3), nsmall = 3, scientific = FALSE), P),

      P = ifelse(as.numeric(P) < 0.001, "< 0.001", P)
      
    ) %>% 
    
    mutate(across(.cols = everything(), .fns = str_trim)) %>% 
    
    mutate(
      
      `Pseudo-F` = ifelse(`Pseudo-F` == "NA", "", `Pseudo-F`), 
      
      `Pseudo-F` = ifelse(is.na(`Pseudo-F`), "", `Pseudo-F`), 
      
      P = ifelse(P == "NA", "", P), 
      
      P = ifelse(is.na(P), "", P), 
      
      Variable = str_replace(Variable, "Host_species", "Host species"), 
      
      Variable = str_replace(Variable, "Month", "Season"), 
      
      Variable = str_replace(Variable, "month", "season")
      
    )
  
  return(tmp1)
  
}

#### Get permanova p values for manuscript ####

get_permanova_pval <- function(permanova.table, variable.label) {
  
  tmp1 <- permanova.table %>% 
    
    bind_rows(.id = "permanova") %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "Variable") %>% 
    
    filter(!is.na(`Pr(>F)`)) %>% 
    
    separate(
      
      col = Variable, 
      
      into = c("Variable", "remaining_number"), 
      
      sep = "\\..."
      
    ) %>% 
    
    select(-remaining_number) %>% 
    
    filter(Variable == variable.label) %>% 
    
    pull(`Pr(>F)`) %>% 
    
    max(.)
  
  return(tmp1)
  
}
```

```{r tableS5}
# Format table input
tableS5_input <- get_permanova_table(bc_permanova) %>% 
  
  mutate(`Plant habitat` = map_chr(`Plant habitat`, .f = plant_hab))

# Save
write_tsv(
  
  tableS5_input, 
  
  file = "data/results/tableS5.txt"
  
)
```

```{r temporalR2}
bc_temporal_R2 <- bc_permanova %>%
  
  map(., .f = as_tibble, rownames = NA) %>%
  
  map(., .f = rownames_to_column, var = "variable") %>%
  
  bind_rows(.id = "ID") %>%
  
  filter(variable == "year" | variable == "month" | variable == "year:month") %>%
  
  group_by(ID) %>%
  
  summarise(temporal_R2 = sum(R2)) %>%
  
  ungroup()
```

### PCoA  

```{r functions_PCoA}
#### PCoA ####

run_pcoa <- function(x) {
  
  tmp1 <- wcmdscale(x, eig = TRUE)
  
  return(tmp1)
  
}

#### PCoA eigenvalues ####

get_pcoa_axis_contributions <- function(x) {
  
  tmp1 <- 100 * (x$eig / sum(x$eig))
  
  tmp2 <- format(round(tmp1[1], 1), nsmall = 1)
  
  tmp3 <- format(round(tmp1[2], 1), nsmall = 1)
  
  tmp4 <- list(tmp2, tmp3) %>% 
    
    set_names(nm = c("eig1", "eig2"))
  
  return(tmp4)
  
}

#### PCoA loadings ####

get_pcoa_loadings <- function(pcoa.ord, metadata.table) {
  
  tmp1 <- scores(pcoa.ord) %>% 
    
    as_tibble(rownames = "sample_unique_id") %>% 
    
    inner_join(metadata.table, by = "sample_unique_id") %>% 
    
    droplevels()
  
  return(tmp1)
  
}

#### PCoA centroids ####

get_pcoa_centroids <- function(pcoa.ord, metadata.table) {
  
  tmp1 <- scores(pcoa.ord) %>% 
    
    as_tibble(rownames = "sample_unique_id") %>% 
    
    inner_join(metadata.table, by = "sample_unique_id") %>% 
    
    group_by(host_species, genotype, seasonality) %>%
    
    summarise(
      
      mean_cent1 = mean(Dim1), 
      
      mean_cent2 = mean(Dim2), 
      
      ebar_min1 = mean(Dim1) - se(Dim1), 
      
      ebar_max1 = mean(Dim1) + se(Dim1), 
      
      ebar_min2 = mean(Dim2) - se(Dim2), 
      
      ebar_max2 = mean(Dim2) + se(Dim2)
      
    ) %>% 
    
    ungroup() %>% 
    
    droplevels() %>% 
    
    mutate(
      
      host_species = ifelse(host_species == "deltoides", "italic(P*'.'~deltoides)", host_species),
      
      host_species = ifelse(host_species == "trichocarpa", "italic(P*'.'~trichocarpa)", host_species)
      
    )
  
  return(tmp1)
  
}

#### Get seasonality P values from PERMANOVA ####

get_seasonality_permanova_Pval <- function(x) {
  
  tmp1 <- as.data.frame(x) %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "Effect") %>% 
    
    mutate(Effect = str_replace_all(Effect, "\\s", "")) %>% 
    
    filter(Effect == "year:month") %>% 
    
    select(`Pr(>F)`)
  
  tmp2 <- tmp1 %>% 
    
    mutate(pval_label = NA) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` < 0.001, "season%*%year*','~italic(P) < 0.001", pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` <= 0.05 & `Pr(>F)` >= 0.001, paste("season%*%year*','~italic(P) == ", round(`Pr(>F)`, 3), sep = ""), pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` > 0.05, "season%*%year*','~italic(n.s.)", pval_label))
  
  return(tmp2)
  
}

#### Plot PCoA ####

plot_pcoa <- function(fig.data.means, axis.contributions, fig.title, stat.labels.P) {
  
  tmp1 <- axis.contributions$eig1
  
  tmp2 <- axis.contributions$eig2
  
  tmp3 <- ggplot() + 
    
    geom_errorbar(
      
      data = fig.data.means,
      
      aes(
        
        x = mean_cent1,
        
        y = mean_cent2,
        
        ymin = ebar_min2,
        
        ymax = ebar_max2,
        
        colour = seasonality
        
      ),
      
      width = 0,
      
      linewidth = 0.25
      
    ) +
    
    geom_errorbarh(
      
      data = fig.data.means,
      
      aes(
        
        y = mean_cent2,
        
        xmin = ebar_min1,
        
        xmax = ebar_max1,
        
        colour = seasonality
        
      ),
      
      height = 0,
      
      linewidth = 0.25
      
    ) +
    
    geom_point(
      
      data = fig.data.means, 
      
      aes(
        
        x = mean_cent1,
        
        y = mean_cent2,
        
        fill = seasonality,
        
        shape = host_species
        
      ),
      
      size = 2
      
    ) + 
    
    scale_fill_manual(
      
      name = NULL,
      
      values = pcoa_colors[c(1, 3, 6, 9:12, 14)],
      
      labels = function(x) seasonality_labels2(x)
      
    ) + 
    
    scale_colour_manual(
      
      name = NULL, 
      
      values = pcoa_colors[c(1, 3, 6, 9:12, 14)],
      
      guide = "none"
      
    ) +
    
    scale_shape_manual(
      
      name = NULL, 
      
      values = c(24, 21), 
      
      labels = function(x) parse(text = x)
      
    ) +
    
    geom_text(
      
      data = stat.labels.P,
      
      aes(x = -Inf, y = Inf, vjust = 1.5, hjust = -0.25, label = pval_label),
      
      parse = TRUE,
      
      size = 3
      
    ) +
    
    labs(
      
      title = fig.title, 
      
      x = bquote('PCo1 (' * .(tmp1) * '%)'),
      
      y = bquote('PCo2 (' * .(tmp2) * '%)')
      
    ) + 
    
    scale_y_continuous(expand = expansion(c(0.05, 0.1))) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      plot.title = element_text(colour = "black", face = "bold", size = 12, hjust = 0), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.ticks.x = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8), 
      
      axis.title.y = element_text(colour = "black", size = 10), 
      
      axis.title.x = element_text(colour = "black", size = 10), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8, hjust = 0), 
      
      plot.margin = margin(0.1, 0.5, 0.1, 0.1, "cm"), 
      
      legend.position = "bottom"
      
    ) +
    
    guides(
      
      shape = guide_legend(override.aes = list(
        
        alpha = 0.75, 
        
        pch = c(24, 21),  
        
        size = 2.5, 
        
        colour = "black", 
        
        fill = "black"
          
      ),
      
      nrow = 2,
      
      byrow = TRUE
      
      ),
      
      fill = guide_legend(override.aes = list(
        
        pch = 21,  
        
        size = 2.5, 
        
        colour = "black"
          
      ),
      
      nrow = 2,
      
      byrow = TRUE
      
      )
      
    )
  
  return(tmp3)
  
}
```

```{r PCoA}
# PCoA
bc_pcoa <- map(bc_dist, .f = run_pcoa)

# PCoA axis contributions
bc_pcoa_eig <- map(bc_pcoa, .f = get_pcoa_axis_contributions)

# PCoA sample loadings
bc_pcoa_loadings <- map2(bc_pcoa, bc_metadata_habsplit_filtered, .f = get_pcoa_loadings)

# PCoA loading centroids
bc_pcoa_centroids <- map2(bc_pcoa, bc_metadata_habsplit_filtered, .f = get_pcoa_centroids)

# Get February time points
feb_dates <- bc_pcoa_centroids$bc_16s_rh %>% 
  
  mutate(across(!c(host_species, seasonality), ~replace(., is.numeric(.), NA))) %>% 
  
  filter(seasonality == "Feb 2018" | seasonality == "Feb 2019") %>% 
  
  distinct()

# Add February time points to 16S LE data
bc_pcoa_centroids$bc_16s_le <- bc_pcoa_centroids$bc_16s_le %>% 
  
  bind_rows(., feb_dates) %>% 
  
  mutate(seasonality = factor(
    
    seasonality, 
    
    levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018", 
               "Feb 2019", "June 2019", "Aug 2019", "Oct 2019"))
    
  )

# Add February time points to ITS LE data
bc_pcoa_centroids$bc_its_le <- bc_pcoa_centroids$bc_its_le %>% 
  
  bind_rows(., feb_dates) %>% 
  
  mutate(seasonality = factor(
    
    seasonality, 
    
    levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018", 
               "Feb 2019", "June 2019", "Aug 2019", "Oct 2019"))
    
  )

# Get month x year P values from PERMANOVA
bc_permanova_seasonality_pval <- map(bc_permanova, .f = get_seasonality_permanova_Pval)

# Create a vector of figure titles
pcoa_fig_titles <- c("(a) Bacteria/archaea, leaf end.", "(c) Bacteria/archaea, root end.", "(e) Bacteria/archaea, rhizo.", 
                     "(b) Fungi, leaf end.", "(d) Fungi, root end.", "(f) Fungi, rhizo.")

pcoa_colors <- as.vector(parula(n = 14))

# Generate figures
bc_pcoa_figs <- pmap(
  
  list(bc_pcoa_centroids, bc_pcoa_eig, pcoa_fig_titles, bc_permanova_seasonality_pval), 
  
  .f = plot_pcoa
  
)
```

## Fig. 1  

```{r fig1}
# Get shared legend
fig1_legend <- get_legend(
  
  bc_pcoa_figs$bc_16s_le +
    
    theme(legend.position = "bottom")
  
)

# Arrange figures in grid
fig1_grid <- plot_grid(
  
  bc_pcoa_figs$bc_16s_le + theme(legend.position = "none"), 
  
  bc_pcoa_figs$bc_its_le + theme(legend.position = "none"), 
  
  bc_pcoa_figs$bc_16s_re + theme(legend.position = "none"), 
  
  bc_pcoa_figs$bc_its_re + theme(legend.position = "none"), 
  
  bc_pcoa_figs$bc_16s_rh + theme(legend.position = "none"), 
  
  bc_pcoa_figs$bc_its_rh + theme(legend.position = "none"), 
  
  align = 'vh', 
  
  nrow = 3, 
  
  ncol = 2
  
)

# Construct figure
fig1 <- plot_grid(fig1_grid, fig1_legend, ncol = 1, rel_heights = c(1, .067))

# Save figure
ggsave2(
  
  filename = "data/results/fig1.png", 
  
  plot = fig1, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 9, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/fig1.pdf", 
  
  plot = fig1, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 9, 
  
  units = "in"
  
)
```

# ASV summary data  

## ASV totals and seqs  

### Get overall sample sequence counts  

```{r sample_seq_counts_overall}
# Sample counts
bc_sample_counts_overall <- bc_asv_filtered %>% 
  
  map(., .f = sample_seq_counts) %>% 
  
  map2(bc_metadata_filtered, ., .f = inner_join, by = "sample_unique_id") %>% 
  
  set_names(nm = c("16S", "ITS")) %>% 
  
  map(., .f = arrange, sample_n_seqs) %>% 
  
  set_names(nm = c("bc_16s_sample_counts", "bc_its_sample_counts"))
```

### Subsample overall datasets to consistent depth  

```{r functions_subsample_to_consistent_depth_overall_datasets}
#### Drop outliers ####

drop_outlier <- function(asv.table, outlier.filter) {
  
  tmp1 <- asv.table %>% 
    
    filter(!sample_unique_id %in% outlier.filter)
  
  return(tmp1)
  
}
```

```{r subsample_to_consistent_depth_overall_datasets}
# Create list of minimum sample counts
bc_sample_seq_cutoffs_overall <- list(2000, 1000) %>% 
  
  set_names(nm = c("bc_16s_sample_seq_min", "bc_its_sample_seq_min"))

bc_sample_seq_min_overall <- map2(
  
  bc_sample_counts_overall, 
  
  bc_sample_seq_cutoffs_overall, 
  
  .f = min_sample_n_seqs
  
)

# Split asv tables by habitat
bc_asv_df_SRS_input_overall <- bc_asv_filtered %>% 
  
  # Drop outlier samples
  map2(., c(bc_16s_hill_le_outlier, bc_its_hill_le_outlier), .f = drop_outlier) %>% 
  
  map2(bc_metadata_filtered, ., .f = inner_join, by = "sample_unique_id")  %>% 
  
  set_names(nm = c("bc_16s_asv_df", "bc_its_asv_df")) %>% 
  
  # Drop samples below minimum sequence depth
  map2(., bc_sample_seq_cutoffs_overall, .f = trim_to_seq_min) %>% 
  
  # Drop ASVs with no seqs
  map(., .f = drop_0seq_asv) %>% 
  
  # Convert to wide format data frame
  map(., .f = asv_long_to_wide_df_SRS)

# Subsample with ranked scaling
bc_asv_df_SRS_output_overall <- map2(
  
  bc_asv_df_SRS_input_overall, 
  
  bc_sample_seq_min_overall, 
  
  .f = srs2
  
)

# Clean up memory
rm(bc_asv_df_SRS_input_overall)

gc()
```

## ASV counts  

### ASV totals, Table S2b  

```{r functions_tableS2b_asv_totals}
#### Get ASV totals ####

get_asv_totals <- function(asv.table, metadata.table, community, subsample) {
  
  # Format ASV table
  tmp1 <- asv.table %>% 
    
    as_tibble(., rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    inner_join(metadata.table, by = "sample_unique_id")
  
  # Get overall count
  tmp2 <- tmp1 %>% 
    
    select(asv_id) %>% 
    
    distinct() %>% 
    
    summarise(asv_total = n()) %>% 
    
    mutate(plant_habitat = "Overall") %>% 
    
    select(plant_habitat, asv_total)
  
  # Get ASV totals by habitat and format table
  tmp3 <- tmp1 %>%  
    
    group_by(plant_habitat, asv_id) %>% 
    
    mutate(asv_total = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    filter(asv_total > 0) %>% 
    
    select(asv_id, plant_habitat) %>% 
    
    distinct() %>% 
    
    group_by(plant_habitat) %>% 
    
    summarise(asv_total = n()) %>% 
    
    ungroup() %>% 
    
    bind_rows(., tmp2) %>% 
    
    mutate(plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere", "Overall")
      
    )) %>% 
    
    rename(
      
      `Plant-associated habitat` = "plant_habitat", 
      
      ASVs = "asv_total"
      
    ) %>% 
    
    mutate(Community = community) %>% 
    
    select(Community, everything()) %>% 
    
    mutate(ASVs = format(ASVs, big.mark = ","))
  
  # Return
  return(tmp3)
  
}
```

```{r tableS2b_asv_totals}
# Community
bc_asv_count_community <- c("Bacteria/Archaea", "Fungi")

# Get ASV counts when subsampled as overall data set
tableS2b_input <- pmap(
  
  list(
    
    bc_asv_df_SRS_output_overall, 
    
    bc_metadata_filtered, 
    
    bc_asv_count_community
    
  ), 
  
  .f = get_asv_totals
  
) %>% 
  
  bind_rows(.)

# Save
write_tsv(
  
  tableS2b_input,
  
  file = "data/results/tableS2b.txt"
  
)
```

### ASV totals subsampled by habitat, table S2c  

```{r functions_table_S2c_asv_totals}
#### Get ASV totals for subsampled dataset by habitat ####

get_habsplit_asv_total <- function(asv.table, community) {
  
  # Format ASV tables
  tmp1 <- tibble(
    
    ID = names(asv.table), 
    
    ASVs = map_dbl(asv.table, .f = ncol)
    
  ) %>% 
    
    mutate(
      
      ID = str_remove(ID, "bc_"), 
      
      ID = str_remove(ID, "_asv_df")
      
    ) %>% 
    
    separate(
      
      ID, 
      
      into = c("Community", "plant_habitat"), 
      
      sep = "_"
      
    ) %>% 
    
    mutate(
      
      Community = ifelse(Community == "16s", "Bacteria/Archaea", Community), 
      
      Community = ifelse(Community == "its", "Fungi", Community), 
      
      plant_habitat = ifelse(plant_habitat == "le", "Leaf endosphere", plant_habitat), 
      
      plant_habitat = ifelse(plant_habitat == "re", "Root endosphere", plant_habitat), 
      
      plant_habitat = ifelse(plant_habitat == "rh", "Rhizosphere", plant_habitat)
      
    )
  
  # Get total number of unique ASVs
  tmp2 <- unlist(map(asv.table, .f = colnames))
  
  tmp3 <- length(unique(tmp2))
  
  # Combine into single table
  tmp4 <- tibble(
    
    Community = community, 
    
    plant_habitat = "Overall", 
    
    ASVs = tmp3
    
  ) %>% 
    
    bind_rows(tmp1, .) %>% 
    
    mutate(plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere", "Overall")
      
    )) %>% 
    
    rename(`Plant-associated habitat` = "plant_habitat") %>% 
    
    mutate(ASVs = format(ASVs, big.mark = ","))
  
  # Return
  return(tmp4)
  
}
```

```{r tableS2c_asv_totals_subhab}
# Get ASV totals for communities subsampled by plant-associated habitat
tableS2c_input <- bind_rows(
  
  get_habsplit_asv_total(
    
    bc_asv_habsplit_df_SRS_output[1:3], bc_asv_count_community[1]
    
  ), 
  
  get_habsplit_asv_total(
    
    bc_asv_habsplit_df_SRS_output[4:6], bc_asv_count_community[2]
    
  )
  
)

# Save
write_tsv(
  
  tableS2c_input,
  
  file = "data/results/tableS2c.txt"
  
)
```

## Overall alpha diversity  

### Calculate overall alpha diversity  

```{r calculate_hill_div_overall}
# Hill div
bc_hill_div_overall <- map(bc_asv_df_SRS_output_overall, .f = hill_div) %>% 
  
  map2(bc_metadata_filtered, ., .f = inner_join, by = "sample_unique_id") %>% 
  
  set_names(nm = "bc_16s", "bc_its")
```

### Overall alpha diversity ANOVA  

```{r functions_hill_div_aov_overall}
#### Hill anova overall ####

hill_aov_overall <- function(x) {
  
  tmp1 <- aov(hill ~ sample_type * host_species, data = x)
  
  return(tmp1)
  
}

#### Hill list anova overall ####

get_hill_div_aov_list_overall <- function(x) {
  
  tmp1 <- x %>% 
    
    group_by(hill_index) %>% 
    
    group_split() %>% 
    
    set_names(nm = c("hill0", "hill1", "hill2"))
  
  tmp2 <- map(tmp1, .f = hill_aov_overall)
  
  return(tmp2)
  
}

#### Hill list anova summary ####

get_hill_div_aov_summary_list <- function(x) {
  
  tmp1 <- x %>% 
    
    map(., .f = summary.aov)
  
  return(tmp1)
  
}

#### Hill Div overall CLD inner function ####

hill_div_overall_cld_inner <- function(x, y) {
  
  tmp1 <- x$sample_type$Letters
  
  tmp2 <- tibble(
    
    sample_type = names(tmp1), 
    
    cld = str_remove_all(tmp1, " "), 
    
    hill_index = y
    
  ) %>% 
    
    select(hill_index, everything())
  
  return(tmp2)
  
}

#### Hill Div overall CLD ####

hill_div_overall_cld <- function(anova.model, tukeyhsd.model) {
  
  tmp1 <- map2(
    
    anova.model, tukeyhsd.model, 
    
    .f = multcompView::multcompLetters4
    
  )
  
  tmp2 <- names(anova.model)
  
  tmp3 <- map2(tmp1, tmp2, .f = hill_div_overall_cld_inner) %>% 
    
    bind_rows() %>% 
    
    mutate(plant_habitat = map_chr(sample_type, .f = plant_hab)) %>%
    
    mutate(plant_habitat = factor(
      
      plant_habitat,
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere"))
      
    ) %>%
    
    mutate(
      
      hill_label = NA, 
      
      hill_label = ifelse(hill_index == "hill0", "phantom()^0*italic(D)", hill_label), 
      
      hill_label = ifelse(hill_index == "hill1", "phantom()^1*italic(D)", hill_label), 
      
      hill_label = ifelse(hill_index == "hill2", "phantom()^2*italic(D)", hill_label)
      
    )
  
  return(tmp3)
  
}

#### Get HSD hill div p values for manuscript ####

get_tukeyHSD_hill_div_pvalues <- function(x) {
  
  tmp1 <- x
  
  tmp2 <- tmp1$hill0$sample_type %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "comparison")
  
  tmp3 <- tmp1$hill1$sample_type %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "comparison")
  
  tmp4 <- tmp1$hill2$sample_type %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "comparison")
  
  tmp5 <- bind_rows(tmp2, tmp3, tmp4)
  
  return(tmp5)
  
}
```

```{r hill_div_aov_overall}
# Run ANOVA
bc_hill_div_aov_overall <- map(
  
  bc_hill_div_overall, 
  
  .f = get_hill_div_aov_list_overall
  
)

# Get ANOVA summaries
bc_hill_div_aov_summary_overall <- map(
  
  bc_hill_div_aov_overall, 
  
  .f = get_hill_div_aov_summary_list
  
)

# Run Tukey's HSD
bc_hill_div_tukeyHSD_overall <- map(bc_hill_div_aov_overall, .f = get_tukeyHSD)

# Get CLD
bc_hill_div_cld_overall <- map2(
  
  bc_hill_div_aov_overall, 
  
  bc_hill_div_tukeyHSD_overall, 
  
  .f = hill_div_overall_cld
  
)

#### Get supplemental hill div anova overall tables ####

get_hill_div_aov_overall_tables <- function(anova.summary) {
  
  tmp1 <- map(anova.summary, .f = get_hill_div_aov_tables_inner) %>% 
    
    bind_rows(.id = "Community") %>% 
    
    mutate(
      
      Community = str_remove(Community, "bc_"), 
      
      Community = str_remove(Community, "bc_"), 
      
      Community = toupper(Community)
      
    )
  
  tmp2 <- tmp1 %>% 
    
    rename(
      
      P = "Pr(>F)", 
      
      `Sum of sqs.` = "Sum Sq"
      
    ) %>% 
    
    mutate(
      
      `Sum of sqs.` = format(round(`Sum of sqs.`, 1), nsmall = 1), 
      
      `F value` = format(round(`F value`, 2), nsmall = 2), 
      
      P = ifelse(P >= 0.001, format(round(P, 3), nsmall = 3, scientific = FALSE), P), 
      
      P = ifelse(as.numeric(P) < 0.001, "< 0.001", P)
      
    ) %>% 
    
    mutate(across(.cols = everything(), .fns = str_trim)) %>% 
    
    mutate(
      
      `F value` = ifelse(`F value` == "NA", "", `F value`), 
      
      `F value` = ifelse(is.na(`F value`), "", `F value`), 
      
      P = ifelse(P == "NA", "", P), 
      
      P = ifelse(is.na(P), "", P)
      
    ) %>% 
    
    mutate(
      
      Variable = str_to_title(Variable), 
      
      Variable = str_replace_all(Variable, "\\:", " x "), 
      
      hill_index = ifelse(hill_index == "hill0", "q = 0", hill_index), 
      
      hill_index = ifelse(hill_index == "hill1", "q = 1", hill_index), 
      
      hill_index = ifelse(hill_index == "hill2", "q = 2", hill_index), 
      
    ) %>% 
    
    rename(`Hill index` = "hill_index") %>% 
    
    select(-`Mean Sq`) %>%
    
    mutate(
      
      Variable = str_replace(Variable, "Sample_type", "Plant habitat"),
      
      Variable = str_replace(Variable, "sample_type", "plant habitat"),
      
      Variable = str_replace_all(Variable, "_", " ")
      
    )
  
  return(tmp2)
  
}
```

### Table S3  

```{r tableS3}
# Table S3 input, 16S
tableS3_input <- get_hill_div_aov_overall_tables(bc_hill_div_aov_summary_overall)

# Save
write_tsv(
  
  tableS3_input,
  
  file = "data/results/tableS3.txt"
  
)
```

### Overall Hill diversity, Fig. S3  

```{r functions_figS3}
#### Add Hill index label ####

add_hill_label <- function(x) {
  
  tmp1 <- x %>% 
    
    mutate(
      
      hill_label = NA, 
      
      hill_label = ifelse(hill_index == "hill0", "phantom()^0*italic(D)", hill_label), 
      
      hill_label = ifelse(hill_index == "hill1", "phantom()^1*italic(D)", hill_label), 
      
      hill_label = ifelse(hill_index == "hill2", "phantom()^2*italic(D)", hill_label)
      
    )
  
  return(tmp1)
  
}

#### Plot overall hill diversity ####

plot_overall_hill_div <- function(hill.div, cdl.stats, plot.title) {
  
  tmp1 <- ggplot() + 
    
    scale_y_continuous(expand = expansion(mult = c(0.05, 0.15))) + 
    
    geom_point(
      
      data = hill.div, 
      
      aes(x = plant_habitat, y = hill, colour = plant_habitat, fill = plant_habitat), 
      
      position = position_jitter(width = 0.2, seed = 12345), 
      
      pch = 21, 
      
      alpha = 0.5, 
      
      size = 1
      
    ) + 
    
    geom_boxplot(
      
      data = hill.div, 
      
      aes(plant_habitat, y = hill, colour = plant_habitat), 
      
      fill = "white", 
      
      outlier.shape = NA,
      
      alpha = 0.5, 
      
      show.legend = FALSE
      
    ) + 
    
    # stat_summary(
    #   
    #   data = hill.div, 
    #   
    #   aes(x = plant_habitat, y = hill), 
    #   
    #   fun.data = mean_cl_normal,
    #   
    #   geom = "errorbar", 
    #   
    #   colour = "black",
    #   
    #   linewidth = 0.5, 
    #   
    #   width = 0
    #   
    # ) +
    
    stat_summary(
      
      data = hill.div, 
      
      aes(x = plant_habitat, y = hill, fill = plant_habitat), 
      
      fun = mean, 
      
      colour = "black",
      
      geom = "point", 
      
      pch = 21,
      
      size = 1.5
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    geom_text(
      
      data = cdl.stats,
      
      aes(x = plant_habitat, y = Inf, vjust = 1.55, label = cld),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    facet_wrap(
      
      ~ hill_label, 
      
      labeller = label_parsed, 
      
      scales = "free_y"
      
    ) + 
    
    labs(
      
      title = plot.title, 
      
      x = NULL, 
      
      y = "Alpha diversity"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )
  
  return(tmp1)
  
}
```

```{r figS3}
# Format figure data
bc_hill_div_overall_fig_input <- map(
  
  bc_hill_div_overall, 
  
  .f = add_hill_label
  
)

# Figure titles
bc_hill_div_overall_fig_titles <- c("(a)", "(b)")

# Generate plots
bc_hill_div_overall_fig <- pmap(
  
  list(
    
    bc_hill_div_overall_fig_input, 
    
    bc_hill_div_cld_overall, 
    
    bc_hill_div_overall_fig_titles
    
  ), 
  
  .f = plot_overall_hill_div
  
)

# Compile figure
figS3 <- plot_grid(
  
  bc_hill_div_overall_fig$bc_16s, 
  
  bc_hill_div_overall_fig$bc_its, 
  
  align = 'vh', 
  
  nrow = 2, 
  
  ncol = 1
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS3.png", 
  
  plot = figS3, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 7, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS3.pdf", 
  
  plot = figS3, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 7, 
  
  units = "in"
  
)
```

## Overall beta diversity  

### Permanova overall  

```{r functions_tableS4_permanova}
#### PERMANOVA overall ####

permanova_overall <- function(dist.mat, env.table) {
  
  # Create filter
  tmp1 <- rownames(as.matrix(dist.mat))
  
  # Order env data
  tmp2 <- env.table %>% 
    
    droplevels() %>% 
    
    filter(sample_unique_id %in% tmp1) %>% 
    
    arrange(match(sample_unique_id, tmp1))
  
  # Set seed
  set.seed(12345)
  
  # PERMANOVA
  tmp3 <- adonis2(
    
    dist.mat ~ sample_type * host_species,
    
    data = tmp2,
    
    permutations = 9999, 
    
    parallel = 10
    
  )
  
  # Return
  return(tmp3)
  
}

#### Get overall PERMANOVA p value ####
get_permanova_overall_pval <- function(x) {
  
  tmp1 <- x$`Pr(>F)`[1]
  
  return(tmp1)
  
}
```

```{r tableS4_permanova}
# Bray curtis distance matrices
bc_dist_overall <- map(bc_asv_df_SRS_output_overall, .f = get_bray_dist) %>% 
  
  set_names(nm = c("bc_16s", "bc_its"))

# PERMANOVA
bc_permanova_overall <- map2(bc_dist_overall, bc_metadata_filtered, .f = permanova_overall)

write_rds(

  bc_permanova_overall,

  file = "data/working/bc_permanova_overall.rds"

)

# Read in
bc_permanova_overall <- read_rds(
  
  file = "data/working/bc_permanova_overall.rds"
  
)
```

### Overall PERMANOVA results, Table S4  

```{r functions_compile_tableS4}
#### Get PERMANOVA tables for Table S4 ####

get_permanova_overall_table <- function(permanova.table, community) {
  
  tmp1 <- permanova.table %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "Variable") %>% 
    
    mutate(
      
      Community = community, 
      
      Variable = ifelse(Variable == "sample_type", "Plant-associated habitat", Variable)
      
    ) %>% 
    
    select(Community, everything()) %>% 
    
    rename(
      
      P = "Pr(>F)", 
      
      `Sum of sqs.` = "SumOfSqs", 
      
      `Pseudo-F` = "F"
      
    ) %>% 
    
    mutate(
      
      `Sum of sqs.` = format(round(`Sum of sqs.`, 1), nsmall = 1), 
      
      R2 = format(round(R2, 3), nsmall = 3), 
      
      `Pseudo-F` = format(round(`Pseudo-F`, 2), nsmall = 2), 
      
      P = ifelse(P > 0.001, format(P, digits = 3), P), 
      
      P = ifelse(P <= 0.001, "< 0.001", P)
      
    ) %>% 
    
    mutate(across(.cols = everything(), .fns = str_trim)) %>% 
    
    mutate(
      
      `Pseudo-F` = ifelse(`Pseudo-F` == "NA", "", `Pseudo-F`), 
      
      `Pseudo-F` = ifelse(is.na(`Pseudo-F`), "", `Pseudo-F`), 
      
      P = ifelse(P == "NA", "", P), 
      
      P = ifelse(is.na(P), "", P)
      
    ) %>% 
    
    mutate(Variable = str_to_title(Variable)) %>% 
    
    mutate(
      
      Variable = str_replace(
        
        Variable,
        
        "Plant-Associated Habitat",
        
        "Plant-associated habitat"
        
      ),
      
      Variable = str_replace(
        
        Variable,
        
        "Sample_type",
        
        "Plant-associated habitat"
        
      ),
      
      Variable = str_replace(
        
        Variable,
        
        "Host_species",
        
        "Host species"
        
      ),
      
      Variable = str_replace(
        
        Variable,
        
        "host_species",
        
        "host species"
        
      ),
      
      Variable = str_replace_all(Variable, "\\:", " x ")
      
    )
  
  return(tmp1)
  
}
```

```{r compile_tableS4}
# Get PERMANOVA tables
tableS4_input <- map2(
  
  bc_permanova_overall,
  
  bc_asv_count_community,
  
  .f = get_permanova_overall_table
  
) %>% 
  
  bind_rows(.)

# Save
write_tsv(
  
  tableS4_input,
  
  file = "data/results/tableS4.txt"
  
)
```

## Fig. S4  

```{r functions_figS4_pcoa}
#### PCoA centroids overall ####

get_pcoa_centroids_overall <- function(pcoa.ord, metadata.table) {
  
  tmp1 <- scores(pcoa.ord) %>% 
    
    as_tibble(rownames = "sample_unique_id") %>% 
    
    inner_join(metadata.table, by = "sample_unique_id") %>% 
    
    group_by(plant_habitat, host_species, genotype) %>% 
    
    summarise(
      
      mean_cent1 = mean(Dim1), 
      
      mean_cent2 = mean(Dim2), 
      
      ebar_min1 = mean(Dim1) - se(Dim1), 
      
      ebar_max1 = mean(Dim1) + se(Dim1), 
      
      ebar_min2 = mean(Dim2) - se(Dim2), 
      
      ebar_max2 = mean(Dim2) + se(Dim2)
      
    ) %>% 
    
    ungroup() %>% 
    
    droplevels() %>%
    
    mutate(
      
      host_species = ifelse(host_species == "deltoides", "italic(P*'.'~deltoides)", host_species),
      
      host_species = ifelse(host_species == "trichocarpa", "italic(P*'.'~trichocarpa)", host_species)
      
    )
  
  return(tmp1)
  
}

#### Get plant habitat P values from PERMANOVA ####

get_planthab_permanova_Pval <- function(x) {
  
  tmp1 <- as.data.frame(x) %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "Effect") %>% 
    
    mutate(Effect = str_replace_all(Effect, "\\s", "")) %>% 
    
    filter(Effect == "sample_type") %>% 
    
    select(`Pr(>F)`)
  
  tmp2 <- tmp1 %>% 
    
    mutate(pval_label = NA) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` < 0.001, "plant-associated~habitat*','~italic(P) < 0.001", pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` <= 0.05 & `Pr(>F)` >= 0.001, paste("plant-associated~habitat*','~italic(P) == ", round(`Pr(>F)`, 3), sep = ""), pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` > 0.05, "plant-associated~habitat*','~italic(n.s.)", pval_label))
  
  return(tmp2)
  
}

#### Plot PCoA overall ####

plot_pcoa_overall <- function(
    
  fig.data.means,
  
  axis.contributions,
  
  fig.title,
  
  stat.labels.P
  
) {
  
  tmp0 <- fig.data.means
  
  tmp1 <- axis.contributions$eig1
  
  tmp2 <- axis.contributions$eig2
  
  tmp3 <- ggplot() + 
    
    geom_errorbar(
      
      data = tmp0,
      
      aes(
        
        x = mean_cent1,
        
        y = mean_cent2,
        
        ymin = ebar_min2,
        
        ymax = ebar_max2
        
      ),
      
      colour = "black",
      
      width = 0,
      
      linewidth = 0.5
      
    ) +
    
    geom_errorbarh(
      
      data = tmp0,
      
      aes(
        
        y = mean_cent2,
        
        xmin = ebar_min1,
        
        xmax = ebar_max1
        
      ),
      
      colour = "black",
      
      height = 0,
      
      linewidth = 0.5
      
    ) +
    
    geom_point(
      
      data = tmp0, 
      
      aes(
        
        x = mean_cent1,
        
        y = mean_cent2,
        
        fill = plant_habitat,
        
        shape = host_species
        
      ),
      
      colour = "black",
      
      size = 2
      
    ) +
    
    scale_shape_manual(
      
      name = "Host species",
      
      values = c(21, 24), 
      
      labels = function(x) parse(text = x)
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      name = "Plant-associated\nhabitat", 
      
      begin = 0.8, 
      
      end = 0
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    geom_text(
      
      data = stat.labels.P,
      
      aes(x = -Inf, y = Inf, vjust = 1.5, hjust = -0.25, label = pval_label),
      
      parse = TRUE,
      
      size = 3
      
    ) +
    
    labs(
      
      title = fig.title, 
      
      x = bquote('PCo1 (' * .(tmp1) * '%)'),
      
      y = bquote('PCo2 (' * .(tmp2) * '%)')
      
    ) + 
    
    scale_y_continuous(expand = expansion(c(0.05, 0.1))) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      plot.title = element_text(colour = "black", face = "bold", size = 12, hjust = 0), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.ticks.x = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8), 
      
      axis.title.y = element_text(colour = "black", size = 10), 
      
      axis.title.x = element_text(colour = "black", size = 10), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      plot.margin = margin(0.1, 0.5, 0.1, 0.1, "cm")
      
    ) +
    
    guides(
      
      shape = guide_legend(override.aes = list(
        
        alpha = 0.75, 
        
        pch = c(21, 24),  
        
        size = 2.5, 
        
        colour = "black", 
        
        fill = "black"
          
      )
      
      ),
      
      fill = guide_legend(override.aes = list(
        
        pch = 21,  
        
        size = 2.5, 
        
        colour = "black"
          
      )
      
      )
      
    )
  
  return(tmp3)
  
}
```

```{r figS4_pcoa}
# PCoA
bc_pcoa_overall <- map(bc_dist_overall, .f = run_pcoa)

# PCoA axis contributions
bc_pcoa_eig_overall <- map(bc_pcoa_overall, .f = get_pcoa_axis_contributions)

# PCoA sample loadings
bc_pcoa_loadings_overall <- map2(bc_pcoa_overall, bc_metadata_filtered, .f = get_pcoa_loadings)

# PCoA loading centroids
bc_pcoa_centroids_overall <- map2(bc_pcoa_overall, bc_metadata_filtered, .f = get_pcoa_centroids_overall)

# Get pant_habitat P values from PERMANOVA
bc_permanova_overall_pval <- map(bc_permanova_overall, .f = get_planthab_permanova_Pval)

# Create a vector of figure titles
pcoa_overall_fig_titles <- c("(a)", "(b)")

# Generate figures
bc_pcoa_overall_figs <- pmap(
  
  list(bc_pcoa_centroids_overall, bc_pcoa_eig_overall, pcoa_overall_fig_titles, bc_permanova_overall_pval), 
  
  .f = plot_pcoa_overall
  
)

# Get shared legends
figS4_legend1 <- get_legend(
  
  bc_pcoa_overall_figs$bc_16s + 
    
    theme(legend.position = "bottom") + 
    
    guides(fill = "none")
  
)

figS4_legend2 <- get_legend(
  
  bc_pcoa_overall_figs$bc_16s + 
    
    theme(legend.position = "bottom") + 
    
    guides(shape = "none")
  
)

# Arrange figures in grid
figS4_grid <- plot_grid(
  
  bc_pcoa_overall_figs$bc_16s + theme(legend.position = "none"), 
  
  bc_pcoa_overall_figs$bc_its + theme(legend.position = "none"), 
  
  align = 'vh', 
  
  nrow = 1, 
  
  ncol = 2
  
)

# Construct figure
figS4 <- plot_grid(
  
  figS4_grid,
  
  figS4_legend1,
  
  figS4_legend2,
  
  ncol = 1, 
  
  rel_heights = c(1, 0.12, 0.12)
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS4.png", 
  
  plot = figS4, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 4, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS4.pdf", 
  
  plot = figS4, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 4, 
  
  units = "in"
  
)
```

# Community change  

```{r bray_dist_between_seasons_functions}
#### Function to order pairs for duplicate detection ####

dup_pairs <- function(comm_1, comm_2) {
  
  tmp1 <- c(comm_1, comm_2)
  
  tmp2 <- sort(tmp1)
  
  tmp3 <- paste(tmp2[1], tmp2[2], sep = "_vs_")
  
  return(tmp3)
  
}

#### Get unique pairs of samples ####

get_unique_sample_pairs <- function(x) {
  
  tmp1 <- x %>% 
    
    as.matrix() %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "community_1") %>% 
    
    pivot_longer(
      
      -community_1, 
      
      names_to = "community_2", 
      
      values_to = "bray"
      
    ) %>% 
    
    select(-bray) %>% 
    
    mutate(community_pairs = map2_chr(community_1, community_2, dup_pairs)) %>% 
    
    filter(community_1 != community_2) %>% 
    
    distinct(community_pairs, .keep_all = TRUE)
  
  return(tmp1)
  
}

#### Get bray dist tibble ####

get_bray_dist_tibble <- function(dist.mat, unique.pairs) {
  
  tmp1 <- as.matrix(dist.mat) %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "community_1") %>% 
    
    pivot_longer(
      
      -community_1, 
      
      names_to = "community_2", 
      
      values_to = "bray"
      
    ) %>% 
    
    # Create ID for comparison
    unite(
      
      col = "community_pairs", 
      
      community_1, community_2, 
      
      remove = FALSE, 
      
      sep = "_vs_"
      
    ) %>% 
    
    select(-community_1, -community_2) %>% 
    
    inner_join(unique.pairs, ., by = "community_pairs") %>% 
    
    separate(
      
      col = community_1, 
      
      into = c("genotype_1", "cutting_location_1", "replicate_1", "month_1", "year_1", "sample_type_1", "host_species_1"), 
      
      sep = "_", 
      
      remove = FALSE
      
    ) %>% 
    
    separate(
      
      col = community_2, 
      
      into = c("genotype_2", "cutting_location_2", "replicate_2", "month_2", "year_2", "sample_type_2", "host_species_2"), 
      
      sep = "_", 
      
      remove = FALSE
      
    )
  
  return(tmp1)
  
}
```

```{r bray_dist_tibble}
# Get unique sample pairs
bc_sample_pairs <- map(bc_dist, .f = get_unique_sample_pairs)

# Get bray dist tibble
bc_bray_dist_tibble <- map2(
  
  bc_dist,
  
  bc_sample_pairs,
  
  .f = get_bray_dist_tibble
  
)
```

## Fig. 2b  

```{r community_rate_change_functions}
#### Community rate change bray-curtis ####

get_community_rate_change_bray <- function(dist.mat) {
  
  tmp1 <- dist.mat %>% 
    
    filter(genotype_1 == genotype_2) %>% 
    
    unite(
      
      seasonality_1, 
      
      month_1, 
      
      year_1, 
      
      sep = " ", 
      
      remove = FALSE
      
    ) %>% 
    
    unite(
      
      seasonality_2, 
      
      month_2, 
      
      year_2, 
      
      sep = " ", 
      
      remove = FALSE
      
    ) %>% 
    
    mutate(
      
      seasonality_1 = factor(
        
        seasonality_1, 
        
        levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018", 
                   "Feb 2019", "June 2019", "Aug 2019", "Oct 2019")
        
      ), 
      
      seasonality_2 = factor(
        
        seasonality_2, 
        
        levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018", 
                   "Feb 2019", "June 2019", "Aug 2019", "Oct 2019")
        
      )
      
    ) %>% 
    
    inner_join(., bc_sample_date1, by = "seasonality_1") %>% 
    
    inner_join(., bc_sample_date2, by = "seasonality_2") %>% 
    
    mutate(interval = abs(sample_day_1 - sample_day_2))
  
  return(tmp1)
  
}

#### Community rate change means ####

get_community_rate_change_mean <- function(x) {
  
  tmp1 <- x %>%
    
    group_by(sample_type_1, interval) %>%
    
    summarise(mean_bray = mean(bray)) %>%
    
    ungroup()
  
  return(tmp1)
  
}

#### Community rate change LM ####

community_rate_change_lm <- function(x) {
  
  tmp1 <- lm(mean_bray ~ interval, data = x)
  
  tmp2 <- summary(tmp1)
  
  return(tmp2)
  
}

#### Get community rate change LM p values ####

get_community_rate_change_lm_P <- function(x) {
  
  tmp1 <- x$coefficients[2, 4]
  
  tmp2 <- x$adj.r.squared
  
  tmp3 <- tibble(
    
    R2 = tmp2,
    
    P = tmp1
    
  )
  
  return(tmp3)
  
}
```

```{r community_rate_change}
# Get sample day
bc_sample_date <- tibble(
  
  year = c(rep(2018, 4), rep(2019, 4)), 
  
  month = rep(c("Feb", "June", "Aug", "Oct"), 2), 
  
  sample_date = c(
    
    "2018-02-19", "2018-06-07", "2018-08-22", "2018-10-04", 
    
    "2019-02-07", "2019-05-30", "2019-08-29", "2019-10-14"
    
  )
  
) %>% 
  
  mutate(
    
    sample_date =  ymd(sample_date), 
    
    julian_day = yday(sample_date), 
    
    sample_day = ifelse(year == 2019, julian_day + 365, julian_day)
    
  ) %>% 
  
  unite(
    
    col = "seasonality", 
    
    month, 
    
    year, 
    
    sep = " "
    
  ) %>% 
  
  mutate(
    
    seasonality = factor(
      
      seasonality, 
      
      levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018", 
                 "Feb 2019", "June 2019", "Aug 2019", "Oct 2019"))
    
  ) %>% 
  
  select(seasonality, sample_day)

# Time 1
bc_sample_date1 <- bc_sample_date %>% 
  
  rename(
    
    seasonality_1 = "seasonality", 
    
    sample_day_1 = "sample_day", 
    
  ) %>% 
  
  mutate(seasonality_1 = as.character(seasonality_1))

# Time 2
bc_sample_date2 <- bc_sample_date %>% 
  
  rename(
    
    seasonality_2 = "seasonality", 
    
    sample_day_2 = "sample_day", 
    
  ) %>% 
  
  mutate(seasonality_2 = as.character(seasonality_2))

# Get community rate change (bray curtis)
bc_community_rate_change_bray <- map(
  
  bc_bray_dist_tibble,
  
  .f = get_community_rate_change_bray
  
)

# Get means
bc_community_rate_change_bray_means <- map(
  
  bc_community_rate_change_bray,
  
  .f = get_community_rate_change_mean
  
)

# Combine community rate change data
bc_community_rate_change_bray_combined <- bc_community_rate_change_bray %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_asv_df")
    
  ) %>% 
  
  separate(
    
    col = id, 
    
    into = c("community", "sample_type")
    
  ) %>% 
  
  mutate(
    
    community = ifelse(community == "16s", "Bacteria and Archaea", "Fungi"), 
    
    sample_type = toupper(sample_type), 
    
    plant_habitat = map_chr(sample_type, plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Combine community rate change data
bc_community_rate_change_bray_means_combined <- bc_community_rate_change_bray_means %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_asv_df")
    
  ) %>% 
  
  separate(
    
    col = id, 
    
    into = c("community", "sample_type")
    
  ) %>% 
  
  mutate(
    
    community = ifelse(community == "16s", "Bacteria and Archaea", "Fungi"), 
    
    sample_type = toupper(sample_type), 
    
    plant_habitat = map_chr(sample_type, plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Community rate change LM
bc_community_rate_change_bray_lm <- map(
  
  bc_community_rate_change_bray_means, 
  
  .f = community_rate_change_lm
  
)

# Community rate change lm P values
bc_community_rate_change_lm_P <- map(
  
  bc_community_rate_change_bray_lm, 
  
  .f = get_community_rate_change_lm_P
  
) %>% 
  
  bind_rows(.id = "ID") %>% 
  
  mutate(
    
    ID = str_remove(ID, "bc_")
    
  ) %>% 
  
  separate(
    
    col = ID, 
    
    into = c("community", "sample_type")
    
  ) %>% 
  
  mutate(
    
    community = ifelse(community == "16s", "Bacteria and Archaea", "Fungi"), 
    
    sample_type = toupper(sample_type), 
    
    plant_habitat = map_chr(sample_type, plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  ) %>%
  
  mutate(
    
    P = ifelse(P >= 0.001, format(round(P, digits = 3), nsmall = 3, scientific = FALSE), "< 0.001"),
    
    R2 = format(round(R2, digits = 3), nsmall = 3),
    
    p_label = ifelse(
      
      P == "< 0.001",
      
      paste("R[adj.]^2 == ", R2, "*','~", "italic(P) ", P, sep = ""),
      
      paste("R[adj.]^2 == ", R2, "*','~", "italic(P) == ", P, sep = "")
      
    ),
    
    ypos = rep(c(0.25, 0.15, 0.05), 2)
    
  )
```

### Fig. 2b  

```{r fig2b_functions}
#### Plot community rate change ####

plot_community_rate_change <- function(
    
  fig.data,
  
  fig.data.means,
  
  pval.stats,
  
  plot.title
  
) {
  
  tmp1 <- ggplot() + 
    
    scale_y_continuous(
      
      limits = c(0, NA)
      
    ) + 
    
    scale_x_continuous(
      
      limits = c(0, NA)
      
    ) + 
    
    geom_point(
      
      data = fig.data, 
      
      aes(
        
        x = interval, 
        
        y = bray, 
        
        colour = plant_habitat, 
        
        fill = plant_habitat
        
      ), 
      
      pch = 21, 
      
      alpha = 0.1, 
      
      size = 1
      
    ) +
    
    stat_summary(
      
      data = fig.data, 
      
      aes(
        
        x = interval, 
        
        y = bray, 
        
        fill = plant_habitat,
        
        group = plant_habitat
        
      ), 
      
      fun = mean,
      
      geom = "point",
      
      colour = "black",
      
      pch = 21, 
      
      size = 2.5
      
    ) +
    
    geom_smooth(
      
      data = fig.data.means, 
      
      aes(
        
        x = interval, 
        
        y = mean_bray, 
        
        # colour = plant_habitat, 
        
        fill = plant_habitat,
        
        group = plant_habitat
        
      ), 
      
      method = "lm", 
      
      formula = y ~ x, 
      
      se = TRUE, 
      
      alpha = 0.3, 
      
      colour = "black",
      
      show.legend = FALSE,
      
      linewidth = 0.5
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      name = "Plant habitat", 
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0
      
    ) +
    
    geom_text(
      
      data = pval.stats,
      
      aes(x = -Inf, y = -Inf, vjust = -0.25, hjust = -0.3, label = p_label),
      
      parse = TRUE,
      
      size = 3
      
    ) +
    
    facet_grid(
      
      rows = vars(community),
      
      cols = vars(plant_habitat)
      
    ) + 
    
    labs(
      
      title = plot.title, 
      
      x = "Days", 
      
      y = "Bray-Curtis dissimilarity"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 9), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 8), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 9), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    ) + 
    
    guides(
      
      fill = guide_legend(override.aes = list(
        
        alpha = 1, 
        
        pch = 21, 
        
        size = 2.5, 
        
        colour = "black"
        
      ))
      
    )
  
  return(tmp1)
  
}
```

```{r fig2b}
# Get fig2a
fig2b <- plot_community_rate_change(
  
  bc_community_rate_change_bray_combined,
  
  bc_community_rate_change_bray_means_combined,
  
  bc_community_rate_change_lm_P,
  
  "(b)"
  
)
```

### Community change rate, Table S9  

```{r tableS9_functions}
#### Get rate change LM table ####

get_rate_change_lm_table <- function(x) {
  
  tmp1 <- x$coefficients %>% 
    
    as.data.frame(.) %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "variable")
  
}
```

```{r tableS9}
# Get input for Table S9
tableS9_input <- bc_community_rate_change_bray_lm %>% 
  
  map(., .f = get_rate_change_lm_table) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(id = str_remove(id, "bc_")) %>% 
  
  separate(
    
    id, 
    
    into = c("Community", "Plant habitat"), 
    
    sep = "_"
    
  ) %>% 
  
  rename(
    
    Variable = "variable", 
    
    P = "Pr(>|t|)"
    
  ) %>% 
  
  filter(Variable == "interval") %>% 
  
  mutate(
    
    Community = ifelse(Community == "16s", "Bacteria/Archaea", Community), 
    
    Community = ifelse(Community == "its", "Fungi", Community), 
    
    `Plant habitat` = toupper(`Plant habitat`), 
    
    `Plant habitat` = map(`Plant habitat`, .f = plant_hab), 
    
    `Plant habitat` = factor(
      
      `Plant habitat`, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    ), 
    
    Variable = "Time interval", 
    
    lci = Estimate - (qnorm(0.975) * `Std. Error`),
    
    uci = Estimate + (qnorm(0.975) * `Std. Error`),
    
    lci = format(round(365 * lci, digits = 3), nsmall = 3, scientific = FALSE), 
    
    uci = format(round(365 * uci, digits = 3), nsmall = 3, scientific = FALSE), 
    
    Estimate = format(round(365 * Estimate, digits = 3), nsmall = 3, scientific = FALSE), 
    
    P = ifelse(P >= 0.001, format(round(P, digits = 3), nsmall = 3, scientific = FALSE), "< 0.001"), 
    
    `t value` = format(round(`t value`, digits = 2), nsmall = 2)
    
  ) %>% 
  
  mutate(Slope = paste(Estimate, " [", lci, ", ", uci, "]", sep = "")) %>%
  
  select(Community, `Plant habitat`, Slope, `t value`, P)

# Save
write_tsv(
  
  tableS9_input,
  
  file = "data/results/tableS9.txt"
  
)
```

### Rate change correlations, Fig. S14  

```{r figS14_functions}
#### Rate change correlation function ####

get_rate_change_correlation_labels <- function(x) {
  
  tmp1 <- x %>%
    
    select(deltoides, trichocarpa)
  
  tmp2 <- corr.test(tmp1, method = "pearson")
  
  tmp3 <- tmp2$r %>%
    
    as.data.frame(.) %>%
    
    rownames_to_column(var = "host_species") %>%
    
    as_tibble(.) %>%
    
    filter(host_species == "deltoides") %>%
    
    select(host_species, trichocarpa)
  
  tmp4 <- ifelse(
    
    tmp2$p.adj < 0.001,
    
    "< 0.001",
    
    format(round(tmp2$p.adj, 3), nsmall = 3)
    
  )
  
  tmp5 <- tmp3 %>%
    
    mutate(
      
      P = tmp4,
      
      p_label = ifelse(
        
        P == "< 0.001",
        
        paste("r == ", format(round(trichocarpa, 3), nsmall = 3), "*','~", "italic(P) ", P, sep = ""),
        
        paste("r == ", format(round(trichocarpa, 3), nsmall = 3), "*','~", "italic(P) == ", P, sep = "")
        
      )
      
    )
  
  return(tmp5)
  
}
```

```{r figS14}
# Get figS14 input
figS14_input <- bc_community_rate_change_bray_combined %>%
  
  group_by(community, plant_habitat, host_species_1, interval) %>%
  
  summarise(mean_bray = mean(bray)) %>%
  
  ungroup() %>%
  
  pivot_wider(
    
    id_cols = c(community, plant_habitat, interval),
    
    names_from = "host_species_1",
    
    values_from = "mean_bray"
    
  )

# Correlation
figS14_corr_input <- figS14_input %>%

  select(deltoides, trichocarpa)

figS14_corr <- corr.test(figS14_corr_input, method = "pearson")

figS14_siglabel <- figS14_corr$r %>%

  as.data.frame(.) %>%

  rownames_to_column(var = "host_species") %>%

  as_tibble(.) %>%

  filter(host_species == "deltoides") %>%

  select(host_species, trichocarpa) %>%

  mutate(

    xpos = 0.6,

    ypos = 0.95,

    siglabel = paste("r==", round(trichocarpa, 2), "*','~", "italic(P)<0.001", sep = "")

  )

# Get figure
figS14 <- ggplot() +
  
  geom_abline(
    
    slope = 1,
    
    intercept = 0,
    
    linewidth = 0.5,
    
    linetype = 2
    
  ) +
  
  geom_point(
    
    data = figS14_input,
    
    aes(
      
      x = trichocarpa,
      
      y = deltoides,
      
      fill = plant_habitat,
      
      shape = community
      
    ),
    
    colour = "black", 
    
    size = 2,
    
    alpha = 0.6
    
  ) +
  
  scale_shape_manual(
    
    name = "Community",
    
    values = c(21, 22)
    
  ) +
  
  scale_colour_viridis(
    
    discrete = TRUE, 
    
    begin = 0.8, 
    
    end = 0, 
    
    guide = "none"
    
  ) + 
  
  scale_fill_viridis(
    
    name = "Plant habitat",
    
    discrete = TRUE, 
    
    begin = 0.8, 
    
    end = 0
    
  ) + 
  
  geom_text(
    
    data = figS14_siglabel,
    
    aes(x = xpos, y = ypos, label = siglabel),
    
    parse = TRUE,
    
    size = 4
    
  ) +
  
  geom_smooth(
    
    data = figS14_input,
    
    aes(
      
      x = trichocarpa,
      
      y = deltoides,
      
      colour = plant_habitat,
      
      group = interaction(community, plant_habitat)
      
    ),
    
    method = "lm",
    
    formula = y ~ x,
    
    se = FALSE
    
  ) +
  
  labs(
    
    title = NULL,
    
    x = bquote("Bray-Curtis dissimilarity ("~italic(P.~trichocarpa)*")"), 
    
    y = bquote("Bray-Curtis dissimilarity ("~italic(P.~deltoides)*")")
    
  ) +
  
  theme(
    
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
    
    panel.background = element_blank(), 
    
    panel.grid = element_blank(), 
    
    strip.background = element_blank(), 
    
    plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
    
    axis.title = element_text(colour = "black", size = 10), 
    
    axis.ticks = element_line(colour = "black", linewidth = 0.25), 
    
    axis.text.x = element_text(colour = "black", size = 8), 
    
    axis.text.y = element_text(colour = "black", size = 8), 
    
    strip.text = element_text(colour = "black", size = 10), 
    
    legend.key = element_blank(), 
    
    legend.title = element_text(colour = "black", size = 10), 
    
    legend.text = element_text(colour = "black", size = 8)
    
  ) +
  
  guides(
      
      shape = guide_legend(override.aes = list(
        
        alpha = 0.6, 
        
        pch = c(21, 22),  
        
        size = 2.5, 
        
        colour = "black", 
        
        fill = "black"
        
      )),
      
      fill = guide_legend(override.aes = list(
        
        alpha = 0.6, 
        
        pch = 21,  
        
        size = 2.5, 
        
        colour = "black"
        
      ))
      
    )

# Save figure
ggsave2(
  
  filename = "data/results/figS14.png", 
  
  plot = figS14, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS14.pdf", 
  
  plot = figS14, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 5, 
  
  units = "in"
  
)
```

# Environmental conditions  

## Read in environmental data  

```{r load_package_for_reading_in_env_data}
# Load
library(ncdf4)
```

```{r get_environmental_data_functions}
get_env_data <- function(env.file, variable.vec, year.vec) {
  
  tmp1 <- nc_open(env.file)
  
  tmp2 <- ncvar_get(tmp1, "lon", verbose = FALSE)
  
  tmp3 <- ncvar_get(tmp1, "lat", verbose = FALSE)
  
  tmp4 <- ncvar_get(tmp1, "time", verbose = FALSE)
  
  tmp5 <- ncvar_get(tmp1, variable.vec)
  
  tmp6 <- ncatt_get(tmp1, variable.vec, "_FillValue")
  
  nc_close(tmp1)
  
  tmp5[tmp5 == tmp6] <- NA
  
  tmp7 <- vector(mode = "numeric", length = length(tmp4))
  
  for(i in 1 : length(tmp4)) {
    
    tmp7[i] <- mean(tmp5[, , i])
    
  }
  
  tmp8 <- tibble(
    
    variable = variable.vec,
    
    year = year.vec,
    
    time_point = tmp4,
    
    value = tmp7
    
  )
  
  return(tmp8)
  
}

#### Environmental variable labels ####

get_env_label <- function(x) {
  
  tmp1 <- x %>%
    
    mutate(
      
      variable_label = NA, 
      
      variable_label = ifelse(variable == "dayl", "Day length", variable_label), 
      
      variable_label = ifelse(variable == "prcp", "Precip.", variable_label), 
      
      variable_label = ifelse(variable == "srad", "Short-wave rad.", variable_label), 
      
      variable_label = ifelse(variable == "tmax", "Max. temp.", variable_label),
      
      variable_label = ifelse(variable == "tmin", "Min. temp.", variable_label),
      
      variable_label = ifelse(variable == "vp", "Vapor press.", variable_label)
      
    ) %>%
    
    mutate(
      
      variable_label = factor(
        
        variable_label,
        
        levels = c("Day length", "Short-wave rad.", "Min. temp.", 
                   "Max. temp.", "Precip.", "Vapor press.")
        
      )
      
    )
  
  return(tmp1)
  
}
```

```{r get_environmental_data}
# Get list of files
env_files <- list.files(
  
  path = "data/environment",
  
  full.names = TRUE
  
)

# Vector of variables
env_variables <- c(
  
  rep("dayl", 3),
  
  rep("prcp", 3),
  
  rep("srad", 3),
  
  rep("tmax", 3),
  
  rep("tmin", 3),
  
  rep("vp", 3)
  
)

# Vector of years
env_year <- rep(c(2017, 2018, 2019), 6)

# Get env data
env <- pmap(
  
  list(
    
    env_files,
    
    env_variables,
    
    env_year
    
  ),
  
  .f = get_env_data
  
)
```

## Environmental time series, Fig. S2a  

```{r env_time_series_figS2a}
# Combine data
env_combined <- env %>%
  
  bind_rows(.) %>%
  
  filter(year != 2017) %>%
  
  group_by(variable) %>%
  
  mutate(sample_day = 1 : n()) %>%
  
  ungroup() %>%
  
  group_by(variable, year) %>%
  
  mutate(
    
    week = NA,
    
    week = c(rep(1 : 51, each = 7), rep(52, 8))
    
  ) %>%
  
  ungroup() %>%
  
  get_env_label(.)

# Figure
figS2a <- ggplot() +
  
  scale_x_continuous(
    
    breaks = c(0, 182, 365, 547, 730)
    
  ) +
  
  geom_vline(
    
    xintercept = 365,
    
    linetype = 2
    
  ) +
  
  geom_vline(
    
    data = bc_sample_date,
    
    aes(xintercept = sample_day, group = seasonality, colour = seasonality)
    
  ) +
  
  geom_line(
    
    data = env_combined,
    
    aes(x = sample_day, y = value),
    
    colour = "#2A406CFF",
    
    alpha = 0.4,
    
  ) +
  
  scale_colour_viridis(
    
    discrete = TRUE,
    
    option = "plasma",
    
    name = NULL
    
  ) + 
  
  facet_wrap(
    
    ~ variable_label,
    
    scales = "free_y"
    
  ) + 
  
  labs(
    
    title = "(a)",
    
    x = "Day",
    
    y = NULL
    
  ) +
  
  theme(
    
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
    
    panel.background = element_blank(), 
    
    panel.grid = element_blank(), 
    
    strip.background = element_blank(), 
    
    plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
    
    axis.title = element_text(colour = "black", size = 10), 
    
    axis.ticks = element_line(colour = "black", linewidth = 0.25), 
    
    axis.text.x = element_text(colour = "black", size = 8), 
    
    axis.text.y = element_text(colour = "black", size = 8), 
    
    strip.text = element_text(colour = "black", size = 10), 
    
    legend.key = element_blank(),

    legend.title = element_text(colour = "black", size = 10),

    legend.text = element_text(colour = "black", size = 8),

    legend.position = "bottom"
    
  ) +
  
  guides(

    colour = guide_legend(

      nrow = 2,

      byrow = TRUE

    )

  )
```

## Environment PCA, Fig. S2b  

```{r env_pca_functions}
#### Expand sample date ####

expand_sample_date <- function(x) {
  
  tmp1 <- tibble(
    
    seasonality = rep(x$seasonality, 7),
    
    sample_day = c((x$sample_day - 6) : x$sample_day)
    
  )
  
  return(tmp1)
  
}

#### Scale ####

scale_values <- function(x) {
  
  tmp1 <- (x - min(x)) / (max(x) - min(x))
  
  return(tmp1)
  
}
```

```{r env_pca}
# Expand sample date
bc_sample_date_expanded <- bc_sample_date %>%
  
  group_by(seasonality) %>%
  
  group_split(.) %>%
  
  map(., .f = ungroup) %>%
  
  map(., .f = expand_sample_date) %>%
  
  bind_rows(.)

# Get PCA input
env_pca_input <- env_combined %>%
  
  full_join(., bc_sample_date_expanded, by = "sample_day") %>%
  
  filter(!is.na(seasonality)) %>%
  
  group_by(variable, seasonality) %>%
  
  mutate(
    
    rep_number = 1 : n()
    
  ) %>%
  
  ungroup() %>%
  
  unite(
    
    col = "sample_id",
    
    seasonality,
    
    rep_number,
    
    sep = "_",
    
    remove = FALSE
    
  )

# Trim
env_pca_input_trimmed <- env_pca_input %>%
  
  group_by(variable) %>%
  
  mutate(scaled_value = scale_values(value)) %>%
  
  ungroup() %>%
  
  select(sample_id, variable, scaled_value) %>%
  
  pivot_wider(
    
    id_cols = sample_id,
    
    names_from = "variable",
    
    values_from = "scaled_value"
    
  ) %>%
  
  column_to_rownames(var = "sample_id")

# Run PCA
env_pca <- rda(env_pca_input_trimmed, scale = TRUE)

# Summary
env_pca_summary <- summary(env_pca)

# Format eigenvalues
env_pca_pc1_eig <- round(100 * (as.numeric(env_pca_summary$cont$importance[ ,1][2])), 0)

env_pca_pc2_eig <- round(100 * (as.numeric(env_pca_summary$cont$importance[ ,2][2])), 0)

## Format plot loadings
env_pca_points <- as.data.frame(env_pca_summary$sites) %>% 
  
  rownames_to_column(var = "sample_id") %>% 
  
  as_tibble(.) %>%
  
  inner_join(., env_pca_input, by = "sample_id") %>%
  
  group_by(seasonality) %>%
  
  summarise(
    
    pc1_mean = mean(PC1),
    
    pc2_mean = mean(PC2),
    
    pc1_se = se(PC1),
    
    pc2_se = se(PC2)
    
  ) %>%
  
  ungroup()

## Format variable loadings
env_pca_variables <- as.data.frame(env_pca_summary$species) %>% 
  
  rownames_to_column(var = "variable") %>% 
  
  as_tibble(.) %>% 
  
  mutate(
    
    variable_label = NA, 
    
    variable_label = ifelse(variable == "dayl", "Day length", variable_label), 
    
    variable_label = ifelse(variable == "prcp", "Precip.", variable_label), 
    
    variable_label = ifelse(variable == "srad", "Short-wave rad.", variable_label), 
    
    variable_label = ifelse(variable == "tmax", "Max. temp.", variable_label),
    
    variable_label = ifelse(variable == "tmin", "Min. temp.", variable_label),
    
    variable_label = ifelse(variable == "vp", "Vapor press.", variable_label)
    
  ) %>% 
  
  mutate(
    
    PC1.NEW = NA,
    
    PC2.NEW = NA,
    
    PC1.NEW = ifelse(variable == "dayl", PC1 + 0.3, PC1.NEW),
    
    PC2.NEW = ifelse(variable == "dayl", PC2 - 0.1, PC2.NEW),
    
    PC1.NEW = ifelse(variable == "prcp", PC1 - 0.1, PC1.NEW),
    
    PC2.NEW = ifelse(variable == "prcp", PC2 - 0.1, PC2.NEW),
    
    PC1.NEW = ifelse(variable == "srad", PC1 + 0.25, PC1.NEW),
    
    PC2.NEW = ifelse(variable == "srad", PC2 + 0.1, PC2.NEW),
    
    PC1.NEW = ifelse(variable == "tmax", PC1 + 0.3, PC1.NEW),
    
    PC2.NEW = ifelse(variable == "tmax", PC2 + 0.1, PC2.NEW),
    
    PC1.NEW = ifelse(variable == "tmin", PC1 + 0.3, PC1.NEW),
    
    PC2.NEW = ifelse(variable == "tmin", PC2 + 0.1, PC2.NEW),
    
    PC1.NEW = ifelse(variable == "vp", PC1 + 0.3, PC1.NEW),
    
    PC2.NEW = ifelse(variable == "vp", PC2 - 0.1, PC2.NEW)
    
  )
```

### Env distmat and PERMANOVA  

```{r env_distmat}
# Distance matrix
env_dist <- vegdist(env_pca_input_trimmed, method = "euclidean")

# Env table
env_seasonality <- env_pca_input %>%
  
  select(sample_id, seasonality) %>%
  
  distinct()

env_order <- rownames(as.matrix(env_dist))

env_table <- env_pca_input_trimmed %>%
  
  rownames_to_column(var = "sample_id") %>%
  
  inner_join(., env_seasonality, by = "sample_id") %>%
  
  arrange(match(sample_id, env_order)) %>%
  
  separate(
    
    seasonality,
    
    into = c("month", "year"),
    
    sep = " ",
    
    remove = FALSE
    
  )

# PERMANOVA
set.seed(12345)

env_pma <- adonis2(env_dist ~ month * year, data = env_table, permutations = 9999)

env_pma_P <- env_pma$`Pr(>F)`[1] %>%
  
  get_pval_cutoff(.)

env_pma_R2 <- format(round(env_pma$R2[1], 3), nsmall = 3)

env_pma_sig <- tibble(
  
  p_label = paste("atop(R^2 == ", env_pma_R2, ", ", "italic(P) ", env_pma_P, ")", sep = "")
  
)

# Unique pairs of samples
env_sample_pairs <- get_unique_sample_pairs(env_dist) %>%
  
  rename(
    
    sample_id_1 = "community_1",
    
    sample_id_2 = "community_2",
    
    sample_id_pairs = "community_pairs"
    
  )

# Convert to a tibble
env_dist_tibble <- as.matrix(env_dist) %>% 
  
  as_tibble(rownames = NA) %>% 
  
  rownames_to_column(var = "sample_id_1") %>% 
  
  pivot_longer(
    
    -sample_id_1, 
    
    names_to = "sample_id_2", 
    
    values_to = "euclidean"
    
  ) %>% 
  
  # Create ID for comparison
  unite(
    
    col = "sample_id_pairs", 
    
    sample_id_1, sample_id_2, 
    
    remove = FALSE, 
    
    sep = "_vs_"
    
  ) %>% 
  
  select(-sample_id_1, -sample_id_2) %>% 
  
  inner_join(env_sample_pairs, ., by = "sample_id_pairs") %>%
  
  separate(
    
    sample_id_1,
    
    into = c("month_1", "year_1"),
    
    sep = " ",
    
    remove = FALSE
    
  ) %>%
  
  separate(
    
    sample_id_2,
    
    into = c("month_2", "year_2"),
    
    sep = " ",
    
    remove = FALSE
    
  ) %>%
  
  separate(
    
    year_1,
    
    into = c("year_1", "rep_number_1"),
    
    sep = "_"
    
  ) %>%
  
  separate(
    
    year_2,
    
    into = c("year_2", "rep_number_2"),
    
    sep = "_"
    
  ) %>%
  
  unite(
    
    col = seasonality_1,
    
    month_1,
    
    year_1,
    
    sep = " ",
    
    remove = FALSE
    
  ) %>%
  
  unite(
    
    col = seasonality_2,
    
    month_2,
    
    year_2,
    
    sep = " ",
    
    remove = FALSE
    
  ) %>%
  
  inner_join(., bc_sample_date1, by = "seasonality_1") %>%
  
  inner_join(., bc_sample_date2, by = "seasonality_2") %>%
  
  # filter(seasonality_1 != seasonality_2) %>%
  
  mutate(
    
    interval = abs(sample_day_1 - sample_day_2)
    
  ) %>%
  
  mutate(
    
    euclidean_norm = euclidean / interval
    
  ) %>%
  
  unite(
    
    col = seasonality_pairs,
    
    seasonality_1,
    
    seasonality_2,
    
    sep = "_vs_",
    
    remove = FALSE
    
  )
```

```{r figS2b_env_pca}
# Figure
figS2b <- ggplot() + 
  
  ## Horizontal and vertical dashed lines
  geom_hline(
    
    yintercept = 0, 
    
    linetype = "dashed", 
    
    size = 0.5, 
    
    colour = "grey30"
    
  ) + 
  
  geom_vline(
    
    xintercept = 0, 
    
    linetype = "dashed", 
    
    size = 0.5, 
    
    colour = "grey30"
    
  ) +
  
  geom_errorbarh(
    
    data = env_pca_points,
    
    aes(
      
      xmin = pc1_mean - pc1_se,
      
      xmax = pc1_mean + pc1_se,
      
      y = pc2_mean,
      
      colour = seasonality
      
    ),
    
    height = 0,
    
    show.legend = FALSE
    
  ) +
  
  geom_errorbar(
    
    data = env_pca_points,
    
    aes(
      
      x = pc1_mean,
      
      ymin = pc2_mean - pc2_se,
      
      ymax = pc2_mean + pc2_se,
      
      colour = seasonality
      
    ),
    
    width = 0,
    
    show.legend = FALSE
    
  ) +
  
  ## Points
  geom_point(
    
    data = env_pca_points, 
    
    aes(x = pc1_mean, y = pc2_mean, fill = seasonality), 
    
    size = 1.5, 
    
    pch = 21
    
  ) + 
  
  scale_fill_viridis(
    
    discrete = TRUE, 
    
    name = NULL,
    
    labels = function(x) seasonality_labels2(x),
    
    option = "plasma"
    
  ) +
  
  scale_colour_viridis(
    
    discrete = TRUE, 
    
    option = "plasma",
    
    guide = "none"
    
  ) +
  
  ## Add vectors
  geom_segment(
    
    data = env_pca_variables, 
    
    aes(x = 0, y = 0, xend = PC1, yend = PC2), 
    
    colour = "grey30", 
    
    arrow = arrow(angle = 30, length = unit(0.03, "npc"), type = "open")
    
  ) + 
  
  ## Labels
  geom_text(
    
    data = env_pca_variables, 
    
    aes(x = PC1.NEW, y = PC2.NEW, label = variable_label), 
    
    colour = "black", 
    
    size = 2.5
    
  ) +
  
  geom_text(
    
    data = env_pma_sig, 
    
    aes(x = -Inf, y = Inf, hjust = -0.25, vjust = 1.5, label = p_label), 
    
    colour = "black",
    
    parse = TRUE,
    
    size = 3
    
  ) +
  
  ## Scale limits
  scale_x_continuous(limits = c(-1.5, 2.5)) + 
  
  scale_y_continuous(limits = c(-2, 2)) + 
  
  ## Add plot and axis titles
  labs(
    
    title = "(b)", 
    
    x = bquote('PC1 (' * .(env_pca_pc1_eig) * '%)'),
    
    y = bquote('PC2 (' * .(env_pca_pc2_eig) * '%)')
    
  ) +
  
  ## Format background and axis lines
  theme(
    
    panel.border = element_rect(fill = NA, colour = "black", linewidth = 0.5), 
    
    panel.background = element_rect(fill = NA),
    
    panel.grid.major = element_blank(),
    
    panel.grid.minor = element_blank()
    
  ) +
  
  ## Format text
  theme(
    
    plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
    
    axis.title.x = element_text(colour = "black", size = 8), 
    
    axis.title.y = element_text(colour = "black", size = 8), 
    
    axis.text.x = element_text(colour = "black", size = 8), 
    
    axis.text.y = element_text(colour = "black", size = 8), 
    
    axis.ticks = element_line(colour = "black")
    
  ) + 
  
  ## Format legends
  theme( 
    
    legend.key = element_blank(), 
    
    legend.title = element_text(colour = "black", size = 10), 
    
    legend.text = element_text(colour = "black", size = 8),
    
    legend.position = "right"
    
  )
```

## Fig. S2  

```{r figS2}
# Figure
figS2 <- plot_grid(
  
  figS2a,
  
  figS2b,
  
  align = 'v',
  
  axis = 'l',
  
  nrow = 2, 
  
  ncol = 1,
  
  rel_heights = c(0.85, 1)
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS2.png", 
  
  plot = figS2, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS2.pdf", 
  
  plot = figS2, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)
```

## Env correlations  

```{r functions_interyear_correlations}
#### Env correlations ####

get_env_corr <- function(x) {
  
  tmp1 <- x %>%
    
    select(`2018`, `2019`)
  
  tmp2 <- corr.test(tmp1)
  
  tmp3 <- tmp2$r[1, 2]
  
  tmp4 <- tmp2$p[1, 2]
  
  tmp5 <- tibble(
    
    r_value = tmp3,
    
    p_value = tmp4
    
  )
  
  return(tmp5)
  
}
```

```{r env_interyear_correlations}
# Get means
env_combined_means <- env_combined %>%
  
  group_by(variable, year, week) %>%
  
  summarise(mean_value = mean(value)) %>%
  
  ungroup() %>%
  
  pivot_wider(
    
    id_cols = c(variable, week),
    
    names_from = "year",
    
    values_from = "mean_value"
    
  ) %>%
  
  get_env_label(.)

# Split
env_means <- env_combined_means %>%
  
  group_by(variable) %>%
  
  group_split(.) %>%
  
  map(., .f = ungroup) %>%
  
  set_names(nm = "dayl", "prcp", "srad", "tmax", "tmin", "vp")
  
# Correlations
env_corr <- map(
  
  env_means,
  
  .f = get_env_corr
  
) %>%
  
  bind_rows(.id = "variable") %>%
  
  get_env_label(.) %>%
  
  mutate(
    
    p_label = map_chr(p_value, .f = get_pval_cutoff),
    
    p_label = ifelse(p_value > 0.05, paste("== ", format(round(p_value, 3), nsmall = 3), sep = ""), p_label)
    
  ) %>%
  
  mutate(
    
    sig_label = paste("atop(r == ", format(round(r_value, 3), nsmall = 3), ",", "italic(P) ", p_label, ")", sep = "")
    
  )
```

### Env. correlations, Fig. S13a  

```{r figS13a_env_corr}
# Figure
figS13a <- ggplot() +
  
  geom_abline(
    
    slope = 1,
    
    intercept = 0,
    
    linewidth = 0.5,
    
    linetype = 2
    
  ) +
  
  geom_point(
    
    data = env_combined_means,
    
    aes(x = `2018`, y = `2019`),
    
    pch = 21,
    
    fill = "#2A406CFF",
    
    alpha = 0.4
    
  ) +
  
  geom_text(
    
    data = env_corr,
    
    aes(x = -Inf, y = Inf, hjust = -0.25, vjust = 1.5, label = sig_label),
    
    parse = TRUE,
    
    stat = "identity",
    
    size = 2.25
    
  ) +
  
  facet_wrap(
    
    ~ variable_label,
    
    scales = "free"
    
  ) +
  
  labs(
    
    title = "(a)",
    
    x = "2018",
    
    y = "2019"
    
  ) +
  
  theme(
    
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
    
    panel.background = element_blank(), 
    
    panel.grid = element_blank(), 
    
    strip.background = element_blank(), 
    
    plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
    
    axis.title = element_text(colour = "black", size = 10), 
    
    axis.ticks = element_line(colour = "black", linewidth = 0.25), 
    
    axis.text.x = element_text(colour = "black", size = 8), 
    
    axis.text.y = element_text(colour = "black", size = 8), 
    
    strip.text = element_text(colour = "black", size = 10), 
    
    legend.key = element_blank(),

    legend.title = element_text(colour = "black", size = 10),

    legend.text = element_text(colour = "black", size = 8),

    legend.position = "bottom"
    
  )
```

### Fig. S13b  

```{r figS13b_env_rate_change}
# Get rate change
env_rate_change_means <- env_dist_tibble %>%
  
  group_by(interval) %>%
  
  summarise(mean_euclidean = mean(euclidean)) %>%
  
  ungroup()


# Env gam
env_rate_change_gam <- summary(gam(mean_euclidean ~ s(interval), data = env_rate_change_means))

# P
env_rate_change_lm_P <- tibble(
  
  P = env_rate_change_gam$s.table[4],
  
  R2 = env_rate_change_gam$r.sq
  
) %>%
  
  mutate(
    
    P = ifelse(P >= 0.001, format(round(P, digits = 3), nsmall = 3, scientific = FALSE), "< 0.001"),
    
    R2 = format(round(R2, digits = 3), nsmall = 3),
    
    p_label = ifelse(
      
      P == "< 0.001",
      
      paste("R^2 == ", R2, "*','~", "italic(P) ", P, sep = ""),
      
      paste("R^2 == ", R2, "*','~", "italic(P) == ", P, sep = "")
      
    )
    
  )

# Plot env rate change
figS13b <- ggplot() + 
    
    scale_y_continuous(
      
      limits = c(0, NA)
      
    ) + 
    
    scale_x_continuous(
      
      limits = c(0, NA)
      
    ) + 
    
    geom_point(
      
      data = env_dist_tibble, 
      
      aes(
        
        x = interval, 
        
        y = euclidean, 
        
      ), 
      
      fill = "#2A406CFF",
      
      pch = 21, 
      
      alpha = 0.1, 
      
      size = 1
      
    ) +
    
    stat_summary(
      
      data = env_dist_tibble, 
      
      aes(
        
        x = interval, 
        
        y = euclidean
        
      ), 
      
      fun = mean,
      
      geom = "point",
      
      fill = "#2A406CFF",
      
      colour = "black",
      
      pch = 21, 
      
      size = 2.5
      
    ) +
    
    geom_smooth(
      
      data = env_rate_change_means, 
      
      aes(
        
        x = interval, 
        
        y = mean_euclidean, 
        
      ), 
      
      method = "gam", 
      
      formula = y ~ s(x), 
      
      se = TRUE, 
      
      alpha = 0.3, 
      
      fill = "#2A406CFF",
      
      colour = "black",
      
      show.legend = FALSE,
      
      linewidth = 0.5
      
    ) +
    
    geom_text(
      
      data = env_rate_change_lm_P,
      
      aes(x = -Inf, y = -Inf, vjust = -1.5, hjust = -0.5, label = p_label),
      
      parse = TRUE,
      
      size = 3
      
    ) +
    
    labs(
      
      title = "(b)", 
      
      x = "Days", 
      
      y = "Euclidean distance"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 9), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8), 
      
      axis.text.y = element_text(colour = "black", size = 8)
      
    )
```

```{r figS13}
# Figure
figS13 <- plot_grid(
  
  figS13a,
  
  figS13b,
  
  align = 'v',
  
  axis = 'l',
  
  nrow = 2, 
  
  ncol = 1,
  
  rel_heights = c(1, 1)
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS13.png", 
  
  plot = figS13, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS13.pdf", 
  
  plot = figS13, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 7.5, 
  
  units = "in"
  
)
```

## Composition - environment, Fig. 2a  

```{r fig2a_functions}
#### Community environment LM ####

get_community_env_change_lm <- function(x) {
  
  tmp1 <- lm(mean_bray ~ mean_euclidean, data = x)
  
  tmp2 <- summary(tmp1)
  
  return(tmp2)
  
}
```

## Fig. 2  

```{r fig2}
# Bray curtis
community_change <- bc_community_rate_change_bray_combined %>%
  
  unite(
    
    col = seasonality_pairs,
    
    seasonality_1,
    
    seasonality_2,
    
    sep = "_vs_",
    
    remove = FALSE
    
  )

# Env table
env_change <- env_dist_tibble %>%
  
  select(sample_id_1, sample_id_2, sample_id_pairs, seasonality_pairs, euclidean)

# Combine
community_env_change <- community_change %>%
  
  inner_join(., env_change, by = "seasonality_pairs")

# Means
community_env_change_means <- community_env_change %>%
  
  group_by(community, plant_habitat, seasonality_pairs) %>%
  
  summarise(
    
    mean_bray = mean(bray),
    
    se_bray = se(bray),
    
    mean_euclidean = mean(euclidean),
    
    se_euclidean = se(euclidean)
    
  ) %>%
  
  ungroup()

# LM input
community_env_change_lm_input <- community_env_change_means %>%
  
  group_by(community, plant_habitat) %>%
  
  group_split(.) %>%
  
  map(., .f = ungroup)

# LM
community_env_change_lm <- map(
  
  community_env_change_lm_input,
  
  .f = get_community_env_change_lm
  
)

# Community env change lm P values
bc_community_env_change_lm_P <- map(
  
  community_env_change_lm, 
  
  .f = get_community_rate_change_lm_P
  
) %>%
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh")) %>%
  
  bind_rows(.id = "ID") %>% 
  
  mutate(
    
    ID = str_remove(ID, "bc_")
    
  ) %>% 
  
  separate(
    
    col = ID, 
    
    into = c("community", "sample_type")
    
  ) %>% 
  
  mutate(
    
    community = ifelse(community == "16s", "Bacteria and Archaea", "Fungi"), 
    
    sample_type = toupper(sample_type), 
    
    plant_habitat = map_chr(sample_type, plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  ) %>%
  
  mutate(
    
    P = ifelse(P >= 0.001, format(round(P, digits = 3), nsmall = 3, scientific = FALSE), "< 0.001"),
    
    R2 = format(round(R2, digits = 3), nsmall = 3),
    
    p_label = ifelse(
      
      P == "< 0.001",
      
      paste("R[adj.]^2 == ", R2, "*','~", "italic(P) ", P, sep = ""),
      
      paste("R[adj.]^2 == ", R2, "*','~", "italic(P) == ", P, sep = "")
      
    )
    
  )

# Figure
fig2a <- ggplot() +
  
  scale_y_continuous(
    
    limits = c(0, 1)
    
  ) +
  
  scale_x_continuous(
    
    limits = c(0, NA)
    
  ) +
  
  geom_point(
    
    data = community_env_change_means,
    
    aes(x = mean_euclidean, y = mean_bray, fill = plant_habitat),
    
    pch = 21,
    
    alpha = 0.6,
    
    size = 2
    
  ) +
  
  geom_smooth(
    
    data = community_env_change_means,
    
    aes(
      
      x = mean_euclidean,
      
      y = mean_bray,
      
      fill = plant_habitat,
      
      group = plant_habitat
      
    ), 
    
    method = "lm", 
    
    formula = y ~ x, 
    
    se = TRUE, 
    
    alpha = 0.3, 
    
    colour = "black",
    
    show.legend = FALSE,
    
    linewidth = 0.5
    
  ) + 
  
  scale_colour_viridis(
    
    discrete = TRUE, 
    
    begin = 0.8, 
    
    end = 0, 
    
    guide = "none"
    
  ) + 
  
  scale_fill_viridis(
    
    name = "Plant habitat", 
    
    discrete = TRUE, 
    
    begin = 0.8, 
    
    end = 0,
    
    guide = "none"
    
  ) +
  
  geom_text(
    
    data = bc_community_env_change_lm_P,
    
    aes(x = -Inf, y = -Inf, vjust = -0.25, hjust = -0.3, label = p_label),
    
    parse = TRUE,
    
    size = 3
    
  ) +
  
  facet_grid(
    
    rows = vars(community),
    
    cols = vars(plant_habitat)
    
  ) + 
  
  labs(
    
    title = "(a)", 
    
    x = "Environment (Euclidean distance)", 
    
    y = "Community composition (Bray-Curtis dissimilarity)"
    
  ) + 
  
  theme(
    
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
    
    panel.background = element_blank(), 
    
    panel.grid = element_blank(), 
    
    strip.background = element_blank(), 
    
    plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
    
    axis.title = element_text(colour = "black", size = 9), 
    
    axis.ticks = element_line(colour = "black", linewidth = 0.25), 
    
    axis.text.x = element_text(colour = "black", size = 8), 
    
    axis.text.y = element_text(colour = "black", size = 8), 
    
    strip.text = element_text(colour = "black", size = 8)#, 
    
    # legend.key = element_blank(), 
    # 
    # legend.title = element_text(colour = "black", size = 9), 
    # 
    # legend.text = element_text(colour = "black", size = 8), 
    # 
    # legend.position = "bottom"
    
  ) #+ 
  
  # guides(
  #   
  #   fill = guide_legend(override.aes = list(
  #     
  #     alpha = 1, 
  #     
  #     pch = 21, 
  #     
  #     size = 2.5, 
  #     
  #     colour = "black"
  #     
  #   ))
  #   
  # )

# Grid 2
fig2 <- plot_grid(
  
  fig2a,
  
  fig2b + theme(legend.position = "none"),
  
  align = "v",
  
  axis = "l",
  
  nrow = 2,
  
  ncol = 1
  
)

# Save figure
ggsave2(
  
  filename = "data/results/fig2.png", 
  
  plot = fig2, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/fig2.pdf", 
  
  plot = fig2, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)
```

```{r tableS8}
# Get input for Table S8
tableS8_input <- community_env_change_lm %>% 
  
  map(., .f = get_rate_change_lm_table) %>% 
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh")) %>%
  
  bind_rows(.id = "id") %>% 
  
  mutate(id = str_remove(id, "bc_")) %>% 
  
  separate(
    
    id, 
    
    into = c("Community", "Plant habitat"), 
    
    sep = "_"
    
  ) %>% 
  
  rename(
    
    Variable = "variable", 
    
    P = "Pr(>|t|)"
    
  ) %>% 
  
  filter(Variable == "mean_euclidean") %>% 
  
  mutate(
    
    Community = ifelse(Community == "16s", "Bacteria/Archaea", Community), 
    
    Community = ifelse(Community == "its", "Fungi", Community), 
    
    `Plant habitat` = toupper(`Plant habitat`), 
    
    `Plant habitat` = map(`Plant habitat`, .f = plant_hab), 
    
    `Plant habitat` = factor(
      
      `Plant habitat`, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    ),
    
    lci = Estimate - (qnorm(0.975) * `Std. Error`),
    
    uci = Estimate + (qnorm(0.975) * `Std. Error`),
    
    lci = format(round(365 * lci, digits = 3), nsmall = 3, scientific = FALSE), 
    
    uci = format(round(365 * uci, digits = 3), nsmall = 3, scientific = FALSE), 
    
    Estimate = format(round(365 * Estimate, digits = 3), nsmall = 3, scientific = FALSE), 
    
    P = ifelse(P >= 0.001, format(round(P, digits = 3), nsmall = 3, scientific = FALSE), "< 0.001"), 
    
    `t value` = format(round(`t value`, digits = 2), nsmall = 2)
    
  ) %>% 
  
  mutate(Slope = paste(Estimate, " [", lci, ", ", uci, "]", sep = "")) %>%
  
  select(Community, `Plant habitat`, Slope, `t value`, P)

# Save
write_tsv(
  
  tableS8_input,
  
  file = "data/results/tableS8.txt"
  
)
```

# Community dispersion  

## Calculate dispersion  

```{r betadisper_functions}
#### Beta-disper function ####

get_betadisper <- function(dist.mat, env.table) {
  
  # Create filter
  tmp1 <- rownames(as.matrix(dist.mat))
  
  # Order env data
  tmp2 <- env.table %>% 
    
    droplevels() %>% 
    
    filter(sample_unique_id %in% tmp1) %>% 
    
    arrange(match(sample_unique_id, tmp1))
  
  tmp3 <- betadisper(dist.mat, tmp2$seasonality, type = "median")
  
  # Return
  return(tmp3)
  
}

#### Betadisper distances ####

get_betadisper_distance <- function(betadisper.model, metadata.tab) {
  
  tmp1 <- betadisper.model$distances %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rename(dist_to_centroid = "value") %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    inner_join(metadata.tab, ., by = "sample_unique_id")
  
  return(tmp1)
  
}

#### Run betadisper ANOVA ####

betadisper_aov <- function(x) {
  
  tmp1 <- aov(dist_to_centroid ~ year * month * host_species, data = x)
  
  return(tmp1)
  
}

#### Get betadisper seasonality P values ####

get_betadisper_seasonality_aov_Pval <- function(x) {
  
  tmp1 <- get_seasonality_aov_Pval(x)
  
  tmp2 <- tmp1 %>%

    mutate(pval_label = NA) %>%

    mutate(pval_label = ifelse(`Pr(>F)` < 0.001, "season%*%year~italic(P) < 0.001", pval_label)) %>%

    mutate(pval_label = ifelse(`Pr(>F)` <= 0.05 & `Pr(>F)` >= 0.001, paste("season%*%year~italic(P) == ", round(`Pr(>F)`, 3), sep = ""), pval_label)) %>%

    mutate(pval_label = ifelse(`Pr(>F)` > 0.05, "season%*%year~italic(n.s.)", pval_label))
  
  return(tmp2)
  
}

#### Get CLD for dispersion ####

get_betadisper_cld2 <- function(aov.model, tukey.model) {
  
  tmp1 <- multcompView::multcompLetters4(
    
    aov.model, 
    
    tukey.model
    
  )
  
  tmp2 <- tmp1$`year:month`
  
  tmp3 <- tibble(
    
    timepoint = names(tmp2$Letters), 
    
    cld = tmp2$Letters
    
  ) %>%
    
    separate(timepoint, into = c("year", "month"), sep = ":") %>%
    
    unite(seasonality, month, year, sep = " ") %>%
    
    mutate(seasonality = factor(seasonality, levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018",
                                                        "Feb 2019", "June 2019", "Aug 2019", "Oct 2019")))
  
  return(tmp3)
  
}
```

```{r betadisper}
# Run betadisper
bc_betadisper <- map2(bc_dist, bc_metadata_habsplit_filtered, .f = get_betadisper)

# Get betadisper distances
bc_betadisper_distance <- map2(
  
  bc_betadisper, bc_metadata_habsplit_filtered, 
  
  .f = get_betadisper_distance
  
)

# Run ANOVA on betadispersion models
bc_betadisper_aov <- map(bc_betadisper_distance,.f = betadisper_aov)

# Get ANVOA summaries
bc_betadisper_aov_summary <- map(bc_betadisper_aov, .f = summary.aov)

# Get seasonality P values
bc_betadisper_seasonality_Pval <- map(bc_betadisper_aov_summary, .f = get_betadisper_seasonality_aov_Pval)

# Get Tukey's HSD
bc_betadisper_tukeyHSD <- map(bc_betadisper_aov, .f = TukeyHSD)

# Get betadisper CLD for plots
bc_betadisper_seasonality_cld <- map2(
  
  bc_betadisper_aov, 
  
  bc_betadisper_tukeyHSD, 
  
  .f = get_betadisper_cld2
  
)
```

### Table S6  

```{r tableS6_function}
#### Get betadisper anova table ####

get_betadisper_aov_table <- function(x) {
  
  tmp1 <- x[[1]] %>% 
    
    as.data.frame(.) %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "Variable") %>% 
    
    rename(
      
      P = "Pr(>F)", 
      
      `Sum of sqs.` = "Sum Sq"
      
    ) %>% 
    
    mutate(
      
      `Sum of sqs.` = format(round(`Sum of sqs.`, 1), nsmall = 1), 
      
      `F value` = format(round(`F value`, 2), nsmall = 2), 
      
      P = ifelse(P >= 0.001, format(round(P, 3), nsmall = 3, scientific = FALSE), P), 
      
      P = ifelse(as.numeric(P) < 0.001, "< 0.001", P)
      
    ) %>% 
    
    mutate(across(.cols = everything(), .fns = str_trim)) %>% 
    
    mutate(
      
      `F value` = ifelse(`F value` == "NA", "", `F value`), 
      
      `F value` = ifelse(is.na(`F value`), "", `F value`), 
      
      P = ifelse(P == "NA", "", P), 
      
      P = ifelse(is.na(P), "", P)
      
    ) %>% 
    
    mutate(
      
      Variable = str_to_title(Variable), 
      
      Variable = str_replace_all(Variable, "\\:", " x "),
      
      Variable = str_replace_all(Variable, "\\_", " ")
      
    ) %>% 
    
    select(-`Mean Sq`)
  
  return(tmp1)
  
}
```

```{r tableS6_input}
# Get table S6 input
tableS6_input <- map(
  
  bc_betadisper_aov_summary, 
  
  .f = get_betadisper_aov_table
  
) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(id = str_remove(id, "bc_")) %>% 
  
  separate(
    
    col = id, 
    
    into = c("Community", "Plant habitat")
    
  ) %>% 
  
  mutate(
    
    Community = ifelse(Community == "16s", "Bacteria/Archaea", Community), 
    
    Community = ifelse(Community == "its", "Fungi", Community), 
    
    `Plant habitat` = toupper(`Plant habitat`), 
    
    `Plant habitat` = map_chr(`Plant habitat`, .f = plant_hab), 
    
    `Plant habitat` = factor(
      
      `Plant habitat`, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Save
write_tsv(
  
  tableS6_input,
  
  file = "data/results/tableS6.txt"
  
)
```

## Plot dispersion  

### Fig. S8  

```{r figS8_function}
#### Prep betadisper fig data ####

get_betadisper_fig_data <- function(x) {
  
  tmp1 <- bind_rows(x, .id = "id") %>% 
    
    mutate(id = str_remove(id, "bc_")) %>% 
    
    separate(
      
      col = id, 
      
      into = c("community", "plant_habitat")
      
    ) %>% 
    
    mutate(
      
      plant_habitat = toupper(plant_habitat), 
      
      plant_habitat = map_chr(plant_habitat, .f = plant_hab), 
      
      plant_habitat = factor(
        
        plant_habitat, 
        
        levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
        
      )
      
    )
  
  return(tmp1)
  
}

#### Plot community dispersion 2 ####

plot_betadisper2 <- function(beta.disper, cdl.stats, stat.labels.P, plot.title, community) {
  
  tmp0 <- beta.disper %>% 
    
    group_by(plant_habitat, year, seasonality) %>% 
    
    summarise(
      
      betadisper_mean = mean(dist_to_centroid), 
      
      betadisper_se = se(dist_to_centroid)
      
    ) %>% 
    
    ungroup()
  
  tmp1 <- ggplot() + 
    
    scale_y_continuous(expand = expansion(mult = c(0.15, 0.15))) + 
    
    geom_point(
      
      data = beta.disper, 
      
      aes(x = seasonality, y = dist_to_centroid, colour = plant_habitat, fill = plant_habitat), 
      
      position = position_jitter(width = 0.2, seed = 12345), 
      
      pch = 21, 
      
      alpha = 0.5, 
      
      size = 1
      
    ) + 
    
    geom_boxplot(
      
      data = beta.disper, 
      
      aes(x = seasonality, y = dist_to_centroid, colour = plant_habitat), 
      
      outlier.shape = NA,
      
      fill = "white", 
      
      alpha = 0.4, 
      
      show.legend = FALSE
      
    ) + 
    
    geom_line(
      
      data = tmp0, 
      
      aes(x = seasonality, y = betadisper_mean, group = year, colour = plant_habitat), 
      
      show.legend = FALSE
      
    ) + 
    
    # stat_summary(
    #   
    #   data = beta.disper, 
    #   
    #   aes(x = seasonality, y = dist_to_centroid, colour = plant_habitat), 
    #   
    #   fun.data = mean_cl_normal,
    #   
    #   geom = "errorbar", 
    #   
    #   linewidth = 0.25, 
    #   
    #   width = 0
    #   
    # ) +
    
    stat_summary(
      
      data = beta.disper, 
      
      aes(x = seasonality, y = dist_to_centroid, fill = plant_habitat), 
      
      fun = mean, 
      
      geom = "point",
      
      colour = "black",
      
      pch = 21,
      
      size = 2
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    geom_text(
      
      data = cdl.stats,
      
      aes(x = seasonality, y = Inf, vjust = 1.55, label = cld),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    geom_text(
      
      data = stat.labels.P,
      
      aes(x = 3, y = -Inf, vjust = -0.25, label = pval_label),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    facet_wrap(
      
      ~ plant_habitat, 
      
      scales = "free_y", 
      
      nrow = 3, 
      
      ncol = 1
      
    ) + 
    
    labs(
      
      title = plot.title, 
      
      x = NULL, 
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )
  
  if(community == "16s") {
    
    tmp2 <- tmp1 + 
      
      labs(y = "Bacterial/archaeal community dispersion\n(distance to centroid)")
    
  } else if(community == "its") {
    
    tmp2 <- tmp1 + 
      
      labs(y = "Fungal community dispersion\n(distance to centroid)")
    
  } else {
    
    tmp2 <- tmp1
    
  }
  
  return(tmp2)
  
}
```

```{r generate_betadisper_plots_figS8}
# Fig. S8a data (bacteria/archaea)
bc_betadisper_data_figS8a <- bc_betadisper_distance[1:3] %>% 
  
  bind_rows()

# Fig. S8b data (fungi)
bc_betadisper_data_figS8b <- bc_betadisper_distance[4:6] %>% 
  
  bind_rows()

# Fig. S8a CLD (bacteria/archaea)
bc_betadisper_cld_figS8a <- get_betadisper_fig_data(bc_betadisper_seasonality_cld[1:3])

# Fig. S8b CLD (fungi)
bc_betadisper_cld_figS8b <- get_betadisper_fig_data(bc_betadisper_seasonality_cld[4:6])

# Fig. S8a P (bacteria/archaea)
bc_betadisper_pval_figS8a <- get_betadisper_fig_data(bc_betadisper_seasonality_Pval[1:3])

# Fig. S8b P (fungi)
bc_betadisper_pval_figS8b <- get_betadisper_fig_data(bc_betadisper_seasonality_Pval[4:6])

# Fig S8a
figS8a <- plot_betadisper2(
  
  bc_betadisper_data_figS8a, 
  
  bc_betadisper_cld_figS8a, 
  
  bc_betadisper_pval_figS8a, 
  
  "(a)",  
  
  "16s"
  
)

# Fig S8b
figS8b <- plot_betadisper2(
  
  bc_betadisper_data_figS8b, 
  
  bc_betadisper_cld_figS8b, 
  
  bc_betadisper_pval_figS8b, 
  
  "(b)",  
  
  "16s"
  
)

# Compile figure
figS8 <- plot_grid(
  
  figS8a, 
  
  figS8b, 
  
  align = 'vh', 
  
  nrow = 1, 
  
  ncol = 2
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS8.png", 
  
  plot = figS8, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS8.pdf", 
  
  plot = figS8, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)
```

# Core communities  

## Abundance-occupancy  

```{r abund_occupancy_functions}
#### Abundance-occupancy ####

get_abund_occupancy <- function(asv.tab, metadata.tab, cutoff) {
  
  tmp1 <- asv.tab %>%
    
    as_tibble(rownames = NA) %>%
    
    rownames_to_column(var = "sample_unique_id") %>%
    
    pivot_longer(
      
      -sample_unique_id,
      
      names_to = "asv_id",
      
      values_to = "n_seqs"
      
    ) %>%
    
    inner_join(metadata.tab, ., by = "sample_unique_id")
  
  tmp2 <- tmp1 %>% 
    
    mutate(asv_presence = ifelse(n_seqs > 0, 1, 0)) %>% 
    
    group_by(year, month) %>% 
    
    mutate(sample_n = n_distinct(sample_unique_id)) %>% 
    
    ungroup() %>% 
    
    group_by(asv_id, year, month) %>% 
    
    mutate(local_presence = sum(asv_presence)) %>% 
    
    ungroup() %>% 
    
    mutate(asv_local_occupancy = local_presence / sample_n) %>% 
    
    group_by(asv_id) %>% 
    
    mutate(max_local_occupancy = max(asv_local_occupancy))  %>% 
    
    ungroup() %>% 
    
    mutate(global_sample_n = n_distinct(sample_unique_id)) %>% 
    
    group_by(asv_id) %>% 
    
    mutate(global_presence = sum(asv_presence)) %>% 
    
    ungroup() %>% 
    
    mutate(asv_global_occupancy = global_presence / global_sample_n) %>% 
    
    # Calculate relabund
    group_by(sample_unique_id) %>% 
    
    mutate(sample_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    mutate(relabund = n_seqs / sample_n_seqs) %>% 
    
    group_by(asv_id) %>% 
    
    mutate(mean_relabund = mean(relabund)) %>% 
    
    ungroup() %>% 
    
    mutate(log_relabund = log10(mean_relabund)) %>% 
    
    select(asv_id, sample_type, plant_habitat, max_local_occupancy, asv_global_occupancy, mean_relabund, log_relabund) %>% 
    
    distinct() %>% 
    
    mutate(
      
      core_id = ifelse(max_local_occupancy >= cutoff, "Core", "Not core"), 
      
      core_id = factor(core_id, levels = c("Core", "Not core"))
      
    )
  
  return(tmp2)
  
}
```

```{r abund_occupancy}
# Determine abundance-occupancy for ASVs and set core cutoff to 70% of a time point
bc_abund_occupancy <- pmap(
  
  list(bc_asv_habsplit_df_SRS_output, bc_metadata_habsplit_filtered, 0.7), 
  
  .f = get_abund_occupancy
  
) %>% 
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh")) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_le"), 
    
    id = str_remove(id, "_re"), 
    
    id = str_remove(id, "_rh"), 
    
    id = ifelse(id == "16s", "Bacteria and Archaea", id), 
    
    id = ifelse(id == "its", "Fungi", id), 
    
    id = factor(id, levels = c("Bacteria and Archaea", "Fungi"))
    
  ) %>% 
  
  rename(Community = "id")
```

### Abund-occupancy, Fig. S15  

```{r figS15_functions}
#### Get abundance-occupancy plots 2 ####

get_abund_occupancy_fig2 <- function(abund.occ) {
  
  tmp0 <- bc_abund_occupancy %>% 
    
    filter(core_id == "Not core")
  
  tmp1 <- bc_abund_occupancy %>% 
    
    filter(core_id == "Core")
  
  tmp2 <- ggplot() + 
    
    geom_point(
      
      data = tmp0, 
      
      aes(x = log_relabund, y = asv_global_occupancy), 
      
      colour = "grey", 
      
      fill = "grey", 
      
      alpha = 0.5, 
      
      pch = 21, 
      
      size = 2
      
    ) + 
    
    geom_point(
      
      data = tmp1, 
      
      aes(x = log_relabund, y = asv_global_occupancy, fill = plant_habitat, colour = plant_habitat), 
      
      alpha = 0.5, 
      
      pch = 21, 
      
      size = 2
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      name = "Plant habitat", 
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0
      
    ) + 
    
    facet_grid(
      
      cols = vars(plant_habitat), 
      
      rows = vars(Community), 
      
      scales = "free_x"
      
    ) + 
    
    labs(
      
      title = NULL, 
      
      y = "Occupancy", 
      
      x = bquote(italic(log)*'(ASV relative abundance)')
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )
  
  return(tmp2)
  
}
```

```{r figS15}
# Get abundance-occupancy plots
figS15 <- get_abund_occupancy_fig2(bc_abund_occupancy)

# Save figure
ggsave2(
  
  filename = "data/results/figS15.png", 
  
  plot = figS15, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS15.pdf", 
  
  plot = figS15, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)
```

## Contribution to beta-diversity  

```{r core_contribution_functions}
#### Core asv filter ####

get_core_asv_filter <- function(x) {
  
  tmp1 <- x %>% 
    
    filter(core_id == "Core") %>% 
    
    select(asv_id) %>% 
    
    distinct() %>% 
    
    pull(asv_id)
  
  return(tmp1)
  
}

#### Get core ASV table for distmat ####

get_core_asv_df <- function(asv.tab, core.filter) {
  
  tmp1 <- asv.tab %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>%
    
    filter(asv_id %in% core.filter) %>% 
    
    drop_0seq_samples(.) %>% 

    asv_long_to_wide_df(.)
  
  return(tmp1)
  
}

#### Get core sample totals ####

get_core_sample_n_seqs <- function(x) {
  
  tmp1 <- x %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    select(sample_unique_id, asv_id, n_seqs) %>% 
    
    group_by(sample_unique_id) %>% 
    
    summarise(sample_n_seqs = sum(n_seqs)) %>% 
    
    ungroup()
  
  return(tmp1)
  
}

#### Get scaling factor for core bray dist ####

get_core_bray_scaling <- function(core.sample.seqs, unique.pairs, full.sample.seqs) {
  
  tmp1 <- core.sample.seqs %>% 
    
    mutate(sample_id2 = sample_unique_id) %>% 
    
    expand(sample_unique_id, sample_id2, .name_repair = "unique") %>% 
    
    unite(col = community_pairs, sample_unique_id, sample_id2, sep = "_vs_", remove = FALSE) %>% 
    
    inner_join(., core.sample.seqs, by = "sample_unique_id") %>% 
    
    rename(
      
      sample_id1 = "sample_unique_id", 
      
      sample_unique_id = "sample_id2", 
      
      sample_n_seqs_1 = "sample_n_seqs"
      
    ) %>% 
    
    inner_join(., core.sample.seqs, by = "sample_unique_id") %>% 
    
    inner_join(unique.pairs, ., by = "community_pairs") %>% 
    
    rename(sample_n_seqs_2 = "sample_n_seqs") %>% 
    
    select(community_pairs, community_1, community_2, sample_n_seqs_1, sample_n_seqs_2) %>% 
    
    pivot_longer(
      
      cols = c(sample_n_seqs_1, sample_n_seqs_2), 
      
      names_to = "sample_unique_id", 
      
      values_to = "sample_n_seqs"
      
    ) %>% 
    
    group_by(community_pairs) %>% 
    
    summarise(sample_n_seqs = sum(sample_n_seqs)) %>% 
    
    ungroup() %>% 
    
    # Calculate factor for correcting core bray curtis distances
    mutate(scaling_factor = sample_n_seqs / (2 * full.sample.seqs))
  
  return(tmp1)
  
}

#### Get bray dist tibble ####

get_bray_dist_tibble <- function(dist.mat, unique.pairs) {
  
  tmp1 <- as.matrix(dist.mat) %>% 
  
  as_tibble(rownames = NA) %>% 
  
  rownames_to_column(var = "community_1") %>% 
  
  pivot_longer(
    
    -community_1, 
    
    names_to = "community_2", 
    
    values_to = "bray"
    
  ) %>% 
  
  # Create ID for comparison
  unite(
    
    col = "community_pairs", 
    
    community_1, community_2, 
    
    remove = FALSE, 
    
    sep = "_vs_"
    
  ) %>% 
  
  select(-community_1, -community_2) %>% 
  
  inner_join(unique.pairs, ., by = "community_pairs") %>% 
  
  separate(
    
    col = community_1, 
    
    into = c("genotype_1", "cutting_location_1", "replicate_1", "month_1", "year_1", "sample_type_1", "host_species_1"), 
    
    sep = "_", 
    
    remove = FALSE
    
  ) %>% 
  
  separate(
    
    col = community_2, 
    
    into = c("genotype_2", "cutting_location_2", "replicate_2", "month_2", "year_2", "sample_type_2", "host_species_2"), 
    
    sep = "_", 
    
    remove = FALSE
    
  )
  
  return(tmp1)
  
}

#### Calculate core contributions ####

get_core_contribution <- function(core.dist, full.dist, unique.pairs, scaling.factors) {
  
  tmp1 <- get_bray_dist_tibble(core.dist, unique.pairs) %>% 
    
    rename(bray_core = "bray")
  
  tmp2 <- get_bray_dist_tibble(full.dist, unique.pairs) %>% 
    
    rename(bray_full = "bray")
  
  tmp3 <- tmp2 %>% 
    
    select(community_pairs, bray_full) %>% 
    
    inner_join(tmp1, ., by = "community_pairs") %>% 
    
    inner_join(., scaling.factors, by = "community_pairs") %>% 
    
    # Scale core
    mutate(bray_core_scaled = bray_core * scaling_factor) %>% 
    
    # Calculate contribution
    mutate(core_contribution = bray_core_scaled / bray_full) %>% 
    
    # Add seasonality
    mutate(
      
      seasonality_1 = paste(month_1, year_1, sep = " "), 
      
      seasonality_2 = paste(month_2, year_2, sep = " ")
      
    )
  
  return(tmp3)
  
}

#### Get overall and seasonal contributions ####

combine_core_contribution_data <- function(core.contributions, list.elements, community) {
  
  # tmp1 <- map_dfr(bc_core_contribution[list.elements], bind_rows) %>%
  # 
  #   mutate(betadiv = "Overall")
  
  tmp2 <- map_dfr(bc_core_contribution[list.elements], bind_rows) %>% 
    
    filter(seasonality_1 == seasonality_2) %>% 
    
    mutate(betadiv = "Within dates")
  
  tmp3 <- map_dfr(bc_core_contribution[list.elements], bind_rows) %>% 
    
    filter(seasonality_1 != seasonality_2) %>% 
    
    mutate(betadiv = "Between dates")
  
  # tmp4 <- bind_rows(tmp1, tmp2, tmp3) %>%
  tmp4 <- bind_rows(tmp2, tmp3) %>%
    
    mutate(
      
      plant_habitat = sapply(sample_type_1, plant_hab, simplify = TRUE), 
      
      plant_habitat = factor(
        
        plant_habitat, 
        
        levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
        
      ), 
      
      # betadiv = factor(betadiv, levels = c("Overall", "Within dates", "Between dates")),
      betadiv = factor(betadiv, levels = c("Within dates", "Between dates")),
      
      community = community
      
    )
  
  return(tmp4)
  
}

#### Get core temporal contribution, mean and CI for manuscript ####

get_core_contribution_mean_CI <- function(x) {
  
  tmp1 <- x %>% 
    
    filter(seasonality_1 != seasonality_2) %>% 
    
    mutate(core_contribution = 100 * core_contribution) %>% 
    
    summarise(
      
      mean_core_contribution = mean(core_contribution), 
      
      lci_core_contribution = se(core_contribution), 
      
      uci_core_contribution = se(core_contribution)
      
    )
  
  return(tmp1)
  
}

#### Core contribution ANOVA ####

get_core_contribution_aov <- function(x) {
  
  tmp1 <- aov(core_contribution ~ betadiv * plant_habitat, data = x)
  
  return(tmp1)
  
}

#### Get type by habitat P values ####

get_core_contribution_aov_Pval <- function(x) {
  
  tmp1 <- x[[1]]$`Pr(>F)`[3]
  
  tmp2 <- tibble(
    
    `Pr(>F)` = tmp1
    
  ) %>%
    
    mutate(pval_label = NA) %>%
    
    mutate(pval_label = ifelse(`Pr(>F)` < 0.001, "type%*%habitat~italic(P) < 0.001", pval_label)) %>%
    
    mutate(pval_label = ifelse(`Pr(>F)` <= 0.05 & `Pr(>F)` >= 0.001, paste("type%*%habitat~italic(P) == ", round(`Pr(>F)`, 3), sep = ""), pval_label)) %>%
    
    mutate(pval_label = ifelse(`Pr(>F)` > 0.05, "type%*%habitat~italic(n.s.)", pval_label))
  
  return(tmp2)
  
}

#### Get core contribution cld 2 ####

get_core_contribution_cld2 <- function(anova.model, tukey.model) {
  
  tmp1 <- multcompView::multcompLetters4(
    
    bc_core_contribution_aov$`Bacteria and Archaea`, 
    
    bc_core_contribution_tukeyHSD$`Bacteria and Archaea`
    
  )
  
  tmp2 <- tmp1$`betadiv:plant_habitat`$Letters
  
  tmp3 <- tibble(
    
    Variable = names(tmp2), 
    
    cld = str_remove_all(tmp2, " ")
    
  ) %>% 
    
    separate(
      
      col = Variable, 
      
      into = c("type", "plant_habitat"), 
      
      sep = ":"
      
    ) %>% 
    
    mutate(
      
      type = factor(type, levels = c("Within dates", "Between dates")), 
      
      plant_habitat = factor(
        
        plant_habitat, 
        
        levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
        
      )
      
    ) %>% 
    
    rename(cld_label = "cld")
  
  return(tmp3)
  
}
```

```{r core_contribution}
# Get core ASV filter
bc_core_asv_filter <- bc_abund_occupancy %>% 
  
  group_by(Community, plant_habitat) %>% 
  
  group_split(.) %>% 
  
  map(., .f = ungroup) %>% 
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh")) %>% 
  
  map(., .f = get_core_asv_filter)

# Get core asv table for distance matrix calculation
bc_asv_core_df <- map2(bc_asv_habsplit_df_SRS_output, bc_core_asv_filter, .f = get_core_asv_df)

# Core asv distmat
bc_core_dist <- map(bc_asv_core_df, .f = get_bray_dist)

# Get core sample seq totals
bc_core_sample_nseqs <- map(bc_asv_core_df, .f = get_core_sample_n_seqs) %>% 
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh"))

# Get factors for scaling core bray curtis dissimilarities
bc_core_scale <- pmap(
  
  list(bc_core_sample_nseqs, bc_sample_pairs, bc_sample_seq_min), 
  
  .f = get_core_bray_scaling
  
)

# Get core contributions
bc_core_contribution <- pmap(
  
  list(bc_core_dist, bc_dist, bc_sample_pairs, bc_core_scale), 
  
  .f = get_core_contribution
  
)


# Core contribution mean for manuscript
bc_core_contribution_mean_CI <- map(
  
  bc_core_contribution, 
  
  .f = get_core_contribution_mean_CI
  
) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_asv_df")
    
  ) %>% 
  
  separate(
    
    col = id, 
    
    into = c("Community", "plant_habitat"), 
    
    sep = "_"
    
  )  %>% 
    
    mutate(
      
      plant_habitat = toupper(plant_habitat), 
      
      plant_habitat = map_chr(plant_habitat, .f = plant_hab)
      
    ) %>% 
  
  arrange(Community, desc(mean_core_contribution))

# Combine data for figures
bc_core_contribution_input <- list(
  
  combine_core_contribution_data(bc_core_contribution, c(1:3), "Bacteria and Archaea"), 
  
  combine_core_contribution_data(bc_core_contribution, c(4:6), "Fungi")
  
) %>% 
  
  set_names(nm = c("Bacteria and Archaea", "Fungi"))

# Run ANOVA for contributions
bc_core_contribution_aov <- map(
  
  bc_core_contribution_input, 
  
  .f = get_core_contribution_aov
  
)

# Get ANOVA summaries
bc_core_contribution_aov_summary <- map(
  
  bc_core_contribution_aov, 
  
  .f = summary.aov
  
)

# Run TukeyHSD
bc_core_contribution_tukeyHSD <- map(
  
  bc_core_contribution_aov, 
  
  .f = TukeyHSD
  
)

# P values
bc_core_contribution_aov_Pval <- map(
  
  bc_core_contribution_aov_summary, 
  
  .f = get_core_contribution_aov_Pval
  
) %>% 
  
  bind_rows(.id = "Community") %>% 
  
  mutate(Community = factor(Community, levels = c("Bacteria and Archaea", "Fungi")))

# CLD communities
cld_comm <- c("Bacteria and Archaea", "Fungi")

# CLD
bc_core_contribution_cld <- map2(
  
  bc_core_contribution_aov, 
  
  bc_core_contribution_tukeyHSD, 
  
  .f = get_core_contribution_cld2
  
) %>% 
  
  bind_rows(.id = "Community") %>% 
  
  mutate(Community = factor(Community, levels = c("Bacteria and Archaea", "Fungi")))
```

### Core contribution, Fig. S16  

```{r figS16_functions}
#### Core contribution figure 2 ####

get_core_contribution_fig <- function(core.contribution, core.pval, core.cld) {
  
  tmp1 <- ggplot() + 
    
    geom_point(
      
      data = core.contribution, 
      
      aes(x = betadiv, y = core_contribution, colour = plant_habitat, fill = plant_habitat), 
      
      position = position_jitter(width = 0.25, seed = 12345), 
      
      alpha = 0.25, 
      
      pch = 21, 
      
      size = 0.01
      
    ) + 
    
    geom_boxplot(
      
      data = core.contribution, 
      
      aes(x = betadiv, y = core_contribution, colour = plant_habitat), 
      
      fill = "white", 
      
      outlier.shape = NA,
      
      alpha = 0.4, 
      
      show.legend = FALSE
      
    ) +
    
    # stat_summary(
    #   
    #   data = core.contribution, 
    #   
    #   aes(x = betadiv, y = core_contribution), 
    #   
    #   colour = "black",
    #   
    #   fun.data = mean_cl_normal,
    #   
    #   geom = "errorbar", 
    #   
    #   linewidth = 0.5,
    #   
    #   width = 0,
    #   
    #   show.legend = TRUE
    #   
    # ) + 
    
    stat_summary(
      
      data = core.contribution, 
      
      aes(x = betadiv, y = core_contribution, fill = plant_habitat), 
      
      colour = "black",
      
      fun = "mean", 
      
      geom = "point", 
      
      pch = 21, 
      
      size = 2.5, 
      
      show.legend = TRUE
      
    ) +
    
    geom_text(
      
      data = core.pval,
      
      aes(x = 1.5, y = -Inf, vjust = -0.5, label = pval_label),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    geom_text(
      
      data = core.cld,
      
      aes(x = type, y = Inf, vjust = 1.5, label = cld_label),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      name = "Plant habitat", 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      name = "Plant habitat", 
      
      begin = 0.8, 
      
      end = 0, 
      
    ) + 
    
    labs(
      
      title = NULL, 
      
      x = "Component of beta-diversity", 
      
      y = "Contribution of core community"
      
    ) + 
    
    scale_y_continuous(
      
      limits = c(-0.01, 1.01), 
      
      expand = expansion(mult = c(0.15, 0.15)), 
      
      breaks = c(0, 0.5, 1)
      
    ) + 
    
    facet_grid(
      
      cols = vars(plant_habitat), 
      
      rows = vars(Community)
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0.5), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text = element_text(colour = "black", size = 6), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    ) + 
    
    guides(
      
      fill = guide_legend(override.aes = list(alpha = 1, pch = 21, size = 2.5))
      
    )
  
  return(tmp1)
  
}
```

```{r figS16}
# Prepare figure input
bc_core_contribution_fig_input <- bc_core_contribution_input %>% 
  
  bind_rows(.id = "Community") %>% 
  
  mutate(Community = factor(Community, levels = c("Bacteria and Archaea", "Fungi")))

# Get figure
figS16 <- get_core_contribution_fig(
  
  bc_core_contribution_fig_input, 
  
  bc_core_contribution_aov_Pval, 
  
  bc_core_contribution_cld
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS16.png", 
  
  plot = figS16, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS16.pdf", 
  
  plot = figS16, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)
```

```{r mean_core_contributions}
# Means
mean_core_contributions <- bc_core_contribution_fig_input %>%
  
  group_by(community, plant_habitat, betadiv) %>%
  
  summarise(mean_core_contribution = mean(core_contribution)) %>%
  
  ungroup() %>%
  
  filter(betadiv == "Between dates") %>%
  
  arrange(mean_core_contribution)
```

## Core group relabund  

### Hierarchical clustering  

```{r core_group_relabund_hclust_functions}
#### Core ASV table ####

get_core_asv_table <- function(asv.tab) {
  
  tmp1 <- asv.tab %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.)
  
  return(tmp1)
  
}

#### Z score ####

z_score <- function(x) {
  
  tmp1 <- x - mean(x)
  
  tmp2 <- sd(x)
  
  tmp3 <- tmp1 / tmp2
  
  return(tmp3)
  
}

#### Standardize core ASVs ####

stand_asv_df <- function(asv.table, seq.min) {
  
  tmp1 <- asv.table %>% 
    
    mutate(asv_relabund = n_seqs / seq.min) %>% 
    
    group_by(sample_unique_id) %>% 
    
    mutate(asv_stand = z_score(asv_relabund)) %>% 
    
    ungroup() %>% 
    
    select(sample_unique_id, asv_id, asv_stand) %>% 
    
    pivot_wider(
      
      id_cols = sample_unique_id, 
      
      names_from = "asv_id", 
      
      values_from = "asv_stand"
      
    ) %>% 
    
    column_to_rownames(var = "sample_unique_id") %>% 
    
    as.data.frame(.)
  
  return(tmp1)
  
}

#### Pearson dist ####

pearson_dist <- function(x) {
  
  tmp1 <- as.dist(1 - cor(x, method = "pearson"))
  
  return(tmp1)
  
}

#### Get core Hclust ####

get_core_hclust <- function(x) {
  
  set.seed(12345)
  
  tmp1 <- hclust(x, method = "complete")
  
  return(tmp1)
  
}

#### H-clust variance explained ####

hclust_var <- function(x, y) {
  
  tmp1 <- kmeans(x, centers = y, iter.max = 20)
  
  tmp2 <- tmp1$betweenss / tmp1$totss
  
  return(tmp2)
  
}

#### Get hclust variance explained ####

get_hclust_var <- function(x, vec.length) {
  
  tmp1 <- vector(mode = "numeric", length = vec.length)
  
  for(i in 2:vec.length) {
    
    tmp1[i] <- hclust_var(x, i)
    
  }
  
  return(tmp1)
  
}
```

```{r core_group_relabund_hclust}
# Standardize core ASVs for correlation distance matrix
bc_asv_core_stand_df <- map(bc_asv_core_df, .f = get_core_asv_table) %>% 
  
  map2(., bc_sample_seq_min, .f = stand_asv_df)

# Correlation-based distance matrix
bc_core_pearson_dist <- map(bc_asv_core_stand_df, .f = pearson_dist)

# Hclust
bc_core_hclust <- map(bc_core_pearson_dist, .f = get_core_hclust)

core_hclust_vec_length <-  c(20, 20, 20, 20, 5, 20)

# Hclust variance explained
bc_core_hclust_var <- map2(bc_asv_core_stand_df, core_hclust_vec_length, .f = get_hclust_var)
```

### Scree plots, Fig. S17  

```{r figS17_functions}
#### Get scree input ####

get_scree_input <- function(x) {
  
  tmp1 <- tibble(
    
    n_groups = c(1 : length(x)), 
    
    variance_explained = x
    
  )
  
  return(tmp1)
  
}

#### Get scree plots ####

get_scree_plots <- function(scree.input) {
  
  tmp1 <- ggplot() + 
    
    scale_x_continuous(
      
      limits = c(0, 20)
      
    ) + 
    
    geom_line(
      
      data = scree.input, 
      
      aes(x = n_groups, y = variance_explained, colour = plant_habitat)
      
    ) + 
    
    geom_point(
      
      data = scree.input, 
      
      aes(x = n_groups, y = variance_explained, fill = plant_habitat), 
      
      colour = "black", 
      
      pch = 21, 
      
      size = 2
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    labs(
      
      title = NULL, 
      
      x = "Number of groups", 
      
      y = "Variance explained"
      
    ) + 
    
    facet_grid(
      
      cols = vars(plant_habitat), 
      
      rows = vars(Community)
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )
  
  return(tmp1)
  
}
```

```{r figS17}
# Get input for scree plots
bc_scree_input <- map(
  
  bc_core_hclust_var, 
  
  .f = get_scree_input
  
) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_asv_df")
    
  ) %>% 
  
  separate(
    
    id, 
    
    into = c("Community", "plant_habitat"), 
    
    sep = "_"
    
  ) %>% 
  
  mutate(
    
    Community = ifelse(Community == "16s", "Bacteria and Archaea", Community), 
    
    Community = ifelse(Community == "its", "Fungi", Community), 
    
    Community = factor(Community, levels = c("Bacteria and Archaea", "Fungi")), 
    
    plant_habitat = toupper(plant_habitat), 
    
    plant_habitat = map(plant_habitat, .f = plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Scree plots
figS17 <- get_scree_plots(bc_scree_input)

# Save figure
ggsave2(
  
  filename = "data/results/figS17.png", 
  
  plot = figS17, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS17.pdf", 
  
  plot = figS17, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)
```

### Group selection

```{r core_group_relabund_group_selection_functions}
#### Get core groups ####

get_core_groups <- function(hclust.result, k.value) {
  
  tmp1 <- cutree(hclust.result, k = k.value, h = NULL) %>% 
    
    enframe(name = "asv_id", value = "eco_group")
  
  return(tmp1)
  
}

#### Core group relabund ####

get_core_group_relabund <- function(asv.tab, sample.seq.min, core.groups, metadata.tab) {
  
  tmp1 <- asv.tab %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    pivot_longer(
      
      -sample_unique_id, 
      
      names_to = "asv_id", 
      
      values_to = "n_seqs"
      
    ) %>% 
    
    mutate(asv_relabund = n_seqs / sample.seq.min)
  
  tmp2 <- tmp1 %>% 
    
    inner_join(., core.groups, by = "asv_id") %>% 
    
    group_by(sample_unique_id, eco_group) %>% 
    
    summarise(group_relabund = sum(asv_relabund)) %>% 
    
    ungroup() %>% 
    
    mutate(eco_group = paste("Group", eco_group, sep = " ")) %>% 
    
    inner_join(metadata.tab, ., by = "sample_unique_id")
  
  return(tmp2)
  
}
```

```{r core_group_relabund_group_selection}
# K values
bc_core_k_values <- c(4, 5, 4, 5, 2, 5)

# Get core groups
bc_core_groups <- map2(bc_core_hclust, bc_core_k_values , .f = get_core_groups)

# Get core group relabund
bc_core_groups_relabund <- pmap(
  
  list(bc_asv_core_df, bc_sample_seq_min, bc_core_groups, bc_metadata_habsplit_filtered), 
  
  .f = get_core_group_relabund
  
) %>% 
  
  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh")) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_le"), 
    
    id = str_remove(id, "_re"), 
    
    id = str_remove(id, "_rh")
    
  ) %>% 
  
  rename(Community = "id") %>% 
  
  mutate(
    
    Community = ifelse(Community == "16s", "Bacteria and Archaea", Community), 
    
    Community = ifelse(Community == "its", "Fungi", Community), 
    
    eco_group = factor(
      
      eco_group, 
      
      levels = c("Group 1", "Group 2", "Group 3", "Group 4", "Group 5")
      
    )
    
  )
```

### Core groups, Fig. 3  

```{r fig3_functions}
#### Core groups figures 2 ####

get_core_group_figs2 <- function(fig.data) {
  
  tmp1 <- fig.data %>% 
    
    group_by(plant_habitat, Community, eco_group, year, month, seasonality) %>% 
    
    summarise(mean_relabund = mean(group_relabund)) %>% 
    
    ungroup()
  
  tmp2 <- ggplot() + 
    
    scale_x_discrete(
      
      labels = function(x) seasonality_labels2(x)
      
    ) +
    
    stat_summary(
      
      data = fig.data, 
      
      aes(x = seasonality, y = group_relabund, group = interaction(eco_group, year)), 
      
      fun.data = mean_cl_normal, 
      
      geom = "errorbar", 
      
      colour = "black", 
      
      width = 0
      
    ) + 
    
    stat_summary(
      
      data = fig.data, 
      
      aes(
        
        x = seasonality, 
        
        y = group_relabund, 
        
        group = interaction(eco_group, year), 
        
        fill = eco_group
        
      ), 
      
      fun.data = mean_se, 
      
      geom = "point", 
      
      pch = 21, 
      
      size = 2
      
    ) + 
    
    geom_line(
      
      data = tmp1, 
      
      aes(x = seasonality, y = mean_relabund, group = interaction(eco_group, year), colour = eco_group), 
      
      show.legend = FALSE
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      option = "plasma", 
      
      name = NULL
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      option = "plasma", 
      
      guide = "none"
      
    ) + 
    
    facet_grid(
      
      cols = vars(Community), 
      
      rows = vars(plant_habitat)
      
    ) + 
    
    labs(
      
      title = NULL, 
      
      x = NULL, 
      
      y = "Relative abundance"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0.5), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 45, hjust = 1), 
      
      strip.background = element_blank(), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "right"
      
    ) + 
    
    guides(
      
      fill = guide_legend(override.aes = list(pch = 21, alpha = 1, size = 2.5))
      
    )
  
  return(tmp2)
  
}
```

```{r fig3}
# Plot core groups
fig3 <- get_core_group_figs2(bc_core_groups_relabund)

# Save figure
ggsave2(
  
  filename = "data/results/fig3.png", 
  
  plot = fig3, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/fig3.pdf", 
  
  plot = fig3, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)
```

### Table S10  

```{r tableS10_functions}
#### Get uncultured genera ####

get_uncultured <- function(x) {
  
  tmp1 <- ifelse(x == "uncultured", NA, x)
  
  return(tmp1)
  
}

#### Get unclassified genera ####

get_unclassified_genera <- function(
    
  domain.col,
  
  phylum.col,
  
  class.col,
  
  order.col,
  
  family.col,
  
  genus.col
  
) {
  
  if(is.na(phylum.col) & is.na(class.col) & is.na(order.col) & is.na(family.col) & is.na(genus.col)) {
    
    tmp1 <- paste("Uncl.", domain.col, sep = " ")
    
  } else if(!is.na(phylum.col) & is.na(class.col) & is.na(order.col) & is.na(family.col) & is.na(genus.col)) {
    
    tmp1 <- paste("Uncl.", phylum.col, sep = " ")
    
  } else if(!is.na(phylum.col) & !is.na(class.col) & is.na(order.col) & is.na(family.col) & is.na(genus.col)) {
    
    tmp1 <- paste("Uncl.", class.col, sep = " ")
    
  } else if(!is.na(phylum.col) & !is.na(class.col) & !is.na(order.col) & is.na(family.col) & is.na(genus.col)) {
    
    tmp1 <- paste("Uncl.", order.col, sep = " ")
    
  } else if(!is.na(phylum.col) & !is.na(class.col) & !is.na(order.col) & !is.na(family.col) & is.na(genus.col)) {
    
    tmp1 <- paste("Uncl.", family.col, sep = " ")
    
  } else {
    
    tmp1 <- genus.col
    
  }
  
  return(tmp1)
  
}

#### Get unidentified orders ####

get_unidentified_genera <- function(
    
  domain.col,
  
  phylum.col,
  
  class.col,
  
  order.col,
  
  family.col,
  
  genus.col
  
) {
  
  if(phylum.col == "unidentified" & class.col == "unidentified" & order.col == "unidentified" & family.col == "unidentified" & genus.col == "unidentified") {
    
    tmp1 <- paste("Uncl.", domain.col, sep = " ")
    
  } else if(phylum.col != "unidentified" & class.col == "unidentified" & order.col == "unidentified" & family.col == "unidentified" & genus.col == "unidentified") {
    
    tmp1 <- paste("Uncl.", phylum.col, sep = " ")
    
  } else if(phylum.col != "unidentified" & class.col != "unidentified" & order.col == "unidentified" & family.col == "unidentified" & genus.col == "unidentified") {
    
    tmp1 <- paste("Uncl.", class.col, sep = " ")
    
  } else if(phylum.col != "unidentified" & class.col != "unidentified" & order.col != "unidentified" & family.col == "unidentified" & genus.col == "unidentified") {
    
    tmp1 <- paste("Uncl.", order.col, sep = " ")
    
  } else if(phylum.col != "unidentified" & class.col != "unidentified" & order.col != "unidentified" & family.col != "unidentified" & genus.col == "unidentified") {
    
    tmp1 <- paste("Uncl.", family.col, sep = " ")
    
  } else {
    
    tmp1 <- genus.col
    
  }
  
  return(tmp1)
  
}

#### Clean unclassified ####

clean_unclassified <- function(x) {
  
  tmp1 <- ifelse(x == "unidentified" | is.na(x), "unclassified", x)
  
  return(tmp1)
  
}

#### Get core group taxonomy ####

get_core_group_tax <- function(asv.tab, core.groups, community) {
  
  if(community == "Bacteria and Archaea") {
    
    tmp0 <- bc_tax_filtered$bc_16s_tax_filtered %>% 
      
      rename(Domain_or_kingdom = "Domain")
    
  } else if(community == "Fungi") {
    
    tmp0 <- bc_tax_filtered$bc_its_tax_filtered %>% 
      
      rename(Domain_or_kingdom = "Kingdom")
    
  } else {
    
    print("Error: Pick a community")
    
  }
  
  tmp1 <- asv.tab %>% 
    
    inner_join(tmp0, ., by = "asv_id") %>% 
    
    inner_join(., core.groups, by = "asv_id") %>%
    
    mutate(
      
      Domain_or_kingdom = get_uncultured(Domain_or_kingdom),
      
      Phylum = get_uncultured(Phylum),
      
      Class = get_uncultured(Class),
      
      Order = get_uncultured(Order),
      
      Family = get_uncultured(Family),
      
      Genus = get_uncultured(Genus),
      
      Genus = pmap_chr(
        
        list(Domain_or_kingdom, Phylum, Class, Order, Family, Genus), 
        
        .f = get_unclassified_genera), 
      
      Genus = pmap_chr(
        
        list(Domain_or_kingdom, Phylum, Class, Order, Family, Genus), 
        
        .f = get_unidentified_genera),
      
      Domain_or_kingdom = clean_unclassified(Domain_or_kingdom),
      
      Phylum = clean_unclassified(Phylum),
      
      Class = clean_unclassified(Class),
      
      Order = clean_unclassified(Order),
      
      Family = clean_unclassified(Family),
      
      Genus = clean_unclassified(Genus)
      
    ) %>%
      
    group_by(Domain_or_kingdom, Phylum, Class, Order, Family, Genus, eco_group)
  
  tmp2 <- tmp1 %>%
    
    summarise(tax_n_seqs = sum(n_seqs)) %>%
    
    ungroup() %>% 
    
    group_by(eco_group) %>% 
    
    mutate(tax_relabund = tax_n_seqs / sum(tax_n_seqs)) %>% 
    
    arrange(eco_group, desc(tax_relabund), .by_group = TRUE) %>%
    
    ungroup()
  
  return(tmp2)
  
}
```

```{r tableS10}
# Community vector
comm_vec <- c(rep("Bacteria and Archaea", 3), rep("Fungi", 3))

# Convert asv to long format
bc_core_group_asv <- map(bc_asv_core_df, .f = get_core_asv_table)

# Get group taxonomy
tableS10_input <- pmap(
  
  list(bc_core_group_asv, bc_core_groups, comm_vec), 
  
  .f = get_core_group_tax
  
) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_asv_df")
    
  ) %>% 
  
  separate(
    
    col = id, 
    
    into = c("Community", "Plant habitat"), 
    
    sep = "_"
    
  ) %>% 
  
  mutate(
    
    Community = ifelse(Community == "16s", "Bacteria/Archaea", Community), 
    
    Community = ifelse(Community == "its", "Fungi", Community), 
    
    Community = factor(Community, levels = c("Bacteria/Archaea", "Fungi")), 
    
    `Plant habitat` = toupper(`Plant habitat`), 
    
    `Plant habitat` = map_chr(`Plant habitat`, .f = plant_hab), 
    
    `Plant habitat` = factor(
      
      `Plant habitat`, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere"))
    
  ) %>% 
  
  mutate(
    
    Phylum = ifelse(tax_relabund < 0.1, "Other", Phylum), 
    
    Class = ifelse(tax_relabund < 0.1, "Other", Class), 
    
    Order = ifelse(tax_relabund < 0.1, "Other", Order), 
    
    Family = ifelse(tax_relabund < 0.1, "Other", Family), 
    
    Genus = ifelse(tax_relabund < 0.1, "Other", Genus)
    
  ) %>% 
  
  group_by(
    
    Community, `Plant habitat`, Domain_or_kingdom, 
    
    Phylum, Class, Order, Family, Genus, eco_group
    
  ) %>% 
  
  summarise(`% of group` = 100 * (sum(tax_relabund))) %>% 
  
  ungroup() %>% 
  
  group_by(Community, `Plant habitat`, eco_group) %>% 
  
  arrange(desc(`% of group`), .by_group = TRUE) %>% 
  
  ungroup() %>% 
  
  rename(Group = "eco_group") %>% 
  
  filter(Phylum != "Other")

# Save
write_tsv(
  
  tableS10_input,
  
  file = "data/results/tableS10.txt"
  
)
```

### Key core groups ANOVA  

```{r core_groups_anova_functions}
#### Get core group ANOVA input ####

get_core_group_aov_input <- function(core.relabund, core.group.list) {
  
  tmp1 <- core.relabund %>%
    
    mutate(eco_group_number = str_remove(eco_group, "Group ")) %>%
    
    mutate(eco_group_number = as.numeric(eco_group_number)) %>%
    
    filter(eco_group_number %in% core.group.list)
  
  tmp2 <- unique(tmp1$eco_group)
  
  tmp3 <- tmp1 %>%
    
    group_by(eco_group) %>%
    
    group_split(.) %>%
    
    map(., .f = ungroup) %>%
    
    set_names(nm = tmp2)
  
  return(tmp3)
  
}

#### Core groups ANOVA ####

core_group_aov <- function(x) {
  
  tmp1 <- aov(group_relabund ~ year * month * host_species, data = x)
  
  return(tmp1)
  
}

#### Get core group ANOVA ####

get_core_group_aov <- function(x) {
  
  tmp1 <- map(x, .f = core_group_aov)
  
  return(tmp1)
  
}

#### Get core group ANOVA summary ####

get_core_group_aov_summary <- function(x) {
  
  tmp1 <- map(x, .f = summary.aov)
  
  return(tmp1)
  
}

#### Get core group ANOVA P value ####

get_core_group_seasonality_aov_Pval <- function(x) {
  
  tmp1 <- map(x, .f = get_betadisper_seasonality_aov_Pval)
  
  return(tmp1)
  
}

#### Get core group Tukey HSD ####

get_core_group_tukeyHSD <- function(x) {
  
  tmp1 <- map(x, .f = TukeyHSD)
  
  return(tmp1)
  
}
```

```{r core_groups_anova}
# Format data
bc_core_groups_relabund_habsplit <- bc_core_groups_relabund %>%
  
  group_by(Community, plant_habitat) %>%
  
  group_split(.) %>%
  
  map(., .f = ungroup) %>% 
  
  set_names(nm = names(bc_betadisper))

# Get ANOVA input data
bc_core_groups_relabund_input <- map2(
  
  bc_core_groups_relabund_habsplit,
  
  list(
    
    c(1, 4),
    
    c(1 : 5),
    
    c(1 : 4),
    
    c(2, 5),
    
    2,
    
    c(2, 3)
    
  ),
  
  .f = get_core_group_aov_input
  
)

# ANOVA
bc_core_groups_relabund_aov <- map(
  
  bc_core_groups_relabund_input,
  
  .f = get_core_group_aov
  
)

# Summary
bc_core_groups_relabund_aov_summary <- map(
  
  bc_core_groups_relabund_aov,
  
  .f = get_core_group_aov_summary
  
)
```

### Core group ANOVA table, Table S11

```{r core_group_aov_table_functions}
#### Core group ANOVA table inner ####

get_core_group_aov_table_inner <- function(x, y) {
  
  tmp1 <- x %>%
    
    mutate(eco_group = y)
  
  return(tmp1)
  
}

#### Core group ANOVA table ####

get_core_group_aov_table <- function(x) {
  
  tmp0 <- names(x)
  
  tmp1 <- map(x, .f = get_betadisper_aov_table) %>%
    
    map2(., tmp0, .f = get_core_group_aov_table_inner) %>%
    
    bind_rows() %>%
    
    select(eco_group, everything())
  
  return(tmp1)
  
}
```

```{r tableS11}
# Get Table S11
tableS11_input <- map(
  
  bc_core_groups_relabund_aov_summary,
  
  .f = get_core_group_aov_table
  
) %>% 
  
  bind_rows(.id = "ID") %>%
  
  mutate(ID = str_remove(ID, "bc_")) %>%
  
  separate(
    
    col = ID,
    
    into = c("Community", "Plant habitat"),
    
    sep = "_"
    
  )

# Save
write_tsv(
  
  tableS11_input,
  
  file = "data/results/tableS11.txt"
  
)
```

### Core group host species, Fig. S18  

```{r figS18}
# Get figS17 input
figS18_input <- bc_core_groups_relabund %>%
  
  group_by(Community, plant_habitat, host_species, seasonality, eco_group) %>%
  
  summarise(mean_group_relabund = mean(group_relabund)) %>%
  
  ungroup() %>%
  
  pivot_wider(
    
    id_cols = c(Community, plant_habitat, seasonality, eco_group),
    
    names_from = "host_species",
    
    values_from = "mean_group_relabund"
    
  )

# Correlation
figS18_corr_input <- figS18_input %>%

  select(deltoides, trichocarpa)

figS18_corr <- corr.test(figS18_corr_input, method = "pearson")

figS18_siglabel <- figS18_corr$r %>%

  as.data.frame(.) %>%

  rownames_to_column(var = "host_species") %>%

  as_tibble(.) %>%

  filter(host_species == "deltoides") %>%

  select(host_species, trichocarpa) %>%

  mutate(

    xpos = 0.125,

    ypos = 0.65,

    siglabel = paste("r==", round(trichocarpa, 2), "*','~", "italic(P)<0.001", sep = "")

  )

# Get figure
figS18 <- ggplot() +
  
  geom_abline(
    
    slope = 1,
    
    intercept = 0,
    
    linewidth = 0.5,
    
    linetype = 2
    
  ) +
  
  geom_point(
    
    data = figS18_input,
    
    aes(
      
      x = trichocarpa,
      
      y = deltoides,
      
      fill = plant_habitat,
      
      shape = Community
      
    ),
    
    colour = "black", 
    
    size = 2,
    
    alpha = 0.6
    
  ) +
  
  scale_shape_manual(
    
    name = "Community",
    
    values = c(21, 22)
    
  ) +
  
  scale_colour_viridis(
    
    discrete = TRUE, 
    
    begin = 0.8, 
    
    end = 0, 
    
    guide = "none"
    
  ) + 
  
  scale_fill_viridis(
    
    name = "Plant habitat",
    
    discrete = TRUE, 
    
    begin = 0.8, 
    
    end = 0
    
  ) + 
  
  geom_text(
    
    data = figS18_siglabel,
    
    aes(x = xpos, y = ypos, label = siglabel),
    
    parse = TRUE,
    
    size = 4
    
  ) +
  
  geom_smooth(
    
    data = figS18_input,
    
    aes(
      
      x = trichocarpa,
      
      y = deltoides,
      
      colour = plant_habitat,
      
      group = interaction(Community, plant_habitat)
      
    ),
    
    method = "lm",
    
    formula = y ~ x,
    
    se = FALSE
    
  ) +
  
  labs(
    
    title = NULL,
    
    x = bquote("Bray-Curtis dissimilarity ("~italic(P.~trichocarpa)*")"), 
    
    y = bquote("Bray-Curtis dissimilarity ("~italic(P.~deltoides)*")")
    
  ) +
  
  theme(
    
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
    
    panel.background = element_blank(), 
    
    panel.grid = element_blank(), 
    
    strip.background = element_blank(), 
    
    plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
    
    axis.title = element_text(colour = "black", size = 10), 
    
    axis.ticks = element_line(colour = "black", linewidth = 0.25), 
    
    axis.text.x = element_text(colour = "black", size = 8), 
    
    axis.text.y = element_text(colour = "black", size = 8), 
    
    strip.text = element_text(colour = "black", size = 10), 
    
    legend.key = element_blank(), 
    
    legend.title = element_text(colour = "black", size = 10), 
    
    legend.text = element_text(colour = "black", size = 8)
    
  ) +
  
  guides(
      
      shape = guide_legend(override.aes = list(
        
        alpha = 0.6, 
        
        pch = c(21, 22),  
        
        size = 2.5, 
        
        colour = "black", 
        
        fill = "black"
        
      )),
      
      fill = guide_legend(override.aes = list(
        
        alpha = 0.6, 
        
        pch = 21,  
        
        size = 2.5, 
        
        colour = "black"
        
      ))
      
    )

# Save figure
ggsave2(
  
  filename = "data/results/figS18.png", 
  
  plot = figS18, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS18.pdf", 
  
  plot = figS18, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 5, 
  
  units = "in"
  
)
```

# Network  

## Prepare input data for SPIEC-EASI network analyis  

```{r prepare_data_for_spieceasi_full_input_functions}
#### Trim ASVs for network ####

trim_asv_for_network <- function(asv.tab, prop.samples, sample.seq.min) {
  
  tmp1 <- prop.samples
  
  tmp2 <- asv.tab %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "asv_id") %>% 
    
    pivot_longer(
      
      -asv_id, 
      
      names_to = "sample_unique_id", 
      
      values_to = "n_seqs"
      
    ) %>% 
    
    group_by(sample_unique_id) %>% 
    
    mutate(sample_n_seqs = sum(n_seqs)) %>% 
    
    ungroup()
  
  tmp3 <- length(unique(tmp2$sample_unique_id))
  
  tmp4 <- tmp2 %>%
    
    mutate(pres_abs = ifelse(n_seqs > 0, 1, 0)) %>%
    
    group_by(asv_id) %>%
    
    summarise(samples_pres = sum(pres_abs)) %>%
    
    ungroup() %>%
    
    mutate(prop_samples_pres = samples_pres / tmp3) %>%
    
    filter(prop_samples_pres >= tmp1) %>%
    
    pull(asv_id)
  
  tmp5 <- tmp2 %>%
    
    filter(asv_id %in% tmp4) %>%
    
    group_by(sample_unique_id) %>% 
    
    mutate(remaining_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    filter(remaining_n_seqs >= sample.seq.min) %>% 
    
    select(asv_id, sample_unique_id, n_seqs) %>% 
    
    arrange(asv_id, sample_unique_id) %>% 
    
    asv_long_to_wide_df(.)
  
  return(tmp5)
  
}

#### Get network sample filter ####

get_network_sample_filter <- function(asv.tab.16s, asv.tab.its) {
  
  tmp1 <- asv.tab.16s %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    select(sample_unique_id)
  
  tmp2 <- asv.tab.its %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    select(sample_unique_id)
  
  tmp3 <- inner_join(tmp1, tmp2, by = "sample_unique_id") %>% 
    
    pull(sample_unique_id)
  
  return(tmp3)
  
}

#### Filter ASVs by sample for network ####

filter_asv_for_network <- function(asv.tab, sample.filter) {
  
  tmp1 <- asv.tab %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    filter(sample_unique_id %in% sample.filter) %>% 
    
    drop_0seq_asv(.) %>% 
    
    drop_0seq_samples(.) %>% 
    
    pivot_wider(
      
      names_from = "asv_id", 
      
      values_from = "n_seqs", 
      
      values_fill = 0
      
    ) %>% 
    
    arrange(sample_unique_id) %>% 
    
    as.data.frame() %>% 
    
    column_to_rownames(var = "sample_unique_id")
  
  return(tmp1)
  
}

#### Filter metadata for network input ####

filter_metadata_for_network <- function(asv.tab, metadata.tab) {
  
  tmp1 <- rownames(asv.tab)
  
  tmp2 <- metadata.tab %>% 
    
    filter(sample_unique_id %in% tmp1) %>% 
    
    arrange(sample_unique_id) %>% 
    
    as.data.frame(.) %>% 
    
    column_to_rownames(var = "sample_unique_id")
  
  return(tmp2)
  
}

#### Filter taxonomy for network input ####

filter_taxonomy_for_network <- function(asv.tab, tax.tab) {
  
  tmp1 <- colnames(asv.tab)
  
  tmp2 <- tax.tab %>% 
    
    filter(asv_id %in% tmp1) %>% 
    
    arrange(match(asv_id, tmp1)) %>% 
    
    as.data.frame(.) %>% 
    
    column_to_rownames(var = "asv_id")
  
  return(tmp2)
  
}
```

```{r prepare_data_for_spieceasi_full_input}
# Filter ASVs not present in at least 20% of samples
bc_asv_for_network_trimmed <- pmap(

  list(bc_asv_habsplit_df_SRS_input, 0.2, 100),

  .f = trim_asv_for_network

) %>%

  set_names(nm = c("bc_16s_le", "bc_16s_re", "bc_16s_rh", "bc_its_le", "bc_its_re", "bc_its_rh"))

# Get filter to match 16S and ITS samples
bc_sample_network_filter <- map2(

  bc_asv_for_network_trimmed[1:3],

  bc_asv_for_network_trimmed[4:6],

  .f = get_network_sample_filter

) %>%

  set_names(nm = c("LE", "RE", "RH"))

# Filter ASV tables to only include matched samples between domains (for network analysis)
bc_asv_for_network_filtered <- c(

  map2(

    bc_asv_for_network_trimmed[1:3],

    bc_sample_network_filter,

    .f = filter_asv_for_network

  ),

  map2(

    bc_asv_for_network_trimmed[4:6],

    bc_sample_network_filter,

    .f = filter_asv_for_network

  )

)

# Filter metadata tables to include matched samples between domains (for network analysis)
bc_metadata_for_network_filtered <- map2(

  bc_asv_for_network_filtered,

  bc_metadata_habsplit_filtered,

  .f = filter_metadata_for_network

)

# Filter taxonomy tables to match ASVs
bc_tax_for_network_filtered <- c(

  map2(

    bc_asv_for_network_filtered[1:3],

    bc_tax_filtered[1],

    .f = filter_taxonomy_for_network

  ),

  map2(

    bc_asv_for_network_filtered[4:6],

    bc_tax_filtered[2],

    .f = filter_taxonomy_for_network

  )

)

# Create matrices for SPIEC-EASI input
bc_spieceasi_input <- map(bc_asv_for_network_filtered, .f = as.matrix)

# Save for HPC
write_rds(

  bc_spieceasi_input,

  file = "data/working/bc_spieceasi_input.rds"

)
```

## Construct networks (HPC) using SPIEC-EASI  

```{r full_network_construction}
# bc_spieceasi_input <- readRDS(file = "input/bc_spieceasi_input.rds")
# 
# # LE, SPIEC-EASI ## DONE 1-11
# bc_le_se_network <- spiec.easi(
# 
#   list(bc_spieceasi_input$bc_16s_le, bc_spieceasi_input$bc_its_le),
# 
#   method = "mb",
# 
#   nlambda = 50,
# 
#   lambda.min.ratio = 1e-3,
# 
#   pulsar.params = list(thresh = 0.01, seed = 12345, ncores = 32)
# 
# )
# 
# saveRDS(bc_le_se_network, file = "output/bc_le_spieceasi_network.rds")
# 
# bc_le_edge <- symBeta(getOptBeta(bc_le_se_network), mode = 'maxabs')
# 
# colnames(bc_le_edge) <- c(colnames(bc_spieceasi_input$bc_16s_le), colnames(bc_spieceasi_input$bc_its_le))
# 
# saveRDS(bc_le_edge, file = "output/bc_le_spieceasi_network_matrix.rds")
# 
# # RE, SPIEC-EASI ## DONE 1-11
# bc_re_se_network <- spiec.easi(
# 
#   list(bc_spieceasi_input$bc_16s_re, bc_spieceasi_input$bc_its_re),
# 
#   method = "mb",
# 
#   nlambda = 50,
# 
#   lambda.min.ratio = 1e-3,
# 
#   pulsar.params = list(thresh = 0.01, seed = 12345, ncores = 32)
# 
# )
# 
# saveRDS(bc_re_se_network, file = "output/bc_re_spieceasi_network.rds")
# 
# bc_re_edge <- symBeta(getOptBeta(bc_re_se_network), mode = 'maxabs')
# 
# colnames(bc_re_edge) <- c(colnames(bc_spieceasi_input$bc_16s_re), colnames(bc_spieceasi_input$bc_its_re))
# 
# saveRDS(bc_re_edge, file = "output/bc_re_spieceasi_network_matrix.rds")
# 
# # RH, SPIEC-EASI
# bc_rh_se_network <- spiec.easi(
# 
#   list(bc_spieceasi_input$bc_16s_rh, bc_spieceasi_input$bc_its_rh),
# 
#   method = "mb",
# 
#   nlambda = 50,
# 
#   lambda.min.ratio = 1e-3,
# 
#   pulsar.params = list(thresh = 0.01, seed = 12345, ncores = 32)
# 
# )
# 
# saveRDS(bc_rh_se_network, file = "output/bc_rh_spieceasi_network.rds")
# 
# bc_rh_edge <- symBeta(getOptBeta(bc_rh_se_network), mode = 'maxabs')
# 
# colnames(bc_rh_edge) <- c(colnames(bc_spieceasi_input$bc_16s_rh), colnames(bc_spieceasi_input$bc_its_rh))
# 
# saveRDS(bc_rh_edge, file = "output/bc_rh_spieceasi_network_matrix.rds")
```

## Convert networks to igraph objects  

```{r spieceasi_to_igraph_functions}
#### Make igraph object from matrix ####

make_igraph_from_matrix <- function(x) {
  
  tmp1 <- graph_from_adjacency_matrix(
    
    x, 
    
    mode = "undirected", 
    
    weighted = TRUE
    
  )
  
  return(tmp1)
  
}

#### Get node metadata ####

get_node_metadata <- function(tax.16s, tax.its, network) {
  
  tmp1 <- tax.16s %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "asv_id")
  
  tmp2 <- tax.its %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "asv_id")
  
  tmp3 <- tmp2 %>% 
    
    rename(Domain = "Kingdom") %>% 
    
    bind_rows(tmp1, .)
  
  tmp4 <- tibble(
    
    asv_id = V(network)$name
    
  ) %>%
    
    inner_join(., tmp3, by = "asv_id") %>%
    
    mutate(
      
      v_degree = degree(network),
      
      v_betweenness = betweenness(
        
        network,
        
        weights = abs(E(network)$weight),
        
        directed = FALSE
        
      )
      
    ) %>%
    
    mutate(
      
      hub = ifelse(
        
        v_degree > quantile(v_degree, 0.9) & v_betweenness > quantile(v_betweenness, 0.9),
        
        "hub",
        
        "non_hub"
        
      )
      
    )
  
  return(tmp4)
  
}

#### Get edge metadata ####

get_edge_metadata <- function(tax.16s, tax.its, network) {
  
  tmp1 <- as_long_data_frame(network) %>% 
    
    rename(
      
      from_index = "from", 
      
      to_index = "to", 
      
      edge_weight = "weight", 
      
      from_node = `ver[el[, 1], ]`, 
      
      to_node = `ver2[el[, 2], ]`
      
    ) %>% 
    
    select(
      
      from_node, 
      
      to_node, 
      
      from_index, 
      
      to_index, 
      
      edge_weight
      
    )
  
  tmp2 <- tax.16s %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "asv_id")
  
  tmp3 <- tax.its %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "asv_id")
  
  tmp4 <- tmp3 %>% 
    
    rename(Domain = "Kingdom") %>% 
    
    bind_rows(tmp2, .)
  
  tmp5 <- tmp4 %>% 
    
    rename(node = "asv_id") %>% 
    
    rename_with(.fn = ~ paste("from", .x, sep = "_")) %>% 
    
    inner_join(tmp1, ., by = "from_node")
  
  tmp6 <- tmp4 %>% 
    
    rename(node = "asv_id") %>% 
    
    rename_with(.fn = ~ paste("to", .x, sep = "_")) %>% 
    
    inner_join(tmp5, ., by = "to_node")
  
  return(tmp6)
  
}
```

```{r spieceasi_to_igraph}
# List matrix files
bc_network_matrix_files <- list.files(
  
  "data/working/spieceasi_output", 
  
  full.names = TRUE
  
)

# Read in matrices
bc_network_matrix_input <- map(
  
  bc_network_matrix_files, 
  
  .f = read_rds
  
) %>% 
  
  set_names(nm = c("LE", "RE", "RH"))

# Make igraph objects
bc_network_igraph <- map(
  
  bc_network_matrix_input, 
  
  .f = make_igraph_from_matrix
  
)

# Get taxonomy input
bc_16s_tax_for_network_filtered <- bc_tax_for_network_filtered[1:3]

bc_its_tax_for_network_filtered <- bc_tax_for_network_filtered[4:6]

# Generate node metadata
bc_network_node_metadata <- pmap(
  
  list(
    
    bc_16s_tax_for_network_filtered, 
    
    bc_its_tax_for_network_filtered, 
    
    bc_network_igraph
    
  ), 
  
  .f = get_node_metadata
  
) %>% 
  
  set_names(nm = str_remove_all(names(.), "16s_"))

# Generate edge metadata
bc_network_edge_metadata <- pmap(
  
  list(
    
    bc_16s_tax_for_network_filtered, 
    
    bc_its_tax_for_network_filtered, 
    
    bc_network_igraph
    
  ), 
  
  .f = get_edge_metadata
  
) %>% 
  
  set_names(nm = str_remove_all(names(.), "16s_"))
```

```{r degree_distribution_functions}
#### Get degree distribution ####

get_degree_distribution <- function(x) {
  
  tmp1 <- degree_distribution(x)
  
  tmp2 <- tibble(
    
    Edges = c(0 : (length(tmp1) - 1)), 
    
    `Proportion of ASVs` = tmp1
    
  )
  
  return(tmp2)
  
}
```

```{r degree_distribution}
# Degree distribution
bc_degree_distribution <- map(
  
  bc_network_igraph, 
  
  .f = get_degree_distribution
  
) %>% 
  
  bind_rows(.id = "plant_habitat") %>% 
  
  mutate(
    
    plant_habitat = map_chr(plant_habitat, .f = plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )
```

```{r edge_input}
# Edge input
bc_edge_input <- bc_network_edge_metadata %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = toupper(id)
    
    ) %>% 
  
  rename(sample_type = "id") %>% 
  
  mutate(
    
    plant_habitat = map_chr(sample_type, .f = plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere"))
    
  ) %>% 
  
  mutate(edge_direction = ifelse(edge_weight < 0, "Negative", "Positive"))
```

### Network summary, Table S12  

```{r tableS12_functions}
#### Lower 95% CI ####

get_LCI <- function(x) {
  
  tmp1 <- t.test(x)
  
  tmp2 <- tmp1$conf.int[1]
  
  return(tmp2)
  
}

#### Upper 95% CI ####

get_UCI <- function(x) {
  
  tmp1 <- t.test(x)
  
  tmp2 <- tmp1$conf.int[2]
  
  return(tmp2)
  
}

#### Get network summary data ####

get_network_summary_data = function(network, node.metadata, edge.metadata) {
  
  # Mean shortest path
  tmp0 <- mean_distance(
    
    network,
    
    weights = abs(E(network)$weight),
    
    directed = FALSE,
    
    unconnected = TRUE,
    
    details = FALSE
    
  )
  
  # Nodes
  tmp1 <- node.metadata %>% 
    
    nrow(.)
  
  # Nodes with >= 1 edge
  tmp2 <- node.metadata %>% 
    
    filter(v_degree > 0) %>% 
    
    nrow(.)
  
  # Degree, mean and SE
  tmp3a <- node.metadata %>% 
    
    summarise(
      
      mean_v_degree = mean(v_degree), 
      
      lci_v_degree = get_LCI(v_degree), 
      
      uci_v_degree = get_UCI(v_degree)
      
    )
  
  # Betweenness, mean and SE
  tmp3b <- node.metadata %>% 
    
    summarise(
      
      mean_v_betweenness = mean(v_betweenness), 
      
      lci_v_betweenness = get_LCI(v_betweenness), 
      
      uci_v_betweenness = get_UCI(v_betweenness)
      
    )
  
  # Diversity
  tmp4 <- diversity(
    
    network, 
    
    weights = abs(E(network)$weight), 
    
    vids = V(network)
    
  )
  
  tmp5 <- tibble(
    
    mean_heterogeneity = mean(!is.na(tmp4)), 
    
    lci_heterogeneity = get_LCI(!is.na(tmp4)), 
    
    uci_heterogeneity = get_UCI(!is.na(tmp4))
    
  )
  
  # # Module membership
  # tmp6 <- cluster_edge_betweenness(
  #   
  #   network, 
  #   
  #   weights = abs(E(network)$weight), 
  #   
  #   directed = FALSE, 
  #   
  #   edge.betweenness = TRUE, 
  #   
  #   merges = TRUE, 
  #   
  #   bridges = TRUE, 
  #   
  #   modularity = TRUE, 
  #   
  #   membership = TRUE
  #   
  # )
  
  # # Calculate modularity
  # tmp7 <- modularity(
  #   
  #   network, 
  #   
  #   membership = tmp6$membership, 
  #   
  #   weights = abs(E(network)$weight), 
  #   
  #   resolution = 1, 
  #   
  #   directed = FALSE
  #   
  # )
  
  # Clustering coefficient
  tmp8 <- transitivity(
    
    network, 
    
    type = c("undirected"), 
    
    vids = V(network), 
    
    weights = abs(E(network)$weight), 
    
    isolates = c("NaN")
    
  )
  
  # Density
  tmp9 <- edge_density(network, loops = FALSE)
  
  # Number of edges
  tmp10a <- edge.metadata %>% 
    
    nrow(.)
  
  # Number of positive edges
  tmp10b <- edge.metadata %>% 
    
    filter(edge_weight > 0) %>% 
    
    nrow(.)
  
  # Number of negative edges
  tmp10c <- edge.metadata %>% 
    
    filter(edge_weight < 0) %>% 
    
    nrow(.)
  
  # Number of hub taxa
  tmp11 <- node.metadata %>% 
    
    filter(hub == "hub") %>% 
    
    nrow(.)
  
  # Compile summary data
  tmp12 <- tibble(
    
    Nodes = tmp1, 
    
    `Nodes with edges` = tmp2, 
    
    # Modularity = tmp7, 
    
    `Clustering coefficient` = tmp8, 
    
    Density = tmp9, 
    
    Edges = tmp10a, 
    
    `Positive edges` = tmp10b, 
    
    `Negative edges` = tmp10c, 
    
    `Hub ASVs` = tmp11, 
    
    `Mean shortest path` = tmp0
    
  ) %>% 
    
    bind_cols(
      
      ., 
      
      tmp3a, 
      
      tmp3b, 
      
      tmp5
      
    )
  
  return(tmp12)
  
}
```

```{r tableS12}
# Network summary
network_summary <- pmap(

  list(

    bc_network_igraph,

    bc_network_node_metadata,

    bc_network_edge_metadata

  ),

  .f = get_network_summary_data

)

# Save network summary data
write_rds(

  network_summary,

  file = "data/working/network_summary.rds"

)

# Read in
network_summary <- read_rds(

  file = "data/working/network_summary.rds"

) %>%

  bind_rows(.id = "Plant habitat") %>%

  mutate(

    `Plant habitat` = map_chr(`Plant habitat`, .f = plant_hab),

    `Plant habitat` = factor(

      `Plant habitat`,

      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")

    )

  )

# Table S12
tableS12_input <- network_summary %>% 
  
  mutate(
    
    `Clustering coefficient` = format(round(`Clustering coefficient`, digits = 3), nsmall = 3), 
    
    `Mean shortest path` = format(round(`Mean shortest path`, digits = 3), nsmall = 3), 
    
    Density = format(round(Density, digits = 4), nsmall = 4), 
    
    Degree = paste(
      
      format(round(mean_v_degree, digits = 3), nsmall = 3), 
      
      " [", 
      
      format(round(lci_v_degree, digits = 3), nsmall = 3), 
      
      ", ", 
      
      format(round(uci_v_degree, digits = 3), nsmall = 3), 
      
      "]", 
      
      sep = ""
      
    ), 
    
    Heterogeneity = paste(
      
      format(round(mean_heterogeneity, digits = 3), nsmall = 3), 
      
      " [", 
      
      format(round(lci_heterogeneity, digits = 3), nsmall = 3), 
      
      ", ", 
      
      format(round(uci_heterogeneity, digits = 3), nsmall = 3), 
      
      "]", 
      
      sep = ""
      
    )
    
  ) %>% 
  
  select(
    
    `Plant habitat`, 
    
    Nodes, 
    
    `Nodes with edges`, 
    
    Edges, 
    
    `Positive edges`, 
    
    `Negative edges`, 
    
    Density, 
    
    Degree, 
    
    `Clustering coefficient`, 
    
    Heterogeneity, 
    
    `Mean shortest path`, 
    
    `Hub ASVs`
    
  )  %>% 
  
  mutate(across(Nodes : `Hub ASVs`, as.character)) %>% 
  
  pivot_longer(
    
    cols = -`Plant habitat`, 
    
    names_to = "Network characteristic", 
    
    values_to = "value"
    
  ) %>% 
  
  pivot_wider(
    
    id_cols = `Network characteristic`, 
    
    names_from = "Plant habitat", 
    
    values_from = "value"
    
  )

# Save
write_tsv(
  
  tableS12_input,
  
  file = "data/results/tableS12.txt"
  
)
```

# Key hub ASVs  

## Get important ASVs  

```{r important_asvs_functions}
#### Get abundant ASVs ####

get_abundant_asv <- function(asv.table, threshold) {
  
  tmp1 <- asv.table %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    group_by(asv_id) %>% 
    
    summarise(asv_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    mutate(asv_perc = 100 * (asv_n_seqs / sum(asv_n_seqs))) %>% 
    
    filter(asv_perc >= threshold) %>% 
    
    pull(asv_id)
  
  return(tmp1)
  
}

#### Get important ASVs ####

get_important_asvs <- function(
    
  node.metadata, 
  
  core.asv.16s, 
  
  core.asv.its, 
  
  abund.asv.16s, 
  
  abund.asv.its
  
  ) {
  
  # Get hub ASVs
  tmp1 <- node.metadata %>% 
    
    filter(hub == "hub")
  
  # Filter to include only core taxa and those over a certain abundance threshold
  tmp2 <- tmp1 %>% 
    
    filter(asv_id %in% c(core.asv.16s, core.asv.its)) %>% 
    
    filter(asv_id %in% c(abund.asv.16s, abund.asv.its))
  
  # Return
  return(tmp2)
  
}
```

```{r important_asvs}
# Get filter for ASVs over a certain threshold
bc_abund_asv <- map2(
  
  bc_asv_habsplit_df_SRS_output, 
  
  0.1, 
  
  .f = get_abundant_asv
  
)

# Get ASVs that are hubs, correlated with PCoA, and core taxa
bc_important_asvs <- pmap(
  
  list(
    
    bc_network_node_metadata, 
    
    bc_core_asv_filter[1:3], 
    
    bc_core_asv_filter[4:6], 
    
    bc_abund_asv[1:3], 
    
    bc_abund_asv[4:6]
    
  ), 
  
  .f = get_important_asvs
  
)
```

## Prepare data for heatmaps  

```{r prepare_data_for_heatmaps_functions}
#### Get ASV seq totals for labels ####

get_asv_seq_totals <- function(x) {
  
  tmp1 <- x %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    group_by(asv_id) %>% 
    
    summarise(asv_n_seqs = sum(n_seqs)) %>% 
    
    ungroup()
  
  return(tmp1)
  
}

#### Add ASV labels ####

add_asv_labels <- function(x, community) {
  
  tmp1 <- x %>% 
    
    arrange(desc(asv_n_seqs)) %>% 
    
    mutate(asv_number = c(1 : nrow(.)))
  
  if(community == "bc_16s") {
    
    tmp2 <- tmp1 %>% 
      
      mutate(asv_label = paste("Bac./Arc. ASV", asv_number, sep = " "))
    
  } else if(community == "bc_its") {
    
    tmp2 <- tmp1 %>% 
      
      mutate(asv_label = paste("Fun. ASV", asv_number, sep = " "))
    
  } else {
    
    tmp2 <- tmp1
    
  }
  
  tmp3 <- tmp2 %>% 
    
    select(asv_id, asv_label)
  
  return(tmp3)
  
}

#### Calculate important ASV Z scores ####

get_important_asv_z <- function(asv.table, metadata.table, important.asvs, community) {
  
  tmp1 <- asv.table %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    inner_join(metadata.table, ., by = "sample_unique_id") %>% 
    
    group_by(seasonality, asv_id) %>% 
    
    summarise(mean_n_seqs = mean(n_seqs)) %>% 
    
    ungroup() %>% 
    
    group_by(asv_id) %>% 
    
    mutate(asv_z = z_score(mean_n_seqs)) %>% 
    
    ungroup() %>% 
    
    filter(asv_id %in% important.asvs$asv_id)
  
  if(community == "bc_16s") {
    
    tmp2 <- tmp1 %>% 
      
      inner_join(., bc_asv_labels$bc_16s, by = "asv_id")
    
  } else if(community == "bc_its") {
    
    tmp2 <- tmp1 %>% 
      
      inner_join(., bc_asv_labels$bc_its, by = "asv_id")
    
  }
  
  return(tmp2)
  
}

#### Combine ASV Z ####

combine_asv_z <- function(z.scores.16s, z.scores.its, plant.hab) {
  
  tmp1 <- bind_rows(z.scores.16s, z.scores.its) %>% 
    
    mutate(sample_type = plant.hab)
  
  return(tmp1)
  
}

#### hclust 2 ####

get_hclust2 <- function(x) {
  
  set.seed(12345)
  
  tmp1 <- hclust(x, method = "ward.D2")
  
  return(tmp1)
  
}

#### Get ASV orders ####

get_asv_order <- function(x) {
  
  tmp1 <- x %>% 
    
    select(seasonality, asv_label, asv_z) %>% 
    
    mutate(asv_z = 3 - asv_z) %>% 
    
    pivot_wider(
      
      id_cols = asv_label, 
      
      names_from = "seasonality", 
      
      values_from = "asv_z"
      
    ) %>% 
    
    column_to_rownames(var = "asv_label") %>% 
    
    as.data.frame(.)
  
  # tmp2 <- pearson_dist(tmp1)
  tmp2 <- vegdist(tmp1, method = "euclidean")
  
  # tmp3 <- get_core_hclust(tmp2)
  tmp3 <- get_hclust2(tmp2)

  tmp4 <- as.dendrogram(tmp3)

  tmp5 <- order.dendrogram(tmp4)

  tmp6 <- tibble(

    asv_label = rownames(tmp1),

    asv_order = tmp5

  )
  
  return(tmp6)
  
}

#### Format input for heat maps ####

format_heatmap_input <- function(asv.z, asv.order) {
  
  tmp1 <- asv.z %>% 
    
    inner_join(., asv.order, by = "asv_label") %>% 
    
    arrange(desc(asv_order)) %>% 
    
    mutate(asv_label = factor(asv_label, levels = unique(asv_label)))
  
  return(tmp1)
  
}
```

```{r prepare_data_for_heatmaps}
# Get an index of new labels for ASVs
bc_asv_labels <- map(bc_asv_habsplit_df_SRS_output, .f = get_asv_seq_totals) %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = str_remove(id, "_le_asv_df"), 
    
    id = str_remove(id, "_re_asv_df"), 
    
    id = str_remove(id, "_rh_asv_df")
    
  ) %>% 
  
  rename(community = "id") %>% 
  
  group_by(community, asv_id) %>% 
  
  summarise(asv_n_seqs = sum(asv_n_seqs)) %>% 
  
  ungroup() %>% 
  
  group_by(community) %>% 
  
  group_split(., .keep = FALSE) %>% 
  
  map(., .f = ungroup) %>% 
  
  set_names(nm = c("bc_16s", "bc_its")) %>% 
  
  map2(., c("bc_16s", "bc_its"), .f = add_asv_labels)

# Calculate important ASV z scores
bc_important_asv_z <- pmap(
  
  list(
    
    bc_asv_habsplit_df_SRS_output, 
    
    bc_metadata_habsplit_filtered, 
    
    c(bc_important_asvs, bc_important_asvs), 
    
    c("bc_16s", "bc_16s", "bc_16s", "bc_its", "bc_its", "bc_its")
    
  ), 
  
  .f = get_important_asv_z
  
)

# Combine ASV z scores
bc_important_asv_z_combined <- pmap(
  
  list(
    
    bc_important_asv_z[1:3], 
    
    bc_important_asv_z[4:6], 
    
    c("LE", "RE", "RH")
    
  ), 
  
  .f = combine_asv_z
  
) %>% 
  
  set_names(nm = c("LE", "RE", "RH"))

# Important ASV taxonomy
bc_important_asv_tax_pre <- bc_tax_filtered$bc_its_tax_filtered %>% 
  
  rename(Domain = "Kingdom") %>% 
  
  bind_rows(bc_tax_filtered$bc_16s_tax_filtered, .)

bc_important_asv_tax <- bc_important_asv_z_combined %>% 
  
  bind_rows() %>% 
  
  select(asv_id, asv_label) %>% 
  
  distinct() %>%
  
  inner_join(., bc_important_asv_tax_pre, by = "asv_id")

write_tsv(bc_important_asv_tax, file = "data/results/important_asv_tax.txt")

# Get ASV orders from correlations and hclust
bc_asv_order <- map(
  
  bc_important_asv_z_combined, 
  
  .f = get_asv_order
  
)

# Format input for heat maps
bc_heatmap_input <- map2(
  
  bc_important_asv_z_combined, 
  
  bc_asv_order, 
  
  .f = format_heatmap_input
  
)

# Get z score range
bc_asv_z_score_range <- bc_important_asv_z_combined %>% 
  
  bind_rows(.id = "id") %>% 
  
  summarise(
    
    min_z = min(asv_z), 
    
    max_z = max(asv_z)
    
  )
```

### Hub ASV heatmap, Fig. 4  

```{r fig4_functions}
#### Get hub tax ####

format_tax_label <- function(x) {
  
  tmp1 <- x %>%
    
    mutate(
      
      Phylum = ifelse(Phylum == "uncultured" | Phylum == "unidentified", "unclassified", Phylum), 
      
      Class = ifelse(Class == "uncultured" | Class == "unidentified", "unclassified", Class), 
      
      Order = ifelse(Order == "uncultured" | Order == "unidentified", "unclassified", Order), 
      
      Family = ifelse(Family == "uncultured" | Family == "unidentified", "unclassified", Family), 
      
      Genus = ifelse(Genus == "uncultured" | Genus == "unidentified", "unclassified", Genus)
      
    ) %>%
  
  mutate(
    
    tax_label = ifelse(Genus != "unclassified", Genus, NA),
    
    tax_label = ifelse(
      
      Phylum == "unclassified" & Class == "unclassified" & Order == "unclassified" & Family == "unclassified" & Genus == "unclassified", 
      
      paste("Uncl.", `Domain or kingdom`, sep = " "),
      
      tax_label
      
    ),
    
    tax_label = ifelse(
      
      Phylum != "unclassified" & Class == "unclassified" & Order == "unclassified" & Family == "unclassified" & Genus == "unclassified", 
      
      paste("Uncl.", Phylum, sep = " "),
      
      tax_label
      
    ),
    
    tax_label = ifelse(
      
      Phylum != "unclassified" & Class != "unclassified" & Order == "unclassified" & Family == "unclassified" & Genus == "unclassified", 
      
      paste("Uncl.", Class, sep = " "),
      
      tax_label
      
    ),
    
    tax_label = ifelse(
      
      Phylum != "unclassified" & Class != "unclassified" & Order != "unclassified" & Family == "unclassified" & Genus == "unclassified", 
      
      paste("Uncl.", Order, sep = " "),
      
      tax_label
      
    ),
    
    tax_label = ifelse(
      
      Phylum != "unclassified" & Class != "unclassified" & Order != "unclassified" & Family != "unclassified" & Genus == "unclassified", 
      
      paste("Uncl.", Family, sep = " "),
      
      tax_label
      
    )
    
  )
  
  return(tmp1)
  
}

#### Get hub summary table ####

get_hub_summary <- function(node.metadata) {
  
  tmp1 <- node.metadata %>% 
    
    filter(hub == "hub") %>% 
    
    mutate(
      
      Phylum = ifelse(is.na(Phylum), "unclassified", Phylum), 
      
      Class = ifelse(is.na(Class), "unclassified", Class), 
      
      Order = ifelse(is.na(Order), "unclassified", Order), 
      
      Family = ifelse(is.na(Family), "unclassified", Family), 
      
      Genus = ifelse(is.na(Genus), "unclassified", Genus)
      
    ) %>% 
    
    right_join(bc_asv_labels$bc_16s, ., by = "asv_id") %>% 
    
    right_join(bc_asv_labels$bc_its, ., by = "asv_id") %>% 
    
    rename(
      
      ASV = "asv_label.y", 
      
      `Domain or kingdom` = "Domain"
      
    ) %>% 
    
    mutate(ASV = ifelse(is.na(ASV), asv_label.x, ASV)) %>% 
    
    arrange(desc(v_degree), desc(v_betweenness)) %>% 
    
    rename(
      
      Degree = "v_degree", 
      
      Betweenness = "v_betweenness"
      
    ) %>% 
    
    select(-c(asv_label.x, Species, hub))
  
  return(tmp1)
  
}

#### Heatmap labels ####

heatmap_asv_labels <- function(x) {
  
  tmp1 <- tibble(ASV = x) %>%
    
    inner_join(., bc_hub_tax, by = "ASV") %>%
    
    pull(asv_fig_label)
  
  return(tmp1)
  
}

#### Plot heat maps ####

plot_heat_map <- function(fig.data, plot.title) {
  
  tmp1 <- ggplot() + 
    
    scale_x_discrete(
      
      drop = FALSE, 
      
      expand = expansion(mult = c(0.075, 0.075)), 
      
      labels = function(x) seasonality_labels2(x)
      
    ) + 
    
    scale_y_discrete(labels = function(x) heatmap_asv_labels(x)) +
    
    geom_tile(
      
      data = fig.data, 
      
      aes(x = seasonality, y = asv_label, fill = asv_z), 
      
      colour = "white"
        
    ) + 
    
    scale_fill_viridis(
      
      name = "Z score", 
      
      limits = range(bc_asv_z_score_range$min_z, bc_asv_z_score_range$max_z)
      
    ) + 
    
    labs(
      
      title = plot.title, 
      
      y = NULL, 
      
      x = NULL
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.y = element_text(colour = "black", size = 7), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 22.5, hjust = 1), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "right"
      
    )
  
}
```

```{r fig4}
# Hub taxonomy
bc_hub_tax <- map(
  
  bc_network_node_metadata, 
  
  .f = get_hub_summary
  
) %>% 
  
  bind_rows(.id = "Plant habitat") %>%
  
  format_tax_label(.) %>%
  
  select(ASV, tax_label) %>%
  
  mutate(
    
    tax_label = ifelse(
      
      tax_label == "Nitrososphaeraceae",
      
      "Uncl. Nitrososphaeraceae",
      
      tax_label
      
    ),
    
    tax_label = ifelse(
      
      tax_label == "Vicinamibacteraceae",
      
      "Uncl. Vicinamibacteraceae",
      
      tax_label
      
    ),
    
    asv_fig_label = paste(ASV, " (", tax_label, ")", sep = "")
    
  )

# Generate heat maps
bc_heatmap <- map2(
  
  bc_heatmap_input, 
  
  c("(a)", "(b)", "(c)"), 
  
  .f = plot_heat_map
  
)

# Get shared legend
fig4_legend <- get_legend(
  
  bc_heatmap$LE + 
    
    theme(legend.position = "right")
  
)

# Arrange figures in grid
fig4_grid <- plot_grid(
  
  bc_heatmap$LE + theme(legend.position = "none"), 
  
  bc_heatmap$RE + theme(legend.position = "none"), 
  
  bc_heatmap$RH + theme(legend.position = "none"), 
  
  align = 'v', 
  
  axis = "lr", 
  
  nrow = 3, 
  
  ncol = 1, 
  
  rel_heights = c(0.5, 0.25, 1)
  
)

# Construct figure
fig4 <- plot_grid(fig4_grid, fig4_legend, nrow = 1, rel_widths = c(1, 0.1))

# Save figure
ggsave2(
  
  filename = "data/results/fig4.png", 
  
  plot = fig4, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/fig4.pdf", 
  
  plot = fig4, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)
```

### Hub Z scores by species, Fig. S19  

```{r figS19_functions}
#### Calculate important ASV Z scores 2 ####

get_important_asv_z_hostspecies <- function(asv.table, metadata.table, important.asvs, community) {
  
  tmp1 <- asv.table %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    inner_join(metadata.table, ., by = "sample_unique_id") %>% 
    
    group_by(seasonality, host_species, asv_id) %>% 
    
    summarise(mean_n_seqs = mean(n_seqs)) %>% 
    
    ungroup() %>% 
    
    group_by(asv_id, host_species) %>%

    mutate(asv_z = z_score(mean_n_seqs)) %>%

    ungroup() %>%

    filter(asv_id %in% important.asvs$asv_id)
  
  if(community == "bc_16s") {
    
    tmp2 <- tmp1 %>% 
      
      inner_join(., bc_asv_labels$bc_16s, by = "asv_id")
    
  } else if(community == "bc_its") {
    
    tmp2 <- tmp1 %>% 
      
      inner_join(., bc_asv_labels$bc_its, by = "asv_id")
    
  }
  
  return(tmp2)
  
}

#### Combine ASV Z ####

combine_asv_z_hostspecies <- function(z.scores.16s, z.scores.its, plant.hab) {
  
  tmp1 <- bind_rows(z.scores.16s, z.scores.its) %>% 
    
    mutate(sample_type = plant.hab) %>%
    
    pivot_wider(
      
      id_cols = c(seasonality, asv_id),
      
      names_from = "host_species",
      
      values_from = "asv_z"
      
    )
  
  return(tmp1)
  
}
```

```{r figS19}
# Calculate important ASV z scores
bc_important_asv_z_hostspecies <- pmap(
  
  list(
    
    bc_asv_habsplit_df_SRS_output, 
    
    bc_metadata_habsplit_filtered, 
    
    c(bc_important_asvs, bc_important_asvs), 
    
    c("bc_16s", "bc_16s", "bc_16s", "bc_its", "bc_its", "bc_its")
    
  ), 
  
  .f = get_important_asv_z_hostspecies
  
)

# Combine ASV z scores
bc_important_asv_z_combined_hostspecies <- pmap(
  
  list(
    
    bc_important_asv_z_hostspecies[1:3], 
    
    bc_important_asv_z_hostspecies[4:6], 
    
    c("LE", "RE", "RH")
    
  ), 
  
  .f = combine_asv_z_hostspecies
  
) %>% 
  
  set_names(nm = c("LE", "RE", "RH")) %>%
  
  bind_rows(.id = "plant_habitat") %>%
  
  mutate(
   
    plant_habitat = map_chr(plant_habitat, .f = plant_hab),
    
    plant_habitat = factor(
      
      plant_habitat,
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Correlation
figS19_corr_input <- bc_important_asv_z_combined_hostspecies %>%
  
  select(deltoides, trichocarpa)

figS19_corr <- corr.test(figS19_corr_input, method = "pearson")

figS19_siglabel <- figS19_corr$r %>%
  
  as.data.frame(.) %>%
  
  rownames_to_column(var = "host_species") %>%
  
  as_tibble(.) %>%
  
  filter(host_species == "deltoides") %>%
  
  select(host_species, trichocarpa) %>%
  
  mutate(
    
    xpos = 1.5,
    
    ypos = -1.25,
    
    siglabel = paste("r==", round(trichocarpa, 2), "*','~", "italic(P)<0.001", sep = "")
    
  )

# Get figure
figS19 <- ggplot() +
  
  geom_abline(
    
    slope = 1,
    
    intercept = 0,
    
    linewidth = 0.5,
    
    linetype = 2
    
  ) +
  
  geom_point(
    
    data = bc_important_asv_z_combined_hostspecies,
    
    aes(x = trichocarpa, y = deltoides, fill = plant_habitat),
    
    pch = 21, 
    
    colour = "black", 
    
    size = 2,
    
    alpha = 0.6
    
  ) + 
  
  # geom_smooth(
  #   
  #   data = bc_important_asv_z_combined_hostspecies,
  #   
  #   aes(x = trichocarpa, y = deltoides, group = plant_habitat, colour = plant_habitat),
  #   
  #   method = "lm",
  #   
  #   formula = y ~ x,
  #   
  #   se = FALSE
  #   
  # ) +
  
  # geom_smooth(
  #   
  #   data = bc_important_asv_z_combined_hostspecies,
  #   
  #   aes(x = trichocarpa, y = deltoides),
  #   
  #   colour = "black",
  #   
  #   method = "lm",
  #   
  #   formula = y ~ x
  #   
  # ) + 
  
  scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      name = "Plant habitat",
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0
      
    ) + 
  
  geom_text(
    
    data = figS19_siglabel,
    
    aes(x = xpos, y = ypos, label = siglabel),
    
    parse = TRUE,
    
    size = 4
    
  ) +
    
    labs(
      
      title = NULL,
      
      x = bquote("Hub ASV Z score ("~italic(P.~trichocarpa)*")"), 
      
      y = bquote("Z score ("~italic(P.~deltoides)*")")
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )

# Save figure
ggsave2(
  
  filename = "data/results/figS19.png", 
  
  plot = figS19, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS19.pdf", 
  
  plot = figS19, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 6.5, 
  
  units = "in"
  
)
```

### Table S13  

```{r tableS13}
# Table S13
tableS13_input <- bc_important_asvs %>% 
  
  bind_rows(.id = "id") %>% 
  
  mutate(
    
    id = str_remove(id, "bc_"), 
    
    id = toupper(id)
    
    ) %>% 
  
  select(id, asv_id, v_degree, v_betweenness) %>% 
  
  inner_join(., bc_important_asv_tax, by = "asv_id") %>% 
  
  mutate(
    
    id = map_chr(id, .f = plant_hab), 
    
    id = factor(
      
      id, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  ) %>% 
  
  rename(
    
    `Plant habitat` = "id", 
    
    ASV = "asv_label", 
    
    `Domain or kingdom` = "Domain"
    
  ) %>% 
  
  arrange(`Plant habitat`, desc(v_degree), desc(v_betweenness)) %>% 
  
  mutate(
    
    Family = ifelse(is.na(Family), "unclassified", Family), 
    
    Genus = ifelse(is.na(Genus), "unclassified", Genus)
    
  ) %>%
  
  format_tax_label(.) %>%
  
  select(`Plant habitat`, ASV, `Domain or kingdom`, Phylum, Class, Order, tax_label, v_degree, v_betweenness) %>% 
  
  rename(
    
    Degree = "v_degree",
    
    Betweenness = "v_betweenness",
    
    Classification = "tax_label"
    
  )

# Save
write_tsv(
  
  tableS13_input,
  
  file = "data/results/tableS13.txt"
  
)
```

# Taxonomy summary and patterns  

## Figures S5-S7

```{r tax_summaries_and_figs_functions}
#### Tax summaries ####

get_tax_summary <- function(asv.table, community) {
  
  # Format ASV table
  tmp1 <- asv.table %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.)
  
  # Add taxonomy
  if(community == "Bacteria/Archaea") {
    
    tmp2 <- tmp1 %>% 
      
      inner_join(bc_tax_filtered$bc_16s_tax_filtered, ., by = "asv_id")
    
  } else if(community == "Fungi") {
    
    tmp2 <- tmp1 %>% 
      
      inner_join(bc_tax_filtered$bc_its_tax_filtered, ., by = "asv_id") %>% 
      
      rename(Domain = "Kingdom")
    
  }
  
  # Phylum level
  tmp3 <- tmp2 %>% 
    
    group_by(Domain, Phylum) %>% 
    
    summarise(tax_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    mutate(tax_perc = 100 * (tax_n_seqs / sum(tax_n_seqs))) %>% 
    
    mutate(Phylum = ifelse(tax_perc < 0.1, "Other", Phylum)) %>% 
    
    mutate(
      
      Phylum = ifelse(Domain == "Fungi" & is.na(Phylum), "Uncl. fungi", Phylum), 
      
      Phylum = ifelse(Domain == "Fungi" & Phylum == "unidentified", "Uncl. fungi", Phylum)
      
    ) %>% 
    
    mutate(
      
      Phylum = ifelse(Phylum == "Other" & Domain == "Bacteria", "Other bacteria", Phylum), 
      
      Phylum = ifelse(Phylum == "Other" & Domain == "Archaea", "Other archaea", Phylum), 
      
      Phylum = ifelse(Phylum == "Other" & Domain == "Fungi", "Other fungi", Phylum)
      
    ) %>% 
    
    group_by(Domain, Phylum) %>% 
    
    summarise(`% of seqs.` = sum(tax_perc)) %>% 
    
    ungroup() %>% 
    
    arrange(desc(`% of seqs.`)) %>% 
    
    mutate(
      
      tax_perc_label = format(round(`% of seqs.`, 1), nsmall = 1), 
      
      tax_label = factor(Phylum, levels = Phylum)
      
    )
  
  if(community == "Bacteria/Archaea") {

    tmp4 <- tmp3 %>%

      mutate(Domain = factor(Domain, levels = c("Bacteria", "Archaea")))

  } else if(community == "Fungi") {

    tmp4 <- tmp3

  }

  # Class level
  tmp5 <- tmp2 %>%

    group_by(Domain, Phylum, Class) %>%

    summarise(tax_n_seqs = sum(n_seqs)) %>%

    ungroup() %>%

    mutate(tax_perc = 100 * (tax_n_seqs / sum(tax_n_seqs))) %>%

    mutate(Class = ifelse(tax_perc < 0.1, "Other", Class)) %>%

    mutate(Phylum = ifelse(tax_perc < 0.1, "Other", Phylum)) %>%

    mutate(

      Phylum = ifelse(Domain == "Fungi" & is.na(Phylum), "Uncl. fungi", Phylum),

      Phylum = ifelse(Domain == "Fungi" & Phylum == "unidentified" , "Uncl. fungi", Phylum)

    ) %>%

    mutate(

      Class = ifelse(Domain == "Fungi" & Phylum == "Uncl. fungi" & is.na(Class), "Uncl. fungi", Class),

      Class = ifelse(Domain == "Fungi" & Phylum == "Uncl. fungi" & Class == "unidentified", "Uncl. fungi", Class),

      Class = ifelse(Domain == "Fungi" & Phylum != "Uncl. fungi" & is.na(Class), paste("Uncl.", Phylum, sep = " "), Class),

      Class = ifelse(Domain == "Fungi" & Phylum != "Uncl. fungi" & Class == "unidentified", paste("Uncl.", Phylum, sep = " "), Class), 
      
      Class = ifelse(Class == "uncultured", paste("Uncl.", Phylum, sep = " "), Class)

    ) %>%

    mutate(

      Class = ifelse(Class == "Other" & Domain == "Bacteria", "Other bacteria", Class),

      Class = ifelse(Class == "Other" & Domain == "Archaea", "Other archaea", Class),

      Class = ifelse(Class == "Other" & Domain == "Fungi", "Other fungi", Class),

      Class = ifelse(Class == "Rozellomycotina_cls_Incertae_sedis", "Uncl. Rozellomycotina", Class)

    ) %>%

    group_by(Domain, Phylum, Class) %>%

    summarise(`% of seqs.` = sum(tax_perc)) %>%

    ungroup() %>%

    arrange(desc(`% of seqs.`)) %>%

    mutate(

      tax_perc_label = format(round(`% of seqs.`, 1), nsmall = 1)

    ) %>%

    mutate(tax_label = factor(Class, levels = Class))

  if(community == "Bacteria/Archaea") {

    tmp6 <- tmp5 %>%

      mutate(Domain = factor(Domain, levels = c("Bacteria", "Archaea")))

  } else if(community == "Fungi") {

    tmp6 <- tmp5

  }

  # Combine into a list
  tmp7 <- list(tmp4, tmp6)

  if(community == "Bacteria/Archaea") {

    tmp8 <- tmp7 %>%

      set_names(nm = c("bc_16s_phylum", "bc_16s_class"))

  } else if(community == "Fungi") {

    tmp8 <- tmp7 %>%

      set_names(nm = c("bc_its_phylum", "bc_its_class"))

  }

  return(tmp8)
  
}

#### Tax summary figures ####

plot_tax_summary <- function(tax.data, community, plot.colors, plot.title) {
  
  tmp1 <- ggplot() + 
    
    scale_y_continuous(
      
      limits = c(0, 105), 
      
      expand = expansion(mult = c(0.05, 0.05)), 
      
      breaks = c(0, 25, 50, 75, 100)
      
    ) + 
    
    geom_point(
      
      data = tax.data, 
      
      aes(x = tax_label, y = `% of seqs.`, shape = Domain), 
      
      colour = plot.colors, 
      
      fill = plot.colors, 
      
      size = 2.5, 
      
      alpha = 0.75
      
    ) + 
    
    geom_text(
      
      data = tax.data, 
      
      aes(x = tax_label, y = `% of seqs.` + 12.5, label = tax_perc_label), 
      
      size = 2, 
      
      angle = 45
      
    ) + 
    
    labs(
      
      title = plot.title, 
      
      x = NULL, 
      
      y = "% of sequences"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0.5), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 7, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )
  
  if(community == "Bacteria/Archaea") {
    
    tmp2 <- tmp1 + 
      
      scale_shape_manual(
        
        name = "Domain", 
        
        values = c(21, 23)
        
      )
    
  } else if(community == "Fungi") {
    
    tmp2 <- tmp1 + 
      
      scale_shape_manual(
        
        values = 22, 
        
        guide = "none"
        
      )
    
  }
  
  return(tmp2)
  
}
```

```{r tax_summaries_and_figs}
# Get 16S taxonomy summaries
bc_tax_summary <- map2(
  
  bc_asv_habsplit_df_SRS_output, 
  
  c(rep("Bacteria/Archaea", 3), rep("Fungi", 3)), 
  
  .f = get_tax_summary
  
)

# Generate phylum figures
bc_tax_phylum_summary_figs <- pmap(
  
  list(
    
    list(
      
      bc_tax_summary$bc_16s_le_asv_df$bc_16s_phylum, 
      
      bc_tax_summary$bc_16s_re_asv_df$bc_16s_phylum, 
      
      bc_tax_summary$bc_16s_rh_asv_df$bc_16s_phylum, 
      
      bc_tax_summary$bc_its_le_asv_df$bc_its_phylum, 
      
      bc_tax_summary$bc_its_re_asv_df$bc_its_phylum, 
      
      bc_tax_summary$bc_its_rh_asv_df$bc_its_phylum
      
    ), 
    
    c("Bacteria/Archaea", "Bacteria/Archaea", "Bacteria/Archaea", "Fungi", "Fungi", "Fungi"), 
    
    c("#6DCD59FF", "#31688EFF", "#440154FF", "#6DCD59FF", "#31688EFF", "#440154FF"), 
    
    c("Leaf endosphere", "Root endosphere", "Rhizosphere", "Leaf endosphere", "Root endosphere", "Rhizosphere")
    
  ), 
  
  .f = plot_tax_summary
  
) %>% 
  
  set_names(nm = names(bc_tax_summary))

# Generate class figures
bc_tax_class_summary_figs <- pmap(
  
  list(
    
    list(
      
      bc_tax_summary$bc_16s_le_asv_df$bc_16s_class, 
      
      bc_tax_summary$bc_16s_re_asv_df$bc_16s_class, 
      
      bc_tax_summary$bc_16s_rh_asv_df$bc_16s_class, 
      
      bc_tax_summary$bc_its_le_asv_df$bc_its_class, 
      
      bc_tax_summary$bc_its_re_asv_df$bc_its_class, 
      
      bc_tax_summary$bc_its_rh_asv_df$bc_its_class
      
    ), 
    
    c("Bacteria/Archaea", "Bacteria/Archaea", "Bacteria/Archaea", "Fungi", "Fungi", "Fungi"), 
    
    c("#6DCD59FF", "#31688EFF", "#440154FF", "#6DCD59FF", "#31688EFF", "#440154FF"), 
    
    c("Leaf endosphere", "Root endosphere", "Rhizosphere", "Leaf endosphere", "Root endosphere", "Rhizosphere")
    
  ), 
  
  .f = plot_tax_summary
  
) %>% 
  
  set_names(nm = names(bc_tax_summary))
```

### Fig. S5  

```{r figS5}
# Get shared legend
figS5_legend <- get_legend(
  
  bc_tax_class_summary_figs$bc_16s_le_asv_df + 
    
    theme(legend.position = "bottom") + 
    
    guides(shape = guide_legend(override.aes = list(colour = "black", fill = "black")))
  
)

# Arrange figures in grid
figS5_grid <- plot_grid(
  
  bc_tax_class_summary_figs$bc_16s_le_asv_df + theme(legend.position = "none"), 
  
  bc_tax_class_summary_figs$bc_16s_re_asv_df + theme(legend.position = "none"), 
  
  bc_tax_class_summary_figs$bc_16s_rh_asv_df + theme(legend.position = "none"), 
  
  align = 'vh', 
  
  nrow = 3, 
  
  ncol = 1
  
)

# Construct figure
figS5 <- plot_grid(figS5_grid, figS5_legend, ncol = 1, rel_heights = c(1, 0.067))

# Save figure
ggsave2(
  
  filename = "data/results/figS5.png", 
  
  plot = figS5, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS5.pdf", 
  
  plot = figS5, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)
```

### Fig. S6  

```{r figS6}
# Arrange figures in grid
figS6 <- plot_grid(
  
  bc_tax_class_summary_figs$bc_its_le_asv_df + theme(legend.position = "none"), 
  
  bc_tax_class_summary_figs$bc_its_re_asv_df + theme(legend.position = "none"), 
  
  bc_tax_class_summary_figs$bc_its_rh_asv_df + theme(legend.position = "none"), 
  
  align = 'vh', 
  
  nrow = 3, 
  
  ncol = 1
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS6.png", 
  
  plot = figS6, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS6.pdf", 
  
  plot = figS6, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)
```

# Functional assignments for fungi  

## Functional guild relative abundance  

```{r assign_funguild_to_asv_functions}
#### Format FUNGuild ####

format_funguild_db <- function(x) {
  
  tmp1 <- x %>% 
    
    rename(db_entry = "X1") %>% 
    
    separate(
      
      db_entry, 
      
      into = c("taxon", "guid", "mb_number", "taxonomic_level", "trophic_mode", "guild", "confidence_ranking", "growth_form", "trait", "notes", "citation_source"), 
      
      sep = ","
      
    ) %>% 
    
    mutate(db_entry = paste("db_entry", c(1 : n()), sep = "_")) %>% 
    
    pivot_longer(
      
      -db_entry, 
      
      names_to = "db_name", 
      
      values_to = "db_value"
      
    ) %>% 
    
    mutate(db_value = str_replace_all(db_value, ".*:", "")) %>% 
    
    pivot_wider(
      
      id_cols = "db_entry", 
      
      names_from = "db_name", 
      
      values_from = "db_value"
      
    ) %>% 
    
    mutate(taxonomic_level = as.numeric(taxonomic_level))
  
  return(tmp1)
  
}

#### Assign funguild classifications to ASVs ####

assign_funguild_to_asv <- function(tax.tab, funguild.db) {
  
  tmp1 <- funguild.db %>% 
    
    filter(taxonomic_level == 13) %>% 
    
    rename(Genus = "taxon")
  
  tmp2 <- funguild.db %>% 
    
    filter(taxonomic_level == 9) %>% 
    
    rename(Family = "taxon")
  
  tmp3 <- funguild.db %>% 
    
    filter(taxonomic_level == 7) %>% 
    
    rename(Order = "taxon")
  
  tmp4 <- funguild.db %>% 
    
    filter(taxonomic_level == 3) %>% 
    
    rename(Phylum = "taxon")
  
  tmp5 <- tax.tab %>% 
    
    filter(!is.na(Genus) & Genus != "unidentified") %>% 
    
    left_join(., tmp1, by = "Genus")
  
  tmp6 <- tax.tab %>% 
    
    filter(is.na(Genus) | Genus == "unidentified") %>% 
    
    filter(!is.na(Family) & Family != "unidentified") %>% 
    
    left_join(., tmp2, by = "Family")
  
  tmp7 <- tax.tab %>% 
    
    filter(is.na(Genus) | Genus == "unidentified") %>% 
    
    filter(is.na(Family) | Family == "unidentified") %>% 
    
    filter(!is.na(Order) & Order != "unidentified") %>% 
    
    left_join(., tmp3, by = "Order")
  
  tmp8 <- bind_rows(tmp5, tmp6, tmp7)
  
  tmp9 <- tmp8 %>% 
    
    pull(asv_id)
  
  tmp10 <- tax.tab %>% 
    
    filter(!asv_id %in% tmp9) %>% 
    
    left_join(., tmp4, by = "Phylum") %>% 
    
    bind_rows(tmp8, .)
  
  return(tmp10)
  
}

#### Combine funguild assignments with ASV tables ####

combine_funguild_and_asv_tab <- function(asv.tab, funguild) {
  
  tmp1 <- asv.tab %>% 
    
    as_tibble(rownames = NA) %>% 
    
    rownames_to_column(var = "sample_unique_id") %>% 
    
    asv_pivot_longer2(.) %>% 
    
    inner_join(., funguild, by = "asv_id")
  
  return(tmp1)
  
}

#### Split seq counts among multi-guild assigments ###

split_guild_n_seqs <- function(x) {
  
  tmp1 <- lengths(regmatches(x$guild, gregexpr("\\-", x$guild)))
  
  tmp2 <- paste("guild", c(1 : (tmp1 + 1)), sep = "_")
  
  tmp3 <- x %>% 
    
    separate(guild, into = tmp2, sep = "-") %>% 
    
    mutate(guild_n_seqs_split = guild_n_seqs / (tmp1 + 1)) %>% 
    
    select(-guild_n_seqs) %>% 
    
    pivot_longer(
      
      -c(sample_unique_id, guild_n_seqs_split), 
      
      names_to = "guild_temp", 
      
      values_to = "guild"
      
    ) %>% 
    
    select(sample_unique_id, guild, guild_n_seqs_split)
  
  return(tmp3)
  
}

#### Calculate guild sequence counts ####

get_guild_n_seqs <- function(x, metadata.tab) {
  
  tmp1 <- x %>% 
    
    # mutate(guild = ifelse(is.na(guild) | confidence_ranking == "Possible" | guild == "NULL", "Unclassified", guild)) %>% 
    
    mutate(guild = ifelse(is.na(guild) | guild == "NULL", "Unclassified", guild)) %>% 
    
    group_by(sample_unique_id, guild) %>% 
    
    summarise(guild_n_seqs = sum(n_seqs)) %>% 
    
    ungroup() %>% 
    
    group_by(sample_unique_id, guild) %>% 
    
    group_split() %>% 
    
    map(., .f = ungroup)
  
  tmp2 <- map(tmp1, .f = split_guild_n_seqs) %>% 
    
    bind_rows(.)
  
  tmp3 <- tmp2 %>% 
    
    group_by(sample_unique_id, guild) %>% 
    
    summarise(guild_n_seqs = sum(guild_n_seqs_split)) %>% 
    
    ungroup() %>% 
    
    inner_join(metadata.tab, ., by = "sample_unique_id")
  
  return(tmp3)
  
}
```

```{r assign_funguild_to_asv}
# Format funguild database
funguild_db <- read_tsv(

  file = "data/funguild_db_2.txt",

  col_names = FALSE

) %>%

  format_funguild_db(.)

# Assign ASVs based on funguild
bc_its_funguild_assignments <- assign_funguild_to_asv(

  bc_tax_filtered$bc_its_tax_filtered,

  funguild_db

)

# Prep assignments
bc_its_funguild_assignments_list <- vector("list", 1)
bc_its_funguild_assignments_list[[1]] <- bc_its_funguild_assignments

# Add funguild assignments to ASV tables
bc_its_funguild_asv <- map2(

  bc_asv_habsplit_df_SRS_output[4:6], bc_its_funguild_assignments_list,

  .f = combine_funguild_and_asv_tab

) %>%

  set_names(nm = c("bc_its_le", "bc_its_re", "bc_its_rh"))

# Get functional guild relative abundances
bc_its_funguild <- map2(

  bc_its_funguild_asv,

  bc_metadata_habsplit_filtered[4:6],

  .f = get_guild_n_seqs

)

# Save
save(

  funguild_db,

  bc_its_funguild_assignments,

  bc_its_funguild_assignments_list,

  bc_its_funguild_asv,

  bc_its_funguild,

  file = "data/working/funguild.RData"

)

# Load funguild data
load("data/working/funguild.RData")
```

## Fungal guild summary  

### Get summary  

```{r guild_summary_functions}
#### Funguild summary ####

get_funguild_summary <- function(x) {
  
  tmp1 <- x %>% 
    
    bind_rows(.) %>% 
    
    group_by(plant_habitat, guild) %>% 
    
    summarise(total_n_seqs = sum(guild_n_seqs)) %>% 
    
    ungroup() %>% 
    
    mutate(guild_perc = 100 * (total_n_seqs / sum(total_n_seqs))) %>% 
    
    mutate(guild = ifelse(guild_perc < 0.1, "Other", guild)) %>% 
    
    group_by(plant_habitat, guild) %>% 
    
    summarise(guild_perc = sum(guild_perc)) %>% 
    
    ungroup() %>% 
    
    arrange(desc(guild_perc)) %>% 
    
    mutate(
      
      guild_perc_label = format(round(guild_perc, 1), nsmall = 1), 
      
      guild = factor(guild, levels = guild)
      
    )
  
  return(tmp1)
  
}

#### Plot funguild summary ####

plot_funguild_summary <- function(funguild.data, plot.colors, plot.title) {
  
  tmp1 <- ggplot() + 
    
    scale_y_continuous(
      
      limits = c(0, 105), 
      
      expand = expansion(mult = c(0.05, 0.05)), 
      
      breaks = c(0, 25, 50, 75, 100)
      
    ) + 
    
    geom_point(
      
      data = funguild.data, 
      
      aes(x = guild, y = guild_perc), 
      
      colour = plot.colors, 
      
      fill = plot.colors, 
      
      pch = 22, 
      
      size = 2.5, 
      
      alpha = 0.75
      
    ) + 
    
    geom_text(
      
      data = funguild.data, 
      
      aes(x = guild, y = guild_perc + 12.5, label = guild_perc_label), 
      
      size = 2, 
      
      angle = 45
      
    ) + 
    
    labs(
      
      title = plot.title, 
      
      x = NULL, 
      
      y = "% of sequences"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0.5), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 7, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    )
  
  return(tmp1)
  
}
```

```{r guild_summary}
# Get summary data
bc_its_funguild_summary <- map(
  
  bc_its_funguild, 
  
  .f = get_funguild_summary
  
)

# Get funguild summary figures
bc_its_funguild_summary_figs <- pmap(
  
  list(
    
    bc_its_funguild_summary, 
    
    c("#6DCD59FF", "#31688EFF", "#440154FF"), 
    
    c("Leaf endosphere", "Root endosphere", "Rhizosphere")
    
  ), 
  
  .f = plot_funguild_summary
  
)
```

### Fig. S7  

```{r figS7}
# Arrange figures in grid
figS7 <- plot_grid(
  
  bc_its_funguild_summary_figs$bc_its_le, 
  
  bc_its_funguild_summary_figs$bc_its_re, 
  
  bc_its_funguild_summary_figs$bc_its_rh, 
  
  align = 'vh', 
  
  nrow = 3, 
  
  ncol = 1
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS7.png", 
  
  plot = figS7, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS7.pdf", 
  
  plot = figS7, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 8, 
  
  units = "in"
  
)
```

## Seasonal patterns in fungal guilds  

```{r fungal_guild_seasonal_patterns_anova_functions}
#### Funguild relative abundances ####

format_funguild_relabund <- function(abund.tab, sample.seq.min, metadata.tab) {
  
  tmp1 <- abund.tab %>%
    
    mutate(guild2 = ifelse(grepl("Saprotroph", guild), "Saprotroph", guild)) %>%
    
    group_by(sample_unique_id, guild2) %>%
    
    summarise(guild_n_seqs = sum(guild_n_seqs)) %>%
    
    ungroup() %>%
    
    mutate(guild_relabund = guild_n_seqs / sample.seq.min) %>%
    
    filter(guild2 == "Arbuscular Mycorrhizal" | guild2 == "Ectomycorrhizal" | guild2 == "Endophyte" | guild2 == "Plant Pathogen" | guild2 == "Saprotroph" | guild2 == "Unclassified") %>%
    
    inner_join(metadata.tab, ., by = "sample_unique_id") %>% 
    
    mutate(
      
      guild2 = ifelse(guild2 == "Arbuscular Mycorrhizal", "AMF", guild2),
      
      guild2 = ifelse(guild2 == "Ectomycorrhizal", "ECM", guild2),
      
      guild_label = factor(
        
        guild2,
        
        levels = c("AMF", "ECM", "Endophyte", "Plant Pathogen", "Saprotroph", "Unclassified"))
      
    )
  
  return(tmp1)
  
}

#### Funguild anova ####

funguild_aov <- function(x) {
  
  tmp1 <- aov(guild_relabund ~ year * month * host_species, data = x)
  
  return(tmp1)
  
}

#### Funguild list anova ####

get_funguild_aov_list <- function(x) {
  
  tmp1 <- x %>% 
    
    group_by(guild_label) %>% 
    
    group_split() %>% 
    
    set_names(nm = c("Arbuscular Mycorrhizal", "Ectomycorrhizal", "Endophyte", "Plant Pathogen", "Saprotroph", "Unclassified"))
  
  tmp2 <- map(tmp1, .f = funguild_aov)
  
  return(tmp2)
  
}

#### Funguild list anova summary ####

get_funguild_aov_summary_list <- function(x) {
  
  tmp1 <- x %>% 
    
    map(., .f = summary.aov)
  
  return(tmp1)
  
}

#### Get funguild tukey hsd ####

get_funguild_tukeyHSD <- function(x) {
  
  tmp1 <- map(x, .f = TukeyHSD)
  
  return(tmp1)
  
}

#### Get seasonality P values for funguild ####

get_funguild_seasonality_aov_Pval <- function(x) {
  
  tmp1 <- map(x, .f = get_seasonality_aov_Pval)
  
  tmp2 <- map_dfr(tmp1, .f = bind_rows) %>% 
    
    mutate(guild_label = names(tmp1)) %>% 
    
    mutate(
      
      guild_label = ifelse(guild_label == "Arbuscular Mycorrhizal", "AMF", guild_label),
      
      guild_label = ifelse(guild_label == "Ectomycorrhizal", "ECM", guild_label),
      
      guild_label = factor(
        
        guild_label,
        
        levels = c("AMF", "ECM", "Endophyte", "Plant Pathogen", "Saprotroph", "Unclassified"))
      
    ) %>% 
    
    mutate(pval_label = NA) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` < 0.001, "season%*%year*','~italic(P) < 0.001", pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` <= 0.05 & `Pr(>F)` >= 0.001, paste("season%*%year*','~italic(P) == ", round(`Pr(>F)`, 3), sep = ""), pval_label)) %>% 
    
    mutate(pval_label = ifelse(`Pr(>F)` > 0.05, "season%*%year*','~italic(n.s.)", pval_label))
  
  return(tmp2)
  
}

#### Funguild cld inner ####

funguild_cld_inner <- function(x, y) {
  
  tmp1 <- x$`year:month`$Letters
  
  tmp2 <- tibble(
    
    sample_type = names(tmp1), 
    
    cld = str_remove_all(tmp1, " "), 
    
    guild_label = y
    
  ) %>% 
    
    select(guild_label, everything())
  
  return(tmp2)
  
}

#### Funguild CLD ####

funguild_cld <- function(anova.model, tukeyhsd.model) {
  
  tmp1 <- map2(
    
    anova.model, tukeyhsd.model, 
    
    .f = multcompView::multcompLetters4
    
  )
  
  tmp2 <- names(anova.model)
  
  tmp3 <- map2(tmp1, tmp2, .f = funguild_cld_inner) %>% 
    
    bind_rows() %>% 
    
    rename(timepoint = "sample_type") %>% 
    
    mutate(
      
      guild_label = ifelse(guild_label == "Arbuscular Mycorrhizal", "AMF", guild_label),
      
      guild_label = ifelse(guild_label == "Ectomycorrhizal", "ECM", guild_label),
      
      guild_label = factor(
        
        guild_label,
        
        levels = c("AMF", "ECM", "Endophyte", "Plant Pathogen", "Saprotroph", "Unclassified"))
      
    ) %>% 
    
    separate(timepoint, into = c("year", "month"), sep = ":") %>%
    
    unite(seasonality, month, year, sep = " ") %>%
    
    mutate(seasonality = factor(seasonality, levels = c("Feb 2018", "June 2018", "Aug 2018", "Oct 2018",
                                                        "Feb 2019", "June 2019", "Aug 2019", "Oct 2019")))
  
  return(tmp3)
  
}
```

```{r fungal_guild_seasonal_patterns_anova}
# Format funguild relabund data for ANOVA and figures
bc_its_funguild_relabund <- pmap(
  
  list(bc_its_funguild, bc_sample_seq_min[4:6], bc_metadata_habsplit_filtered[4:6]), 
  
  .f = format_funguild_relabund
  
)

# Run ANOVA
bc_funguild_aov <- map(bc_its_funguild_relabund, .f = get_funguild_aov_list)

# Get ANOVA summaries
bc_funguild_aov_summary <- map(bc_funguild_aov, .f = get_funguild_aov_summary_list)

# Run Tukey's HSD
bc_funguild_seasonality_tukeyHSD <- map(bc_funguild_aov, .f = get_funguild_tukeyHSD)

# Get month x year interaction P values
bc_funguild_seasonality_Pval <- map(bc_funguild_aov_summary, .f = get_funguild_seasonality_aov_Pval)

# Get hill diversity CLD for plots
bc_funguild_seasonality_cld <- map2(
  
  bc_funguild_aov, 
  
  bc_funguild_seasonality_tukeyHSD, 
  
  .f = funguild_cld
  
)
```

### Table S14  

```{r tableS14_functions}
#### Get funguild anova tables inner ####

get_funguild_aov_tables_inner <- function(aov.summary) {
  
  tmp1 <- aov.summary %>% 
    
    map(., .f = `[[`, 1) %>% 
    
    map(., .f = as_tibble, rownames = NA) %>% 
    
    map(., .f = rownames_to_column, var = "Variable") %>% 
    
    bind_rows(.id = "guild_label") %>% 
    
    mutate(Variable = str_squish(Variable))
  
  return(tmp1)
  
}

#### Get funguild anova tables ####

get_funguild_aov_tables <- function(anova.summary) {
  
  tmp1 <- map(anova.summary, .f = get_funguild_aov_tables_inner) %>% 
    
    bind_rows(.id = "Plant habitat") %>% 
    
    mutate(
      
      `Plant habitat` = str_remove(`Plant habitat`, "bc_16s_"), 
      
      `Plant habitat` = str_remove(`Plant habitat`, "bc_its_"), 
      
      `Plant habitat` = toupper(`Plant habitat`)
      
    )
  
  tmp2 <- unique(tmp1$`Plant habitat`)
  
  tmp3 <- tmp1 %>% 
    
    rename(
      
      P = "Pr(>F)", 
      
      `Sum of sqs.` = "Sum Sq"
      
    ) %>% 
    
    mutate(
      
      `Sum of sqs.` = format(round(`Sum of sqs.`, 3), nsmall = 3), 
      
      `F value` = format(round(`F value`, 2), nsmall = 2), 
      
      P = ifelse(P >= 0.001, format(round(P, 3), nsmall = 3, scientific = FALSE), P), 
      
      P = ifelse(as.numeric(P) < 0.001, "< 0.001", P)
      
    ) %>% 
    
    mutate(across(.cols = everything(), .fns = str_trim)) %>% 
    
    mutate(
      
      `F value` = ifelse(`F value` == "NA", "", `F value`), 
      
      `F value` = ifelse(is.na(`F value`), "", `F value`), 
      
      P = ifelse(P == "NA", "", P), 
      
      P = ifelse(is.na(P), "", P)
      
    ) %>% 
    
    mutate(
      
      Variable = str_to_title(Variable), 
      
      Variable = str_replace_all(Variable, "\\:", " x "), 
      
      Variable = str_replace_all(Variable, "\\_", " ")
      
    ) %>%
    
    mutate(
      
      guild_label = ifelse(guild_label == "Arbuscular Mycorrhizal", "AMF", guild_label),
      
      guild_label = ifelse(guild_label == "Ectomycorrhizal", "ECM", guild_label),
      
      guild_label = factor(
        
        guild_label,
        
        levels = c("AMF", "ECM", "Endophyte", "Plant Pathogen", "Saprotroph", "Unclassified"))
      
    ) %>%
    
    rename(Guild = "guild_label") %>% 
    
    select(-`Mean Sq`) %>% 
    
    group_by(`Plant habitat`) %>% 
    
    group_split() %>% 
    
    set_names(nm = tmp2) %>% 
    
    map(., .f = ungroup) %>% 
    
    map(., select, -`Plant habitat`)
  
  return(tmp3)
  
}
```

```{r tableS14}
# Table S14
tableS14_input <- get_funguild_aov_tables(bc_funguild_aov_summary) %>%
  
  bind_rows(.id = "Plant habitat") %>%
  
  mutate(`Plant habitat` = map_chr(`Plant habitat`, .f = plant_hab))

# Save
write_tsv(
  
  tableS14_input,
  
  file = "data/results/tableS14.txt"
  
)
```

## Funguild, Fig. 5  

```{r fig5_functions}
#### Plot funguild ####

plot_funguild2 <- function(guild.data, cld.data, pvals) {
  
  tmp0 <- guild.data %>% 
    
    group_by(plant_habitat, guild_label, year, seasonality) %>% 
    
    summarise(
      
      guild_mean = mean(guild_relabund), 
      
      guild_se = se(guild_relabund)
      
    ) %>% 
    
    ungroup()
  
  tmp1 <- ggplot() + 
    
    scale_y_continuous(
      
      expand = expansion(mult = c(0.25, 0.25))
      
    ) + 
    
    scale_x_discrete(labels = function(x) seasonality_labels2(x)) +
    
    geom_point(
      
      data = guild.data, 
      
      aes(x = seasonality, y = guild_relabund, colour = plant_habitat, fill = plant_habitat), 
      
      position = position_jitter(width = 0.2, seed = 12345), 
      
      pch = 21, 
      
      alpha = 0.5, 
      
      size = 1
      
    ) + 
    
    geom_boxplot(
      
      data = guild.data, 
      
      aes(x = seasonality, y = guild_relabund, colour = plant_habitat), 
      
      fill = "white",
      
      outlier.shape = NA,
      
      alpha = 0.4, 
      
      show.legend = FALSE
      
    ) + 
    
    geom_line(
      
      data = tmp0, 
      
      aes(x = seasonality, y = guild_mean, group = year, colour = plant_habitat), 
      
      show.legend = FALSE
      
    ) + 
    
    # stat_summary(
    #   
    #   data = guild.data, 
    #   
    #   aes(x = seasonality, y = guild_relabund), 
    #   
    #   fun.data = mean_cl_normal,
    #   
    #   geom = "errorbar",
    #   
    #   colour = "black",
    #   
    #   linewidth = 0.25, 
    #   
    #   width = 0
    #   
    # ) +
    
    stat_summary(
      
      data = guild.data, 
      
      aes(x = seasonality, y = guild_relabund, fill = plant_habitat), 
      
      fun = mean, 
      
      colour = "black",
      
      geom = "point", 
      
      pch = 21,
      
      size = 2
      
    ) + 
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    geom_text(
      
      data = cld.data,
      
      aes(x = seasonality, y = Inf, vjust = 1.55, label = cld),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    geom_text(
      
      data = pvals,
      
      aes(x = 4.5, y = -Inf, vjust = -0.25, label = pval_label),
      
      parse = TRUE,
      
      size = 3
      
    ) + 
    
    facet_grid(
      
      cols = vars(plant_habitat), 
      
      rows = vars(guild_label), 
      
      scales = "free_y"
      
    ) + 
    
    labs(
      
      title = NULL, 
      
      x = NULL, 
      
      y = "Guild relative abundance"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10)
      
    )
  
  return(tmp1)
  
}
```

```{r fig5}
# Input
bc_its_funguild_relabund_input <- bc_its_funguild_relabund %>% 
  
  bind_rows() %>% 
  
  mutate(
    
    plant_habitat = map_chr(sample_type, .f = plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Input
bc_funguild_seasonality_cld_input <- bc_funguild_seasonality_cld %>% 
  
  bind_rows(.id = "plant_habitat") %>% 
  
  mutate(
    
    plant_habitat = str_remove(plant_habitat, "bc_its_"), 
    
    plant_habitat = toupper(plant_habitat), 
    
    plant_habitat = map_chr(plant_habitat, .f = plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Input
bc_funguild_seasonality_Pval_input <- bc_funguild_seasonality_Pval %>% 
  
  bind_rows(.id = "plant_habitat") %>% 
  
  mutate(
    
    plant_habitat = str_remove(plant_habitat, "bc_its_"), 
    
    plant_habitat = toupper(plant_habitat), 
    
    plant_habitat = map_chr(plant_habitat, .f = plant_hab), 
    
    plant_habitat = factor(
      
      plant_habitat, 
      
      levels = c("Leaf endosphere", "Root endosphere", "Rhizosphere")
      
    )
    
  )

# Fig 5
fig5 <- plot_funguild2(
  
  bc_its_funguild_relabund_input, 
  
  bc_funguild_seasonality_cld_input, 
  
  bc_funguild_seasonality_Pval_input
  
)

# Save figure
ggsave2(
  
  filename = "data/results/fig5.png", 
  
  plot = fig5, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 8.5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/fig5.pdf", 
  
  plot = fig5, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 8.5, 
  
  units = "in"
  
)
```

### Guilds by host species, Fig. S20  

```{r figS20_functions}
funguild_SI_fig <- function(guild.relabund) {
  
  tmp0 <- guild.relabund %>%

    group_by(plant_habitat, guild_label, host_species, year, seasonality) %>%

    summarise(

      guild_relabund_mean = mean(guild_relabund)

    ) %>%

    ungroup()
  
  tmp1 <- ggplot() +
    
    scale_x_discrete(
      
      labels = function(x) seasonality_labels2(x)
      
    ) +
    
    geom_point(
      
      data = guild.relabund, 
      
      aes(
        
        x = seasonality, 
        
        y = guild_relabund, 
        
        colour = plant_habitat,
        
        fill = plant_habitat, 
        
        group = host_species, 
        
        shape = host_species
        
      ), 
      
      position = position_jitterdodge(
        
        jitter.width = 0.2, 
        
        dodge.width = 0.5, 
        
        seed = 12345
        
      ),
      
      alpha = 0.2, 
      
      size = 1
      
    ) +
    
    geom_line(
      
      data = tmp0, 
      
      aes(
        
        x = seasonality, 
        
        y = guild_relabund_mean, 
        
        group = interaction(guild_label, host_species, year), 
        
        colour = plant_habitat
        
      ), 
      
      linetype = 1,
      
      position = position_dodge(width = 0.5), 
      
      show.legend = FALSE
      
    ) + 
    
    stat_summary(
      
      data = guild.relabund, 
      
      aes(
        
        x = seasonality, 
        
        y = guild_relabund, 
        
        # colour = plant_habitat, 
        
        group = host_species
        
      ), 
      
      position = position_dodge(width = 0.5), 
      
      fun.data = mean_cl_normal,
      
      geom = "errorbar", 
      
      colour = "black",
      
      linewidth = 0.5, 
      
      width = 0.2
      
    ) +
    
    stat_summary(
      
      data = guild.relabund,
      
      aes(
        
        x = seasonality,
        
        y = guild_relabund,
        
        fill = plant_habitat,
        
        #colour = plant_habitat,
        
        group = host_species, 
        
        shape = host_species
        
      ),
      
      colour = "black",
      
      position = position_dodge(width = 0.5), 
      
      fun = mean,
      
      geom = "point",
      
      size = 1.5
      
    ) +
    
    scale_shape_manual(
      
      name = "Host species",
      
      values = c(21, 24), 
      
      labels = function(x) parse(text = x)
      
    ) +
    
    scale_colour_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) + 
    
    scale_fill_viridis(
      
      discrete = TRUE, 
      
      begin = 0.8, 
      
      end = 0, 
      
      guide = "none"
      
    ) +
    
    facet_wrap(
      
      ~ plant_habitat + guild_label, 
      
      scales = "free_y",
      
      nrow = 1
      
    ) + 
    
    labs(
      
      title = NULL, 
      
      x = NULL, 
      
      y = "Guild relative abundance"
      
    ) + 
    
    theme(
      
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), 
      
      panel.background = element_blank(), 
      
      panel.grid = element_blank(), 
      
      strip.background = element_blank(), 
      
      plot.title = element_text(colour = "black", size = 12, hjust = 0, face = "bold"), 
      
      axis.title = element_text(colour = "black", size = 10), 
      
      axis.ticks = element_line(colour = "black", linewidth = 0.25), 
      
      axis.text.x = element_text(colour = "black", size = 8, angle = 45, hjust = 1), 
      
      axis.text.y = element_text(colour = "black", size = 8), 
      
      strip.text = element_text(colour = "black", size = 10), 
      
      legend.key = element_blank(), 
      
      legend.title = element_text(colour = "black", size = 10), 
      
      legend.text = element_text(colour = "black", size = 8), 
      
      legend.position = "bottom"
      
    ) +
    
    guides(
      
      shape = guide_legend(override.aes = list(
        
        alpha = 0.75, 
        
        pch = c(21, 24),  
        
        size = 2.5, 
        
        colour = "black", 
        
        fill = "black"
        
      ))
      
    )
  
  return(tmp1)
  
}
```

```{r figS20}
# Input
figS20_input <- bc_its_funguild_relabund_input %>%
  
  filter(plant_habitat != "Root endosphere") %>%
  
  filter(
    
    (plant_habitat == "Leaf endosphere" & guild_label == "ECM") | 
      
      (plant_habitat == "Leaf endosphere" & guild_label == "Endophyte") |
      
      (plant_habitat == "Rhizosphere" & guild_label == "AMF")
    
  )

figS20 <- funguild_SI_fig(figS20_input)

# Save figure
ggsave2(
  
  filename = "data/results/figS20.png", 
  
  plot = figS20, 
  
  device = "png", 
  
  dpi = 300, 
  
  width = 6.5, 
  
  height = 5, 
  
  units = "in"
  
)

# Save figure
ggsave2(
  
  filename = "data/results/figS20.pdf", 
  
  plot = figS20, 
  
  device = "pdf", 
  
  width = 6.5, 
  
  height = 5, 
  
  units = "in"
  
)
```
